---
title: "ARE_in_hg19"
author: "Benoit Nicolet"
date: "10/9/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(plyr)
library(dplyr)
library(seqinr)
library(stringr)
library(qdapRegex)
library(dplyr)
library(Biostrings)
library(biomaRt)
library(ggplot2)

#BiocManager::install("DESeq2")
#library(motifmatchr)

knitr::opts_knit$set("/home/ben/Analysis/")
setwd("/home/ben/Analysis/")

```


```{r biomart?}

ensembl = useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
#All_filters <- listFilters(ensembl)
#View(All_filters)
#attributes_ens <- listAttributes(ensembl)
#View(attributes_ens)

IDs_genenames_coding <- getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id", "external_gene_name","gene_biotype"), mart = ensembl, filters = "with_ccds", values = TRUE)

```

## Formatting the fasta from ensembl
```{r Formatting the fasta from AURA}
# 
# UTR3_fasta <- readtext::readtext("/Users/stinky/Desktop/Sequences_info/3UTR_mart_export.txt")
# glimpse(UTR3_fasta)
# 
# UTR3_fasta <- gsub("\\\n","",UTR3_fasta) # removing the newlines
# 
# UTR3_fasta <- gsub(">","\\\n>",UTR3_fasta) ## replace > by \n>
# 
# write(UTR3_fasta,"/Users/stinky/Desktop/Sequences_info/3'UTR_ensembl_mart_export_nobreaks.txt")



```


```{r importing the formated Fasta file and processing}
## re-importing ##
#UTR3_fasta <- read.delim("/Users/stinky/Desktop/Sequences_info/3'UTR_ensembl_mart_export_nobreaks.txt",sep="\t",header = F)
UTR3_fasta <- read.delim("3'UTR_ensembl_mart_export_nobreaks.txt",sep="\t",header = F)


UTR3_fasta$gene.id <- mapply(strsplit(as.character(UTR3_fasta$V1),"\\|"),FUN=function(x){(as.character(x)[1])})
UTR3_fasta$gene.id <- gsub(">","",UTR3_fasta$gene.id)
UTR3_fasta$tx.id <- mapply(strsplit(as.character(UTR3_fasta$V1),"\\|"),FUN=function(x){(as.character(x)[2])})

UTR3_fasta$sequence <- UTR3_fasta$tx.id
UTR3_fasta$sequence <- gsub('^...............','',UTR3_fasta$sequence)

UTR3_fasta$tx.id <- gsub("A","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("T","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("G","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("C","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("ENS","ENST",UTR3_fasta$tx.id)

UTR3_fasta$V1 <- NULL


UTR3_fasta$UTR3_length <- mapply(strsplit(as.character(UTR3_fasta$sequence),","),FUN=function(x){nchar(x)})

dim(UTR3_fasta)
UTR3_fasta <- merge(UTR3_fasta, IDs_genenames_coding, by.x="tx.id",by.y="ensembl_transcript_id",all.x=T)
dim(UTR3_fasta)


UTR3_fasta <- UTR3_fasta[!UTR3_fasta$sequence=="Sequence unavailable",]

table(duplicated(UTR3_fasta$tx.id))

## oooh yeah : )

```

## Importing ATTRACT DB
```{r ATtract DB}
## ATTRACT db downloaded on June 11th 2020 from https://attract.cnic.es/index
ATTRACT <- read.delim("ATtRACT/ATtRACT_db.txt")
ATTRACT <- subset(ATTRACT,ATTRACT$Organism=="Homo_sapiens")
ATTRACT <- subset(ATTRACT,ATTRACT$Mutated=="no")
dim(ATTRACT) # 3256 motifs

## Lets filter per score 
# ATTRACT$Score <- gsub("\\*","",ATTRACT$Score)
# ATTRACT$Score <- as.numeric(ATTRACT$Score)
# ATTRACT <- subset(ATTRACT,ATTRACT$Score >=0.1)

## converting Us in Ts
ATTRACT$Motif <- gsub("U","T",ATTRACT$Motif)

table(duplicated(ATTRACT$Motif))
ATTRACT <- ATTRACT[!duplicated(ATTRACT$Motif),] ## Removing duplicates motifs
dim(ATTRACT) # 2271 motifs

```


## Counting other motifs
```{r counting other motifs}
datalist = list()
#length(ATTRACT$Motif)

{
start_time <- Sys.time()
for (i in 1:length(ATTRACT$Motif)) {
    dat <- as.vector(str_count(UTR3_fasta$sequence, pattern=as.character(ATTRACT$Motif[i])))
    datalist[[i]] <- dat # add it to your list
    ## This next part is just to return the progress (per 10% increase)
    k <- round((as.numeric(i-1)/length(ATTRACT$Motif))*100,-1)
    if (round((as.numeric(i)/length(ATTRACT$Motif))*100,-1)>k)
    {print((paste0(round((as.numeric(i)/length(ATTRACT$Motif))*100,-1),"%")))}
    else {}
    flush.console()
    }

end_time <- Sys.time()
end_time - start_time
}

UTR3_database <- do.call(cbind,datalist)

colnames(UTR3_database) <- paste0(ATTRACT$Motif)
rownames(UTR3_database) <- UTR3_fasta$tx.id
dim(UTR3_database)


UTR3_database <- UTR3_database[,order(colnames(UTR3_database))]
UTR3_database_export <- data.frame(UTR3_database)

UTR3_database_export$external_gene_name <- UTR3_fasta$external_gene_name
UTR3_database_export$tx.id <- UTR3_fasta$tx.id

write.table(UTR3_database_export,"RBP_motifs_in_human_transcriptome_3UTR_hg38_from_ATtRACT_db.csv", sep=";",dec=",", row.names =F, quote = F)

dim(UTR3_database[, colSums(UTR3_database) !=0 ]) # 2235 non-empty columns
UTR3_database <- UTR3_database[, colSums(UTR3_database) != 0]

UTR3_database <- UTR3_database[,order(colnames(UTR3_database))]

UTR3_fasta <- cbind(UTR3_fasta,UTR3_database)
#glimpse(UTR3_fasta[UTR3_fasta$external_gene_name=="IFNG",])

```


## Counting ATTTA 's
```{r Counting ATTTAs}

#UTR3_fasta$ARE_ATTTA_count <- string.counter(strings=UTR3_fasta$sequence, pattern="ATTTA")
#UTR3_fasta$ARE_ATTTA_count_homeade_counter <- string.counter(strings=UTR3_fasta$sequence, pattern="ATTTA")

UTR3_fasta$ARE_1xATTTA_count <- str_count(UTR3_fasta$sequence,pattern ="ATTTA")
UTR3_fasta$ARE_2_count <- str_count(UTR3_fasta$sequence, "ATTTATTTA")
UTR3_fasta$ARE_3_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTA")
UTR3_fasta$ARE_4_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTA")
UTR3_fasta$ARE_5_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTATTTA")
UTR3_fasta$ARE_6_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTATTTATTTA")
UTR3_fasta$ARE_7_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTATTTATTTATTTA")
UTR3_fasta$ARE_8_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTATTTATTTATTTATTTA")
UTR3_fasta$ARE_9_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTATTTATTTATTTATTTATTTA")
UTR3_fasta$ARE_10_count <- str_count(UTR3_fasta$sequence, "ATTTATTTATTTATTTATTTATTTATTTATTTATTTATTTA")


## Correcting for double counting : 

UTR3_fasta$ARE_9_count <- UTR3_fasta$ARE_9_count- 
  (UTR3_fasta$ARE_10_count*1)

UTR3_fasta$ARE_8_count <- UTR3_fasta$ARE_8_count- 
  ((UTR3_fasta$ARE_10_count*1)+
  (UTR3_fasta$ARE_9_count*1))

UTR3_fasta$ARE_7_count <- UTR3_fasta$ARE_7_count- 
  ((UTR3_fasta$ARE_10_count*1)+
  (UTR3_fasta$ARE_9_count*1)+
  (UTR3_fasta$ARE_8_count*1))

UTR3_fasta$ARE_6_count <- UTR3_fasta$ARE_6_count- 
  ((UTR3_fasta$ARE_10_count*1)+
  (UTR3_fasta$ARE_9_count*1)+
  (UTR3_fasta$ARE_8_count*1)+
  (UTR3_fasta$ARE_7_count*1))

UTR3_fasta$ARE_5_count <- UTR3_fasta$ARE_5_count- 
  ((UTR3_fasta$ARE_10_count*1)+
  (UTR3_fasta$ARE_9_count*1)+
  (UTR3_fasta$ARE_8_count*1)+
  (UTR3_fasta$ARE_7_count*1)+
  (UTR3_fasta$ARE_6_count*1))

UTR3_fasta$ARE_4_count <- UTR3_fasta$ARE_4_count- 
  ((UTR3_fasta$ARE_10_count*2)+
  (UTR3_fasta$ARE_9_count*2)+
  (UTR3_fasta$ARE_8_count*1)+
  (UTR3_fasta$ARE_7_count*1)+
  (UTR3_fasta$ARE_6_count*1)+
  (UTR3_fasta$ARE_5_count*1))

UTR3_fasta$ARE_3_count <- UTR3_fasta$ARE_3_count- 
  ((UTR3_fasta$ARE_10_count*2)+
  (UTR3_fasta$ARE_9_count*2)+
  (UTR3_fasta$ARE_8_count*2)+
  (UTR3_fasta$ARE_7_count*2)+
  (UTR3_fasta$ARE_6_count*1)+
  (UTR3_fasta$ARE_5_count*1)+
  (UTR3_fasta$ARE_4_count*1))


UTR3_fasta$ARE_2_count <- UTR3_fasta$ARE_2_count- 
  ((UTR3_fasta$ARE_10_count*3)+
  (UTR3_fasta$ARE_9_count*3)+
  (UTR3_fasta$ARE_8_count*3)+
  (UTR3_fasta$ARE_7_count*2)+
  (UTR3_fasta$ARE_6_count*2)+
  (UTR3_fasta$ARE_5_count*2)+
  (UTR3_fasta$ARE_4_count*1)+
  (UTR3_fasta$ARE_3_count*1))

UTR3_fasta$ARE_1xATTTA_count <- UTR3_fasta$ARE_1xATTTA_count- 
  ((UTR3_fasta$ARE_10_count*5)+
  (UTR3_fasta$ARE_9_count*5)+
  (UTR3_fasta$ARE_8_count*4)+
  (UTR3_fasta$ARE_7_count*4)+
  (UTR3_fasta$ARE_6_count*3)+
  (UTR3_fasta$ARE_5_count*3)+
  (UTR3_fasta$ARE_4_count*2)+
  (UTR3_fasta$ARE_3_count*2)+
  (UTR3_fasta$ARE_2_count*1))



## Now counting the counts:

UTR3_fasta$ARE_full_count <-
  (UTR3_fasta$ARE_10_count*10)+
  (UTR3_fasta$ARE_9_count*9)+
  (UTR3_fasta$ARE_8_count*8)+
  (UTR3_fasta$ARE_7_count*7)+
  (UTR3_fasta$ARE_6_count*6)+
  (UTR3_fasta$ARE_5_count*5)+
  (UTR3_fasta$ARE_4_count*4)+
  (UTR3_fasta$ARE_3_count*3)+
  (UTR3_fasta$ARE_2_count*2)+
  UTR3_fasta$ARE_1xATTTA_count



dim(UTR3_fasta) # 43,028
dim(UTR3_fasta[UTR3_fasta$ARE_full_count>0,]) # 24,720 sequences with at least 1 "ATTTA" ARE(s) 

#plot(UTR3_fasta$ARE_1xATTTA_count, UTR3_fasta$ARE_full_count)

UTR3_fasta$ARE_10_count <- NULL
UTR3_fasta$ARE_9_count <- NULL
UTR3_fasta$ARE_8_count <- NULL
UTR3_fasta$ARE_7_count <- NULL
UTR3_fasta$ARE_6_count <- NULL
UTR3_fasta$ARE_5_count <- NULL
UTR3_fasta$ARE_4_count <- NULL
UTR3_fasta$ARE_3_count <- NULL
UTR3_fasta$ARE_2_count <- NULL
UTR3_fasta$ARE_1xATTTA_count <- NULL

```


## Counting other motifs
```{r counting other motifs}

## WAUUUAW
UTR3_fasta$ARE_WAUUUAW <- str_count(UTR3_fasta$sequence,pattern = c("TATTTAT","AATTTAA","TATTTAA","AATTTAT"))

## WWAUUUAWW
UTR3_fasta$ARE_WWAUUUAWW <- str_count(UTR3_fasta$sequence,pattern = c("ATATTTATA","ATATTTATT","TTATTTATA","TTATTTATT","AAATTTAAA","AAATTTAAT","TAATTTAAA","TAATTTAAT","ATATTTAAA","ATATTTAAT","TTATTTAAA","TTATTTAAT","AAATTTATA","AAATTTATT","TAATTTATA","TAATTTATT"))

```

## G-C content
```{r G-C content}

UTR3_fasta$G_count <- str_count(UTR3_fasta$sequence,pattern ="G")
UTR3_fasta$C_count <- str_count(UTR3_fasta$sequence,pattern ="C")

UTR3_fasta$GC_percentage <- ((UTR3_fasta$G_count + UTR3_fasta$C_count)/UTR3_fasta$UTR3_length)*100

UTR3_fasta$G_count <- NULL
UTR3_fasta$C_count <- NULL

```



## distance between AREs (ATTTA)
```{r distance between AREs}

UTR3_fasta$seq_in_between_AREs <- rm_between(UTR3_fasta$sequence, 'ATTTA', 'ATTTA', extract=TRUE)



UTR3_fasta$dist_between_AREs <- mapply(strsplit(as.character(UTR3_fasta$seq_in_between_AREs),", "),FUN=function(x){nchar(x)})

UTR3_fasta$dist_between_AREs <- gsub("c\\(","",UTR3_fasta$dist_between_AREs)
UTR3_fasta$dist_between_AREs <- gsub("\\)","",UTR3_fasta$dist_between_AREs)

UTR3_fasta$dist_between_AREs <- gsub("NA",0, UTR3_fasta$dist_between_AREs)

UTR3_fasta$max_dist_between_AREs <- mapply(strsplit(as.character(UTR3_fasta$dist_between_AREs), ","), FUN=function(x){max(as.numeric(x))})
UTR3_fasta$min_dist_between_AREs <- mapply(strsplit(as.character(UTR3_fasta$dist_between_AREs), ","), FUN=function(x){min(as.numeric(x))})

ARE_dist_split <- strsplit(UTR3_fasta$dist_between_AREs, split = ",", fixed = TRUE)


UTR3_fasta$median_dist_between_AREs <- sapply(ARE_dist_split, function(x) median(as.numeric(x)))
UTR3_fasta$mean_dist_between_AREs <- sapply(ARE_dist_split, function(x) mean(as.numeric(x)))

#UTR3_fasta$dist_between_AREs <- NULL
#UTR3_fasta$seq_in_between_AREs <- NULL


#AREs_motifs<- as.vector("ATTTA","TATTTAT","AATTTAA","TATTTAA","AATTTAT","ATATTTATA","ATATTTATT","TTATTTATA","TTATTTATT","AAATTTAAA","AAATTTAAT","TAATTTAAA","TAATTTAAT","ATATTTAAA","ATATTTAAT","TTATTTAAA","TTATTTAAT","AAATTTATA","AAATTTATT","TAATTTATA","TAATTTATT")


ggplot(UTR3_fasta,aes(x=log10(UTR3_fasta$mean_dist_between_AREs)))+
  geom_density()

ggplot(UTR3_fasta,aes(x=log10(UTR3_fasta$median_dist_between_AREs)))+
  geom_density()

ggplot(UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$max_dist_between_AREs))))+
  geom_density()

ggplot(UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$min_dist_between_AREs))))+
  geom_density()

```


##
```{r making tables pretty for export}
Gene_wise_3UTR_list <- UTR3_fasta
  
Gene_wise_3UTR_list$sequence <- NULL
Gene_wise_3UTR_list$gene_biotype <- NULL
Gene_wise_3UTR_list$ARE_ATTTA_count <- NULL

colnames(Gene_wise_3UTR_list) <- paste0(colnames(Gene_wise_3UTR_list),"_UTR3")
registerDoMC(4)
Gene_wise_3UTR_list <- ddply(Gene_wise_3UTR_list,"external_gene_name", numcolwise(mean), .parallel = T, .progress = T)
#Gene_wise_3UTR_list$ARE_counts <- ceiling(Gene_wise_3UTR_list$ARE_counts)
dim(Gene_wise_3UTR_list) # 18708  2243

write.table(Gene_wise_3UTR_list,"3UTR_motifs_list_from_sequences.csv",sep = ";", dec = ",",row.names = F)

```


```{r AGAAGA motifs count}
UTR3_fasta$ATTTA <- str_count(UTR3_fasta$sequence,pattern ="ATTTA")

UTR3_fasta$AGAAGA <- str_count(UTR3_fasta$sequence,pattern ="AGAAGA")
UTR3_fasta$.GAAGA <- str_count(UTR3_fasta$sequence,pattern =".GAAGA")
UTR3_fasta$AGAAG. <- str_count(UTR3_fasta$sequence,pattern ="AGAAG.")
UTR3_fasta$AGAA <- str_count(UTR3_fasta$sequence,pattern ="AGAA")


# UTR3_fasta <- subset(UTR3_fasta,UTR3_fasta$AGAAGA>0|
#                        UTR3_fasta$.GAAGA>0|
#                        UTR3_fasta$AGAAG.>0|
#                        UTR3_fasta$AGAA>0)

UTR3_fasta$seq_in_between_AGAAGA <- rm_between(UTR3_fasta$sequence, 'AGAAGA', 'AGAAGA', extract=TRUE)



UTR3_fasta$dist_between_AGAAGA <- mapply(strsplit(as.character(UTR3_fasta$seq_in_between_AGAAGA),", "),FUN=function(x){nchar(x)})

UTR3_fasta$dist_between_AGAAGA <- gsub("c\\(","",UTR3_fasta$dist_between_AGAAGA)
UTR3_fasta$dist_between_AGAAGA <- gsub("\\)","",UTR3_fasta$dist_between_AGAAGA)

UTR3_fasta$dist_between_AGAAGA <- gsub("NA",0, UTR3_fasta$dist_between_AGAAGA)

UTR3_fasta$max_dist_between_AGAAGA <- mapply(strsplit(as.character(UTR3_fasta$dist_between_AGAAGA), ","), FUN=function(x){max(as.numeric(x))})
UTR3_fasta$min_dist_between_AGAAGA <- mapply(strsplit(as.character(UTR3_fasta$dist_between_AGAAGA), ","), FUN=function(x){min(as.numeric(x))})

UTR3_fasta$max_dist_between_AGAAGA <- gsub("integer\\(0",0,UTR3_fasta$max_dist_between_AGAAGA)
UTR3_fasta$min_dist_between_AGAAGA <- gsub("integer\\(0",0,UTR3_fasta$min_dist_between_AGAAGA)

AGAAGA_dist_split <- strsplit(UTR3_fasta$dist_between_AGAAGA, split = ",", fixed = TRUE)


UTR3_fasta$median_dist_between_AGAAGA <- sapply(AGAAGA_dist_split, function(x) median(as.numeric(x)))
UTR3_fasta$mean_dist_between_AGAAGA <- sapply(AGAAGA_dist_split, function(x) mean(as.numeric(x)))



ggplot(UTR3_fasta,aes(x=log10(UTR3_fasta$mean_dist_between_AGAAGA)))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(UTR3_fasta$mean_dist_between_AREs)),color="red")

ggplot(UTR3_fasta,aes(x=log10(UTR3_fasta$median_dist_between_AGAAGA)))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(UTR3_fasta$median_dist_between_AREs)),color="red")

ggplot(UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$max_dist_between_AGAAGA))))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$max_dist_between_AREs))),color="red")

ggplot(UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$min_dist_between_AGAAGA))))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$min_dist_between_AREs))),color="red")


ggplot(UTR3_fasta,aes(x=(as.numeric(UTR3_fasta$AGAAGA)),y=(as.numeric(UTR3_fasta$ATTTA))))+
  geom_point()+
  geom_smooth()+
  scale_x_continuous(limits = c(0,110))+
  scale_y_continuous(limits = c(0,110))

UTR3_fasta_AGE <- UTR3_fasta
UTR3_fasta_AGE <- subset(UTR3_fasta_AGE, UTR3_fasta_AGE$AGAAGA>0 |UTR3_fasta_AGE$AGAAG.>0|UTR3_fasta_AGE$.GAAGA>0)
dim(UTR3_fasta_AGE)
dim(UTR3_fasta)

UTR3_fasta_AGE$sequence <- NULL
UTR3_fasta_AGE$gene.id <- NULL
UTR3_fasta_AGE$seq_in_between_AREs <- NULL
UTR3_fasta_AGE$seq_in_between_AGAAGA <- NULL

write.table(UTR3_fasta_AGE,"AGE_motifs_occurence_3UTR_human.csv",sep = ";", dec = ",",row.names = F)

dim(subset(UTR3_fasta_AGE, UTR3_fasta_AGE$AGAAGA>0)) # 14452
dim(UTR3_fasta) #41580

14452/41580*100 # 34.75709% of all transcriptome has >0 AGE motifs : ) 

```



```{r AGAAGA motifs count}
UTR3_fasta$CTCAGGT <- str_count(UTR3_fasta$sequence,pattern ="CTCAGGT")

UTR3_fasta <- subset(UTR3_fasta,UTR3_fasta$CTCAGGT>0)

UTR3_fasta$seq_in_between_AGAAGA <- rm_between(UTR3_fasta$sequence, 'AGAAGA', 'AGAAGA', extract=TRUE)



UTR3_fasta$dist_between_AGAAGA <- mapply(strsplit(as.character(UTR3_fasta$seq_in_between_AGAAGA),", "),FUN=function(x){nchar(x)})

UTR3_fasta$dist_between_AGAAGA <- gsub("c\\(","",UTR3_fasta$dist_between_AGAAGA)
UTR3_fasta$dist_between_AGAAGA <- gsub("\\)","",UTR3_fasta$dist_between_AGAAGA)

UTR3_fasta$dist_between_AGAAGA <- gsub("NA",0, UTR3_fasta$dist_between_AGAAGA)

UTR3_fasta$max_dist_between_AGAAGA <- mapply(strsplit(as.character(UTR3_fasta$dist_between_AGAAGA), ", "), FUN=function(x){max(x)})
UTR3_fasta$min_dist_between_AGAAGA <- mapply(strsplit(as.character(UTR3_fasta$dist_between_AGAAGA), ", "), FUN=function(x){min(x)})

UTR3_fasta$max_dist_between_AGAAGA <- gsub("integer\\(0",0,UTR3_fasta$max_dist_between_AGAAGA)
UTR3_fasta$min_dist_between_AGAAGA <- gsub("integer\\(0",0,UTR3_fasta$min_dist_between_AGAAGA)

AGAAGA_dist_split <- strsplit(UTR3_fasta$dist_between_AGAAGA, split = ", ", fixed = TRUE)


UTR3_fasta$median_dist_between_AGAAGA <- sapply(AGAAGA_dist_split, function(x) median(as.numeric(x)))
UTR3_fasta$mean_dist_between_AGAAGA <- sapply(AGAAGA_dist_split, function(x) mean(as.numeric(x)))



ggplot(UTR3_fasta,aes(x=log10(UTR3_fasta$mean_dist_between_AGAAGA)))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(UTR3_fasta$mean_dist_between_AREs)),color="red")

ggplot(UTR3_fasta,aes(x=log10(UTR3_fasta$median_dist_between_AGAAGA)))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(UTR3_fasta$median_dist_between_AREs)),color="red")

ggplot(UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$max_dist_between_AGAAGA))))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$max_dist_between_AREs))),color="red")

ggplot(UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$min_dist_between_AGAAGA))))+
  geom_density()+
  geom_density(data=UTR3_fasta,aes(x=log10(as.numeric(UTR3_fasta$min_dist_between_AREs))),color="red")


write.table(Gene_wise_3UTR_list,"3UTR_motifs_list_from_sequences.csv",sep = ";", dec = ",",row.names = F)

```

```{r new bits to isolate potential target}

UTR3_fasta$CTTTCTT <- str_count(UTR3_fasta$sequence,pattern ="CTTTCTT")
UTR3_fasta$CTTT <- str_count(UTR3_fasta$sequence,pattern ="CTTT")
UTR3_fasta$AGAAGA <- str_count(UTR3_fasta$sequence,pattern ="AGAAGA")
UTR3_fasta$TCTTCT <- str_count(UTR3_fasta$sequence,pattern ="TCTTCT")
UTR3_fasta$AGGAGA <- str_count(UTR3_fasta$sequence,pattern ="AGGAGA")
UTR3_fasta$CTCAGGT <- str_count(UTR3_fasta$sequence,pattern ="CTCAGGT")
UTR3_fasta$ATTTA <- str_count(UTR3_fasta$sequence,pattern ="ATTTA")

UTR3_fasta <- subset(UTR3_fasta,
       UTR3_fasta$CTTTCTT>0 |
       UTR3_fasta$CTTT>0 |
       UTR3_fasta$AGAAGA>0 |
       UTR3_fasta$AGGAGA>0 |
       UTR3_fasta$CTCAGGT>0 |
       UTR3_fasta$ATTTA>0 )

#write.table(UTR3_fasta,"/home/ben/Analysis/RF_human/3UTR_motifs_for_selection_110321.csv",sep = ";", dec = ",",row.names = F)


ggplot(UTR3_fasta,aes(x=AGAAGA,y=TCTTCT))+
  geom_point()+
  scale_x_continuous(limits = c(0,30))+
  scale_y_continuous(limits = c(0,30))+
  theme(aspect.ratio = 1)



ggplot(UTR3_fasta,aes(x=UTR3_length,y=AGAAGA))+
  geom_point()+
  #scale_x_continuous(limits = c(0,30))+
  scale_y_continuous(limits = c(0,30))+
  theme(aspect.ratio = 1)

ggplot(UTR3_fasta,aes(x=UTR3_length,y=ATTTA))+
  geom_point()+
  #scale_x_continuous(limits = c(0,30))+
  #scale_y_continuous(limits = c(0,30))+
  theme(aspect.ratio = 1)


ggplot(UTR3_fasta,aes(x=AGAAGA,y=ATTTA))+
  geom_point()+
  #scale_x_continuous(limits = c(0,30))+
  #scale_y_continuous(limits = c(0,30))+
  theme(aspect.ratio = 1)


```

