---
title: "Visualizing XGB output in immune subsets"
author: "Benoit Nicolet"
date: "03/05/2022"
output: html_document
---


```{r setup, include=FALSE}

#install.packages("pheatmap")

library(plyr)
library(dplyr)
library(ggplot2)
library(ggfortify)
library(tidyverse)
library(circlize)
library(pheatmap)
library(caret)
library(ggpointdensity)
library(viridis)
library(reshape)
library(limma)


knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set("/home/ben/Analysis/RF_human/immune_system/")
setwd("/home/ben/Analysis/RF_human/immune_system/")

```


```{r importing RNA library}

# Sequence_parameters <- read.delim("/home/ben/Analysis/RF_human/Library_V2_Aug2021/RNA_library_v2_RBP_GC_length_codon_AA_m6A_m5C_AtoI_m1A_m7G_CD8miRDB.csv",sep = ";", dec=",")

```



## RNA Models ##

```{r importing RNA models}

CD8_Tn_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD8_Tn_14_10_2021.RDS")
CD8_Tcm_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD8_Tcm_14_10_2021.RDS")
CD8_Tem_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD8_Tem_14_10_2021.RDS")
CD8_Teff_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD8_Teff_10k_15_10_2021.RDS")

CD4_Tn_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD4_Tn_14_10_2021.RDS")
CD4_Tcm_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD4_Tcm_14_10_2021.RDS")
CD4_Tem_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_10knround_TPM_CD4_Tem_14_10_2021.RDS")

CD8_Tn_RNA<- data.frame(caret::varImp(CD8_Tn_RNA,scale=F)$importance)
CD8_Tcm_RNA<- data.frame(caret::varImp(CD8_Tcm_RNA,scale=F)$importance)
CD8_Tem_RNA<- data.frame(caret::varImp(CD8_Tem_RNA,scale=F)$importance)
CD8_Teff_RNA<- data.frame(caret::varImp(CD8_Teff_RNA,scale=F)$importance)
CD4_Tn_RNA<- data.frame(caret::varImp(CD4_Tn_RNA,scale=F)$importance)
CD4_Tcm_RNA<- data.frame(caret::varImp(CD4_Tcm_RNA,scale=F)$importance)
CD4_Tem_RNA<- data.frame(caret::varImp(CD4_Tem_RNA,scale=F)$importance)


B_naive_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Bnaive_03_05_2022.RDS")
B_mem_S_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_B_S_mem_03_05_2022.RDS")
B_mem_NS_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_B_NS_mem_03_05_2022.RDS")
B_plasma_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_B_plasma_03_05_2022.RDS")

Basophils_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Basophils_03_05_2022.RDS")
Neutrophils_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Neutrophils_03_05_2022.RDS")

mDC_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_mDC_03_05_2022.RDS")
pDC_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_pDC_03_05_2022.RDS")


B_naive_RNA<- data.frame(caret::varImp(B_naive_RNA,scale=F)$importance)
B_mem_S_RNA<- data.frame(caret::varImp(B_mem_S_RNA,scale=F)$importance)
B_mem_NS_RNA<- data.frame(caret::varImp(B_mem_NS_RNA,scale=F)$importance)
B_plasma_RNA<- data.frame(caret::varImp(B_plasma_RNA,scale=F)$importance)
Basophils_RNA<- data.frame(caret::varImp(Basophils_RNA,scale=F)$importance)
Neutrophils_RNA<- data.frame(caret::varImp(Neutrophils_RNA,scale=F)$importance)
mDC_RNA<- data.frame(caret::varImp(mDC_RNA,scale=F)$importance)
pDC_RNA<- data.frame(caret::varImp(pDC_RNA,scale=F)$importance)

Mono_classical_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Mono_classical_03_05_2022.RDS")
Mono_intermediate_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Mono_inter_03_05_2022.RDS")
Mono_non_classical_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Mono_non_classical_03_05_2022.RDS")

Tregs_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_Tregs_03_05_2022.RDS")
NK_RNA <- readRDS("/home/ben/Analysis/RF_human/immune_system/RNA/models/xgb_TPM_NK_03_05_2022.RDS")

Mono_classical_RNA<- data.frame(caret::varImp(Mono_classical_RNA,scale=F)$importance)
Mono_intermediate_RNA<- data.frame(caret::varImp(Mono_intermediate_RNA,scale=F)$importance)
Mono_non_classical_RNA<- data.frame(caret::varImp(Mono_non_classical_RNA,scale=F)$importance)
Tregs_RNA<- data.frame(caret::varImp(Tregs_RNA,scale=F)$importance)
NK_RNA<- data.frame(caret::varImp(NK_RNA,scale=F)$importance)


CD8_Tn_RNA$ID <- rownames(CD8_Tn_RNA)
CD8_Tcm_RNA$ID <- rownames(CD8_Tcm_RNA)
CD8_Tem_RNA$ID <- rownames(CD8_Tem_RNA)
CD8_Teff_RNA$ID <- rownames(CD8_Teff_RNA)
CD4_Tn_RNA$ID <- rownames(CD4_Tn_RNA)
CD4_Tcm_RNA$ID <- rownames(CD4_Tcm_RNA)
CD4_Tem_RNA$ID <- rownames(CD4_Tem_RNA)

B_naive_RNA$ID <- rownames(B_naive_RNA)
B_mem_S_RNA$ID <- rownames(B_mem_S_RNA)
B_mem_NS_RNA$ID <- rownames(B_mem_NS_RNA)
B_plasma_RNA$ID <- rownames(B_plasma_RNA)
Basophils_RNA$ID <- rownames(Basophils_RNA)
Neutrophils_RNA$ID <- rownames(Neutrophils_RNA)
mDC_RNA$ID <- rownames(mDC_RNA)
pDC_RNA$ID <- rownames(pDC_RNA)
Mono_classical_RNA$ID <- rownames(Mono_classical_RNA)
Mono_intermediate_RNA$ID <- rownames(Mono_intermediate_RNA)
Mono_non_classical_RNA$ID <- rownames(Mono_non_classical_RNA)
Tregs_RNA$ID <- rownames(Tregs_RNA)
NK_RNA$ID <- rownames(NK_RNA)





colnames(CD8_Tn_RNA)[1] <- "CD8_Tn_RNA"
colnames(CD8_Tcm_RNA)[1] <- "CD8_Tcm_RNA"
colnames(CD8_Tem_RNA)[1] <- "CD8_Tem_RNA"
colnames(CD8_Teff_RNA)[1] <- "CD8_Teff_RNA"
colnames(CD4_Tn_RNA)[1] <- "CD4_Tn_RNA"
colnames(CD4_Tcm_RNA)[1] <- "CD4_Tcm_RNA"
colnames(CD4_Tem_RNA)[1] <- "CD4_Tem_RNA"

colnames(B_naive_RNA)[1] <- "B_naive_RNA"
colnames(B_mem_S_RNA)[1] <- "B_mem_S_RNA"
colnames(B_mem_NS_RNA)[1] <- "B_mem_NS_RNA"
colnames(B_plasma_RNA)[1] <- "B_plasma_RNA"
colnames(Basophils_RNA)[1] <- "Basophils_RNA"
colnames(Neutrophils_RNA)[1] <- "Neutrophils_RNA"
colnames(mDC_RNA)[1] <- "mDC_RNA"
colnames(pDC_RNA)[1] <- "pDC_RNA"
colnames(Mono_classical_RNA)[1] <- "Mono_classical_RNA"
colnames(Mono_intermediate_RNA)[1] <- "Mono_intermediate_RNA"
colnames(Mono_non_classical_RNA)[1] <- "Mono_non_classical_RNA"
colnames(Tregs_RNA)[1] <- "Tregs_RNA"
colnames(NK_RNA)[1] <- "NK_RNA"


RNA_subsets_models <- list(CD8_Tn_RNA,CD8_Tcm_RNA,CD8_Tem_RNA,CD8_Teff_RNA,CD4_Tn_RNA,CD4_Tcm_RNA,CD4_Tem_RNA,
                          Tregs_RNA,
                          NK_RNA,
                          B_naive_RNA,B_mem_NS_RNA,B_mem_S_RNA,B_plasma_RNA,
                          Basophils_RNA,Neutrophils_RNA,
                          Mono_classical_RNA,Mono_intermediate_RNA,Mono_non_classical_RNA,
                          mDC_RNA,pDC_RNA)

RNA_subsets_models <- plyr::join_all(RNA_subsets_models,by = "ID")
RNA_subsets_models[1:2] <- RNA_subsets_models[2:1]
colnames(RNA_subsets_models)[1:2] <- colnames(RNA_subsets_models)[2:1]
dim(RNA_subsets_models)



## median correction ##
# RNA_subsets_models <- data.frame(RNA_subsets_models[1:1],scale(RNA_subsets_models[2:21],center = F))

RNA_subsets_models[2:21] <- log10(RNA_subsets_models[2:21])
RNA_subsets_models <- do.call(data.frame,lapply(RNA_subsets_models,function(x) replace(x, is.infinite(x),NA)))
RNA_subsets_models[is.na(RNA_subsets_models)]=min(RNA_subsets_models[2:21],na.rm = T)
RNA_subsets_models[2:21] <- RNA_subsets_models[2:21] - min(RNA_subsets_models[2:21])

# rownames(RNA_subsets_models) <- RNA_subsets_models$ID
# 
# RNA_subsets_models <- as.matrix(RNA_subsets_models[2:21])
# 
# RNA_subsets_models <- limma::voom(RNA_subsets_models,design = NULL)
# RNA_subsets_models <- data.frame(RNA_subsets_models)
# RNA_subsets_models <- data.frame("ID"=rownames(RNA_subsets_models),RNA_subsets_models)



# RNA_subsets_models_median_cor <- RNA_subsets_models
# RNA_subsets_models_median_cor[,2:21] <- 2^RNA_subsets_models_median_cor[,2:21]
# RNA_subsets_models_median_cor[is.na(RNA_subsets_models_median_cor)]=0
# RNA_subsets_models_median_cor[,2:21] <- RNA_subsets_models_median_cor[,2:21] - mapply(median,RNA_subsets_models_median_cor[,2:21])
# RNA_subsets_models_median_cor[,2:21] <- log2(RNA_subsets_models_median_cor[,2:21])
# RNA_subsets_models_median_cor <- do.call(data.frame,lapply(RNA_subsets_models_median_cor,function(x) replace(x, is.infinite(x),NA)))
# RNA_subsets_models_median_cor[is.na(RNA_subsets_models_median_cor)]=min(RNA_subsets_models_median_cor[,2:21],na.rm = T)
# RNA_subsets_models <- RNA_subsets_models_median_cor

#RNA_subsets_models[is.na(RNA_subsets_models)]=min(RNA_subsets_models[2:21],na.rm = T)



dim(RNA_subsets_models)
RNA_subsets_models$region <- "Other"
RNA_subsets_models$region[grep("CDS", RNA_subsets_models$ID)] <- "CDS"
RNA_subsets_models$region[grep("UTR5", RNA_subsets_models$ID)] <- "UTR5"
RNA_subsets_models$region[grep("UTR3", RNA_subsets_models$ID)] <- "UTR3"
#RNA_subsets_models$region[grep("codon", RNA_subsets_models$ID)] <- "codon"

dim(RNA_subsets_models)


RNA_subsets_models_CDS <- subset(RNA_subsets_models, RNA_subsets_models$region=="CDS")
dim(RNA_subsets_models_CDS)

RNA_subsets_models_UTR5 <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR5")
dim(RNA_subsets_models_UTR5)

RNA_subsets_models_UTR3 <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")
dim(RNA_subsets_models_UTR3)

RNA_subsets_models_Other <- subset(RNA_subsets_models, RNA_subsets_models$region=="Other")
dim(RNA_subsets_models_Other)


```




```{r corplot}
# RNA_subsets_models_no_NA <- RNA_subsets_models
# RNA_subsets_models_no_NA[is.na(RNA_subsets_models_no_NA)]=min(RNA_subsets_models_no_NA[2:21],na.rm = T)

# col_fun= colorRamp2(c(-3, -1.5, 0, 1.5, 3), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))
col_fun= colorRamp2(c(-3, -1, 0, 0.5, 2), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))


corrplot::corrplot(cor(RNA_subsets_models[,2:21],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "full lib",mar = c(0,0,1,0))

corrplot::corrplot(cor(RNA_subsets_models_UTR5[,2:21],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "5'UTR",mar = c(0,0,1,0))

corrplot::corrplot(cor(RNA_subsets_models_CDS[,2:21],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "CDS",mar = c(0,0,1,0))


corrplot::corrplot(cor(RNA_subsets_models_UTR3[,2:21],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "3'UTR",mar = c(0,0,1,0))

corrplot::corrplot(cor(RNA_subsets_models_Other[,2:21],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "other",mar = c(0,0,1,0))

# cor(RNA_subsets_models[,2:8],method = "pearson",use = "pairwise.complete.obs")

```



```{r corplot and heatmap on top features}



RNA_subsets_models_CDS_top <- RNA_subsets_models_CDS
RNA_subsets_models_CDS_top[2:21] <- 10^(RNA_subsets_models_CDS_top[2:21])
RNA_subsets_models_CDS_top$rowmean <- rowMeans(RNA_subsets_models_CDS_top[,2:21])
RNA_subsets_models_CDS_top <- RNA_subsets_models_CDS_top[order(RNA_subsets_models_CDS_top$rowmean,decreasing = T),]
RNA_subsets_models_CDS_top <- RNA_subsets_models_CDS_top[1:100,]
RNA_subsets_models_CDS_top[2:21] <- log10(RNA_subsets_models_CDS_top[2:21])
RNA_subsets_models_CDS_top <- do.call(data.frame,lapply(RNA_subsets_models_CDS_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_CDS_top)
#RNA_subsets_models_CDS_top$rowmean <- NULL

RNA_subsets_models_UTR5_top <- RNA_subsets_models_UTR5
RNA_subsets_models_UTR5_top[2:21] <- 10^(RNA_subsets_models_UTR5_top[2:21])
RNA_subsets_models_UTR5_top$rowmean <- rowMeans(RNA_subsets_models_UTR5_top[,2:21])
RNA_subsets_models_UTR5_top <- RNA_subsets_models_UTR5_top[order(RNA_subsets_models_UTR5_top$rowmean,decreasing = T),]
RNA_subsets_models_UTR5_top <- RNA_subsets_models_UTR5_top[1:100,]
RNA_subsets_models_UTR5_top[2:21] <- log10(RNA_subsets_models_UTR5_top[2:21])
RNA_subsets_models_UTR5_top <- do.call(data.frame,lapply(RNA_subsets_models_UTR5_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_UTR5_top)


RNA_subsets_models_UTR3_top <- RNA_subsets_models_UTR3
RNA_subsets_models_UTR3_top[2:21] <- 10^(RNA_subsets_models_UTR3_top[2:21])
RNA_subsets_models_UTR3_top$rowmean <- rowMeans(RNA_subsets_models_UTR3_top[,2:21])
RNA_subsets_models_UTR3_top <- RNA_subsets_models_UTR3_top[order(RNA_subsets_models_UTR3_top$rowmean,decreasing = T),]
RNA_subsets_models_UTR3_top <- RNA_subsets_models_UTR3_top[1:100,]
RNA_subsets_models_UTR3_top[2:21] <- log10(RNA_subsets_models_UTR3_top[2:21])
RNA_subsets_models_UTR3_top <- do.call(data.frame,lapply(RNA_subsets_models_UTR3_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_UTR3_top)


RNA_subsets_models_Other_top <- RNA_subsets_models_Other
RNA_subsets_models_Other_top[2:21] <- 10^(RNA_subsets_models_Other_top[2:21])
RNA_subsets_models_Other_top$rowmean <- rowMeans(RNA_subsets_models_Other_top[,2:21])
RNA_subsets_models_Other_top <- RNA_subsets_models_Other_top[order(RNA_subsets_models_Other_top$rowmean,decreasing = T),]
RNA_subsets_models_Other_top <- RNA_subsets_models_Other_top[1:25,]
RNA_subsets_models_Other_top[2:21] <- log10(RNA_subsets_models_Other_top[2:21])
RNA_subsets_models_Other_top <- do.call(data.frame,lapply(RNA_subsets_models_Other_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_Other_top)


## Heatmap time ##
col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(RNA_subsets_models_Other_top[2:21], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_Other_top[2:21]), fontsize_row=4,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_subsets_models_Other$ID, cellwidth = 8, main = "general",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RNA_subsets_models_CDS_top[2:21], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_CDS_top[2:21]), fontsize_row=3 ,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_subsets_models_CDS$ID,cellwidth = 8,main = "CDS",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RNA_subsets_models_UTR5_top[2:21], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_UTR5_top[2:21]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_subsets_models_UTR5$ID,cellwidth = 8,main = "5'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RNA_subsets_models_UTR3_top[2:21], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_UTR3_top[2:21]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_subsets_models_UTR3$ID, cellwidth = 8,main = "3'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)



```






```{r importance per region}

RNA_subsets_models_for_imp_plot <- RNA_subsets_models
# RNA_subsets_models_for_imp_plot[RNA_subsets_models_for_imp_plot==min(RNA_subsets_models_for_imp_plot[,2:21])]=NA
# RNA_subsets_models_for_imp_plot[2:21] <- RNA_subsets_models_for_imp_plot[2:21] - min(RNA_subsets_models_for_imp_plot[2:21])
RNA_subsets_models_for_imp_plot <- melt(RNA_subsets_models_for_imp_plot)


ggplot(RNA_subsets_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 1)


## average feature imp ##
RNA_subsets_models_avg_imp <- data.frame("ID"=RNA_subsets_models$ID,"rowmean"=rowMeans(RNA_subsets_models[,2:21]),"region"=RNA_subsets_models$region)
RNA_subsets_models_avg_imp <- melt(RNA_subsets_models_avg_imp)


ggplot(RNA_subsets_models_avg_imp,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = region), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.8), size = 0.75, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.4, color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.4, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 4)


```



```{r capturing the variance of regions}
RNA_subsets_models_var <- data.frame(RNA_subsets_models)
RNA_subsets_models_var[2:21] <- 10^(RNA_subsets_models_var[2:21])
# RNA_subsets_models_var[RNA_subsets_models_var==min(RNA_subsets_models_var[,2:21],na.rm = T)]=NA

RNA_subsets_models_var[is.na(RNA_subsets_models_var)]=0
RNA_subsets_models_var$row_mean <- rowMeans(as.matrix(RNA_subsets_models_var[2:21]))

RNA_subsets_models_var$row_sd <- matrixStats::rowSds(as.matrix(RNA_subsets_models_var[2:21]))
RNA_subsets_models_var$corrected_sd <- RNA_subsets_models_var$row_sd/RNA_subsets_models_var$row_mean
RNA_subsets_models_var$row_mean <- log10(RNA_subsets_models_var$row_mean)
RNA_subsets_models_var[2:21] <- log10(RNA_subsets_models_var[2:21])

plot(RNA_subsets_models_var$row_sd,RNA_subsets_models_var$row_mean)
plot(RNA_subsets_models_var$corrected_sd,RNA_subsets_models_var$row_mean)

ggplot(RNA_subsets_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RNA_subsets_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  theme(aspect.ratio = 0.5)

ggplot(RNA_subsets_models_var,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

RNA_subsets_models_var$region <- factor(RNA_subsets_models_var$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(RNA_subsets_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(RNA_subsets_models_var$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)



mean(RNA_subsets_models_var[RNA_subsets_models_var$region=="UTR5",]$corrected_sd,na.rm = T) # = -2.38509
mean(RNA_subsets_models_var[RNA_subsets_models_var$region=="CDS",]$corrected_sd,na.rm = T) # = -2.031128
mean(RNA_subsets_models_var[RNA_subsets_models_var$region=="UTR3",]$corrected_sd,na.rm = T) # = -2.061535
mean(RNA_subsets_models_var[RNA_subsets_models_var$region=="Other",]$corrected_sd,na.rm = T) # = -2.781053

median(RNA_subsets_models_var$corrected_sd,na.rm = T)

table(RNA_subsets_models_var[RNA_subsets_models_var$corrected_sd>1.5,]$region)/table(RNA_subsets_models_var$region)*100



```
```{r core vs variable features}

hist(RNA_subsets_models_var$corrected_sd)
median(RNA_subsets_models_var$corrected_sd,na.rm = T)
mean(RNA_subsets_models_var$corrected_sd,na.rm = T)

core_features_RNA <- data.frame(table(RNA_subsets_models_var[RNA_subsets_models_var$corrected_sd<mean(RNA_subsets_models_var$corrected_sd,na.rm = T),]$region)/table(RNA_subsets_models_var$region)*100)
core_features_RNA$group <- "B_core"

core_features_RNA <- rbind(core_features_RNA, data.frame("Var1"=c("UTR5","CDS","UTR3","Other"),"Freq"= 100-core_features_RNA$Freq,"group"="A_variable"))


ggplot(core_features_RNA,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  scale_y_continuous(expand = c(0,0))+
  theme_minimal()+
  theme(aspect.ratio = 2)
  


```




```{r full variablity assessment}

RNA_subsets_models_var$row_min <- do.call(pmin, as.data.frame(RNA_subsets_models_var[2:21]))
table(RNA_subsets_models_var$row_min==0)

dim(RNA_subsets_models_var[(RNA_subsets_models_var$corrected_sd < median(RNA_subsets_models_var$corrected_sd,na.rm = T)) & (RNA_subsets_models_var$row_min>0),])

dim(RNA_subsets_models_var[(RNA_subsets_models_var$row_min>0),])

core_features_RNA <- data.frame(
  table(RNA_subsets_models_var[(RNA_subsets_models_var$corrected_sd < median(RNA_subsets_models_var$corrected_sd,na.rm = T)) & (RNA_subsets_models_var$row_min>0),]$region)/table(RNA_subsets_models_var$region)*100)
core_features_RNA$group <- "C_core"

HV_core_features_RNA <- data.frame(
  table(RNA_subsets_models_var[(RNA_subsets_models_var$corrected_sd < median(RNA_subsets_models_var$corrected_sd,na.rm = T)) & (RNA_subsets_models_var$row_min==0),]$region)/table(RNA_subsets_models_var$region)*100)
HV_core_features_RNA$group <- "A_hyper_variable"

core_features_RNA <- rbind(core_features_RNA, HV_core_features_RNA, data.frame("Var1"=c("UTR5","CDS","UTR3","Other"),"Freq"= 100-(core_features_RNA$Freq + HV_core_features_RNA$Freq),"group"="B_variable"))


ggplot(core_features_RNA,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  # scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)
  

ggplot(RNA_subsets_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point()+
  theme(aspect.ratio = 1)


```


```{r capturing the variance of regions}
non_zero_RNA_subsets_models <-RNA_subsets_models_var

non_zero_RNA_subsets_models <- subset(non_zero_RNA_subsets_models,non_zero_RNA_subsets_models$row_min>0)
dim(non_zero_RNA_subsets_models)

ggplot(non_zero_RNA_subsets_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(non_zero_RNA_subsets_models,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

non_zero_RNA_subsets_models$region <- factor(non_zero_RNA_subsets_models$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(non_zero_RNA_subsets_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(non_zero_RNA_subsets_models$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)






```





```{r }
RNA_subsets_models_codons <- RNA_subsets_models
RNA_subsets_models_codons$region[grep("codon", RNA_subsets_models_codons$ID)] <- "codon"

RNA_subsets_models_codons <- subset(RNA_subsets_models_codons, RNA_subsets_models_codons$region=="codon")

dim(RNA_subsets_models_codons)


#col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green
col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

RNA_subsets_models_codons <- RNA_subsets_models_codons[order(RNA_subsets_models_codons$ID),]


pheatmap(RNA_subsets_models_codons[2:21], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_codons[2:21]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_subsets_models_codons$ID, cellwidth = 8,main = "codons",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

```



```{r }
RNA_subsets_models_aminos <- RNA_subsets_models
RNA_subsets_models_aminos$region[grep("amino", RNA_subsets_models_aminos$ID)] <- "amino"

RNA_subsets_models_aminos <- subset(RNA_subsets_models_aminos, RNA_subsets_models_aminos$region=="amino")

dim(RNA_subsets_models_aminos)


col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(RNA_subsets_models_aminos[2:21], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_aminos[2:21]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_subsets_models_aminos$ID, cellwidth = 8,main = "amino acids",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)



```



```{r AGAA}


RNA_subsets_models_AGAA <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")

RNA_subsets_models_AGAA <- RNA_subsets_models_AGAA[grep("AGAA",RNA_subsets_models_AGAA$ID),]

RNA_subsets_models_AGAA <- subset(RNA_subsets_models_AGAA, rowMeans(RNA_subsets_models_AGAA[,2:21])>6)
dim(RNA_subsets_models_AGAA)

#154360

col_fun_blue = colorRamp2(c(6,8,10), c("#FFFFFF","#2980b9","#000000")) 

pheatmap(RNA_subsets_models_AGAA[2:21], col_fun_blue(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_AGAA[2:21]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =RNA_subsets_models_AGAA$ID,  cellheight = 8, cellwidth = 8,main = "AGAA",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)



```




```{r CTTT}

RNA_subsets_models_CTTT <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")
RNA_subsets_models_CTTT <- RNA_subsets_models_CTTT[grep("CTTT",RNA_subsets_models_CTTT$ID),]

RNA_subsets_models_CTTT <- subset(RNA_subsets_models_CTTT, rowMeans(RNA_subsets_models_CTTT[,2:21])>6.5)
dim(RNA_subsets_models_CTTT)


col_fun_red = colorRamp2(c(6,8,10), c("#FFFFFF","#c0392b","#000000")) 

pheatmap(RNA_subsets_models_CTTT[2:21], col_fun_red(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_CTTT[2:21]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =RNA_subsets_models_CTTT$ID, cellheight = 8, cellwidth = 8,main = "CTTT",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

```




```{r ATTTA}


RNA_subsets_models_ATTTA <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")
RNA_subsets_models_ATTTA <- RNA_subsets_models_ATTTA[grep("ATTTA",RNA_subsets_models_ATTTA$ID),]

RNA_subsets_models_ATTTA <- subset(RNA_subsets_models_ATTTA, rowMeans(RNA_subsets_models_ATTTA[,2:21])>4)
dim(RNA_subsets_models_ATTTA)


col_fun_red = colorRamp2(c(6,8,10), c("#FFFFFF","#8e44ad","#000000")) 

pheatmap(RNA_subsets_models_ATTTA[2:21], col_fun_red(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_ATTTA[2:21]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =RNA_subsets_models_ATTTA$ID, cellheight = 8, cellwidth = 8,main = "ATTTA",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


```




```{r CTCAGG}


RNA_subsets_models_CTCAGG <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")
RNA_subsets_models_CTCAGG <- RNA_subsets_models_CTCAGG[grep("CTCAGG",RNA_subsets_models_CTCAGG$ID),]

RNA_subsets_models_CTCAGG <- subset(RNA_subsets_models_CTCAGG, rowMeans(RNA_subsets_models_CTCAGG[,2:21])>0)
dim(RNA_subsets_models_CTCAGG)


col_fun_orange = colorRamp2(c(6,8,10),  c("#FFFFFF","#f39c12","#000000")) 

pheatmap(RNA_subsets_models_CTCAGG[2:21], col_fun_orange(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_CTCAGG[2:21]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =RNA_subsets_models_CTCAGG$ID, cellheight = 8,cellwidth = 8,main = "CTCAGG",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


```





















## exporting RNA versus features ##

```{r importing RNA data}

RNA_CD8_Tcm_test <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/RNA_CD8_Tcm_test", sep=";")
RNA_CD8_Tcm_train <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/RNA_CD8_Tcm_train", sep=";")

RNA_CD8_Tcm <- rbind.data.frame(RNA_CD8_Tcm_test,RNA_CD8_Tcm_train)


```



```{r RNA modif prot}

ggplot(RNA_CD8_Tcm, aes(x=log10(total_m1A),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(total_m5C),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(total_m6A),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(total_m7G),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(total_AtoI),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```


```{r PTM modif RNA models}

ggplot(RNA_CD8_Tcm, aes(x=log10(Acetylation_count),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(Ubiquitination_count),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(Malonylation_count),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(Methylation_count),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=(drerio_homolog_perc_id_r1),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```




```{r m6a regions}


ggplot(RNA_CD8_Tcm, aes(x=log10(UTR5_m6a),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(CDS_m6a),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=log10(UTR3_m6a),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(total_m6A),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

```


```{r region m7G}


ggplot(RNA_CD8_Tcm, aes(x=UTR5_m7G,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=CDS_m7G,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=UTR3_m7G,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```


```{r 5UTR}

ggplot(RNA_CD8_Tcm, aes(x=log10(UTR5_length_UTR5),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



ggplot(RNA_CD8_Tcm, aes(x=GC_percentage_UTR5,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=GCGC_UTR5,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)





```



```{r CDS Prot}

ggplot(RNA_CD8_Tcm, aes(x=log10(CTTC_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(AGAA_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=(S_amino_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=(C_amino_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=(D_amino_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=(R_amino_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=GC_percentage_CDS,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r condon usage}

ggplot(RNA_CD8_Tcm, aes(x=(AAG_codon_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,30))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=(TGC_codon_CDS),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,10))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```



```{r 3UTR prot}

ggplot(RNA_CD8_Tcm, aes(x=GC_percentage_UTR3,y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(TTTT_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=log10(ATTTA_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=log10(AATAAA_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



```




```{r selected motifs}

ggplot(RNA_CD8_Tcm, aes(x=log10(TATTTA_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=log10(AGAAGA_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RNA_CD8_Tcm, aes(x=log10(CTTTCTT_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(RNA_CD8_Tcm, aes(x=log10(CTCAGGT_UTR3),y=RNA))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```



