---
title: "Analysis of flanking sequence of a motif"
author: "Benoit Nicolet"
date: "19/3/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(plyr)
library(dplyr)
library(seqinr)
library(stringr)
library(qdapRegex)
library(dplyr)
library(Biostrings)
library(biomaRt)
library(ggplot2)
library(data.table)
#BiocManager::install("Logolas")
library(Logolas)
library(ggseqlogo)


#BiocManager::install("DESeq2")
#library(motifmatchr)

knitr::opts_knit$set("/home/ben/Analysis/")
setwd("/home/ben/Analysis/")

```


```{r importing the formated Fasta file and processing}
## re-importing ##
#UTR3_fasta <- read.delim("/Users/stinky/Desktop/Sequences_info/3'UTR_ensembl_mart_export_nobreaks.txt",sep="\t",header = F)
UTR3_fasta <- read.delim("3'UTR_ensembl_mart_export_nobreaks.txt",sep="\t",header = F)


UTR3_fasta$gene.id <- mapply(strsplit(as.character(UTR3_fasta$V1),"\\|"),FUN=function(x){(as.character(x)[1])})
UTR3_fasta$gene.id <- gsub(">","",UTR3_fasta$gene.id)
UTR3_fasta$tx.id <- mapply(strsplit(as.character(UTR3_fasta$V1),"\\|"),FUN=function(x){(as.character(x)[2])})

UTR3_fasta$sequence <- UTR3_fasta$tx.id
UTR3_fasta$sequence <- gsub('^...............','',UTR3_fasta$sequence)

UTR3_fasta$tx.id <- gsub("A","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("T","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("G","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("C","",UTR3_fasta$tx.id)
UTR3_fasta$tx.id <- gsub("ENS","ENST",UTR3_fasta$tx.id)

UTR3_fasta$V1 <- NULL


UTR3_fasta$UTR3_length <- mapply(strsplit(as.character(UTR3_fasta$sequence),","),FUN=function(x){nchar(x)})

dim(UTR3_fasta)
#UTR3_fasta <- merge(UTR3_fasta, IDs_genenames_coding, by.x="tx.id",by.y="ensembl_transcript_id",all.x=T)
dim(UTR3_fasta)


UTR3_fasta <- UTR3_fasta[!UTR3_fasta$sequence=="Sequence unavailable",]

table(duplicated(UTR3_fasta$tx.id))

## oooh yeah : )

```


## ATTTA flanking motif
```{r ATTTA flanking motif}

sequence_input <- UTR3_fasta[1:1000,]
extracted_motif_final_list <- list()

for (line in 1:length(sequence_input$sequence)) {
  extracted_motif_list_per_seq <- list()
  tail_sequence <- sequence_input$sequence[line]
  
  for (l in 1:(stringr::str_count(tail_sequence,"ATTTA"))) {
  extracted_motif <- stringr::str_extract(tail_sequence,"..........ATTTA..........")
  extracted_motif_list_per_seq[[l+1]] <- extracted_motif
  extracted_motif_final_list[[line]] <- do.call(rbind,extracted_motif_list_per_seq)
  
  tail_sequence <- mapply(strsplit(as.character(tail_sequence),stringr::str_extract(tail_sequence,as.character(extracted_motif))),FUN=function(x){(as.character(x)[2])})
  #extracted_motif_final_list[[line]] <- extracted_motif_list_per_seq 
  }
  
  }
extracted_motif_final_list

motifs_db <- do.call(rbind,extracted_motif_final_list)
motifs_db



motifs_db <- data.frame(motifs_db)
motifs_db <- subset(motifs_db,!is.na(motifs_db))

motifs_db$motifs_db <- gsub("T","U",motifs_db$motifs_db)

motifs_db$left <- substring(motifs_db$motifs_db, 1, 10)
motifs_db$right <- substring(motifs_db$motifs_db, 16, 26)

logomaker(motifs_db$left, type = "EDLogo",
          colors = c("blue","red","green","yellow")) # order is C U A G for colors

logomaker(motifs_db$right, type = "EDLogo",
          colors = c("blue","red","green","yellow"))

```




## CTTT flanking motif
```{r CTTT flanking motif}

sequence_input <- UTR3_fasta[1:1000,]
extracted_motif_final_list <- list()

for (line in 1:length(sequence_input$sequence)) {
  extracted_motif_list_per_seq <- list()
  tail_sequence <- sequence_input$sequence[line]
  
  for (l in 1:(stringr::str_count(tail_sequence,"CTTT"))) {
  extracted_motif <- stringr::str_extract(tail_sequence,"..........CTTT..........")
  extracted_motif_list_per_seq[[l+1]] <- extracted_motif
  extracted_motif_final_list[[line]] <- do.call(rbind,extracted_motif_list_per_seq)
  
  tail_sequence <- mapply(strsplit(as.character(tail_sequence),stringr::str_extract(tail_sequence,as.character(extracted_motif))),FUN=function(x){(as.character(x)[2])})
  #extracted_motif_final_list[[line]] <- extracted_motif_list_per_seq 
  }
  
  }
extracted_motif_final_list

motifs_db <- do.call(rbind,extracted_motif_final_list)
motifs_db



motifs_db <- data.frame(motifs_db)
motifs_db <- subset(motifs_db,!is.na(motifs_db))
motifs_db$motifs_db <- gsub("T","U",motifs_db$motifs_db)
dim(motifs_db)

#motifs_db <- gsub("CTTT","NNNN",motifs_db)
motifs_db$left <- substring(motifs_db$motifs_db, 1, 10)
motifs_db$right <- substring(motifs_db$motifs_db, 16, 26)


logomaker(motifs_db$left, type = "EDLogo",
          colors = c("blue","red","green","yellow"))

logomaker(motifs_db$right, type = "EDLogo",
          colors = c("blue","red","green","yellow"))


```



## AGAAGA flanking motif
```{r AGAAGA flanking motif}

sequence_input <- UTR3_fasta
extracted_motif_final_list <- list()

for (line in 1:length(sequence_input$sequence)) {
  extracted_motif_list_per_seq <- list()
  tail_sequence <- sequence_input$sequence[line]
  
  for (l in 1:(stringr::str_count(tail_sequence,"AGAAGA"))) {
  extracted_motif <- stringr::str_extract(tail_sequence,"..........AGAAGA..........")
  extracted_motif_list_per_seq[[l+1]] <- extracted_motif
  extracted_motif_final_list[[line]] <- do.call(rbind,extracted_motif_list_per_seq)
  
  tail_sequence <- mapply(strsplit(as.character(tail_sequence),stringr::str_extract(tail_sequence,as.character(extracted_motif))),FUN=function(x){(as.character(x)[2])})
  #extracted_motif_final_list[[line]] <- extracted_motif_list_per_seq 
  }
  
  }
extracted_motif_final_list

motifs_db <- do.call(rbind,extracted_motif_final_list)
motifs_db



motifs_db <- data.frame(motifs_db)
motifs_db <- subset(motifs_db,!is.na(motifs_db))

dim(motifs_db)

#motifs_db <- gsub("CTTT","NNNN",motifs_db)
motifs_db$left <- substring(motifs_db$motifs_db, 1, 10)
motifs_db$right <- substring(motifs_db$motifs_db, 17, 27)


logomaker(motifs_db$left, type = "EDLogo",
          colors = c("blue","red","green","yellow"))

logomaker(motifs_db$right, type = "EDLogo",
          colors = c("blue","red","green","yellow"))


```

