---
title: "Visualizing RF output in T cell subsets"
author: "Benoit Nicolet"
date: "17/06/2022"
output: html_document
---


```{r setup, include=FALSE}

#install.packages("pheatmap")

library(plyr)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(circlize)
library(pheatmap)
library(caret)
library(ggpointdensity)
library(viridis)
library(reshape)
# library(GOexpress)


knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set("/home/ben/Analysis/RF_human/T_cell_subsets/")
setwd("/home/ben/Analysis/RF_human/T_cell_subsets/")

```


```{r importing RNA library}

Sequence_parameters <- read.delim("/home/ben/Analysis/RF_human/Library_V2_Aug2021/RNA_library_v2_RBP_GC_length_codon_AA_m6A_m5C_AtoI_m1A_m7G_CD8miRDB.csv",sep = ";", dec=",")

```


```{r importing RNA models}

CD8_Tn_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD8_Tn_14_10_2021.RDS")
CD8_Tcm_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD8_Tcm_14_10_2021.RDS")
CD8_Tem_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD8_Tem_14_10_2021.RDS")
CD8_Teff_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD8_Teff_10k_15_10_2021.RDS")

CD4_Tn_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD4_Tn_14_10_2021.RDS")
CD4_Tcm_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD4_Tcm_14_10_2021.RDS")
CD4_Tem_RNA <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/xgb_10knround_TPM_CD4_Tem_14_10_2021.RDS")


CD8_Tn_RNA<- data.frame(varImp(CD8_Tn_RNA)$importance)
CD8_Tcm_RNA<- data.frame(varImp(CD8_Tcm_RNA)$importance)
CD8_Tem_RNA<- data.frame(varImp(CD8_Tem_RNA)$importance)
CD8_Teff_RNA<- data.frame(varImp(CD8_Teff_RNA)$importance)
CD4_Tn_RNA<- data.frame(varImp(CD4_Tn_RNA)$importance)
CD4_Tcm_RNA<- data.frame(varImp(CD4_Tcm_RNA)$importance)
CD4_Tem_RNA<- data.frame(varImp(CD4_Tem_RNA)$importance)

CD8_Tn_RNA$ID <- rownames(CD8_Tn_RNA)
CD8_Tcm_RNA$ID <- rownames(CD8_Tcm_RNA)
CD8_Tem_RNA$ID <- rownames(CD8_Tem_RNA)
CD8_Teff_RNA$ID <- rownames(CD8_Teff_RNA)
CD4_Tn_RNA$ID <- rownames(CD4_Tn_RNA)
CD4_Tcm_RNA$ID <- rownames(CD4_Tcm_RNA)
CD4_Tem_RNA$ID <- rownames(CD4_Tem_RNA)


colnames(CD8_Tn_RNA)[1] <- "CD8_Tn_RNA"
colnames(CD8_Tcm_RNA)[1] <- "CD8_Tcm_RNA"
colnames(CD8_Tem_RNA)[1] <- "CD8_Tem_RNA"
colnames(CD8_Teff_RNA)[1] <- "CD8_Teff_RNA"
colnames(CD4_Tn_RNA)[1] <- "CD4_Tn_RNA"
colnames(CD4_Tcm_RNA)[1] <- "CD4_Tcm_RNA"
colnames(CD4_Tem_RNA)[1] <- "CD4_Tem_RNA"


RNA_subsets_models <- list(CD8_Tn_RNA,CD8_Tcm_RNA,CD8_Tem_RNA,CD8_Teff_RNA,CD4_Tn_RNA,CD4_Tcm_RNA,CD4_Tem_RNA)

RNA_subsets_models <- join_all(RNA_subsets_models,by = "ID")
RNA_subsets_models[1:2] <- RNA_subsets_models[2:1]
colnames(RNA_subsets_models)[1:2] <- colnames(RNA_subsets_models)[2:1]
dim(RNA_subsets_models)

# Min correction #
RNA_subsets_models[2:8] <- log10(RNA_subsets_models[2:8])
RNA_subsets_models <- do.call(data.frame,lapply(RNA_subsets_models,function(x) replace(x, is.infinite(x),NA)))
RNA_subsets_models[is.na(RNA_subsets_models)]=min(RNA_subsets_models[2:8],na.rm = T)
RNA_subsets_models[2:8] <- RNA_subsets_models[2:8] - min(RNA_subsets_models[2:8])

dim(RNA_subsets_models)

RNA_subsets_models$region <- "Other"
RNA_subsets_models$region[grep("CDS", RNA_subsets_models$ID)] <- "CDS"
RNA_subsets_models$region[grep("UTR5", RNA_subsets_models$ID)] <- "UTR5"
RNA_subsets_models$region[grep("UTR3", RNA_subsets_models$ID)] <- "UTR3"
#RNA_subsets_models$region[grep("codon", RNA_subsets_models$ID)] <- "codon"

dim(RNA_subsets_models)







RNA_subsets_models_CDS <- subset(RNA_subsets_models, RNA_subsets_models$region=="CDS")
dim(RNA_subsets_models_CDS)

RNA_subsets_models_UTR5 <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR5")
dim(RNA_subsets_models_UTR5)

RNA_subsets_models_UTR3 <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")
dim(RNA_subsets_models_UTR3)

RNA_subsets_models_Other <- subset(RNA_subsets_models, RNA_subsets_models$region=="Other")
dim(RNA_subsets_models_Other)

```



```{r corplot}
RNA_subsets_models_no_NA <- RNA_subsets_models
RNA_subsets_models_no_NA[is.na(RNA_subsets_models_no_NA)]=0

col_fun= colorRamp2(c(-3, -1.5, 0, 1.5, 3), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))
corrplot::corrplot(corr = (cor(RNA_subsets_models[,2:8],method = "pearson",use = "pairwise.complete.obs")), method = "shade",
                   diag = T,type = "full",order = "original", col.lim =c(0,1), col =col_fun(seq(-8.7,3,by=0.01)))

corrplot::corrplot(corr = (cor(RNA_subsets_models_UTR5[,2:8],method = "pearson",use = "pairwise.complete.obs")), method = "shade",
                   diag = T,type = "full",order = "original", col.lim =c(0,1), col =col_fun(seq(-8.7,3,by=0.01)))

corrplot::corrplot(corr = (cor(RNA_subsets_models_CDS[,2:8],method = "pearson",use = "pairwise.complete.obs")), method = "shade",
                   diag = T,type = "full",order = "original", col.lim =c(0,1), col =col_fun(seq(-8.7,3,by=0.01)))

corrplot::corrplot(corr = (cor(RNA_subsets_models_UTR3[,2:8],method = "pearson",use = "pairwise.complete.obs")), method = "shade",
                   diag = T,type = "full",order = "original", col.lim =c(0,1), col =col_fun(seq(-8.7,3,by=0.01)))


cor(RNA_subsets_models[,2:8],method = "pearson",use = "pairwise.complete.obs")

```



```{r corplot and heatmap on top features}




RNA_subsets_models_CDS_top <- RNA_subsets_models_CDS
RNA_subsets_models_CDS_top[2:8] <- 10^(RNA_subsets_models_CDS_top[2:8])
RNA_subsets_models_CDS_top$rowmean <- rowMeans(RNA_subsets_models_CDS_top[,2:8])
RNA_subsets_models_CDS_top <- RNA_subsets_models_CDS_top[order(RNA_subsets_models_CDS_top$rowmean,decreasing = T),]
RNA_subsets_models_CDS_top <- RNA_subsets_models_CDS_top[1:100,]
RNA_subsets_models_CDS_top[2:8] <- log10(RNA_subsets_models_CDS_top[2:8])
RNA_subsets_models_CDS_top <- do.call(data.frame,lapply(RNA_subsets_models_CDS_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_CDS_top)


RNA_subsets_models_UTR5_top <- RNA_subsets_models_UTR5
RNA_subsets_models_UTR5_top[2:8] <- 10^(RNA_subsets_models_UTR5_top[2:8])
RNA_subsets_models_UTR5_top$rowmean <- rowMeans(RNA_subsets_models_UTR5_top[,2:8])
RNA_subsets_models_UTR5_top <- RNA_subsets_models_UTR5_top[order(RNA_subsets_models_UTR5_top$rowmean,decreasing = T),]
RNA_subsets_models_UTR5_top <- RNA_subsets_models_UTR5_top[1:100,]
RNA_subsets_models_UTR5_top[2:8] <- log10(RNA_subsets_models_UTR5_top[2:8])
RNA_subsets_models_UTR5_top <- do.call(data.frame,lapply(RNA_subsets_models_UTR5_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_UTR5_top)


RNA_subsets_models_UTR3_top <- RNA_subsets_models_UTR3
RNA_subsets_models_UTR3_top[2:8] <- 10^(RNA_subsets_models_UTR3_top[2:8])
RNA_subsets_models_UTR3_top$rowmean <- rowMeans(RNA_subsets_models_UTR3_top[,2:8])
RNA_subsets_models_UTR3_top <- RNA_subsets_models_UTR3_top[order(RNA_subsets_models_UTR3_top$rowmean,decreasing = T),]
RNA_subsets_models_UTR3_top <- RNA_subsets_models_UTR3_top[1:100,]
RNA_subsets_models_UTR3_top[2:8] <- log10(RNA_subsets_models_UTR3_top[2:8])
RNA_subsets_models_UTR3_top <- do.call(data.frame,lapply(RNA_subsets_models_UTR3_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_UTR3_top)


RNA_subsets_models_Other_top <- RNA_subsets_models_Other
RNA_subsets_models_Other_top[2:8] <- 10^(RNA_subsets_models_Other_top[2:8])
RNA_subsets_models_Other_top$rowmean <- rowMeans(RNA_subsets_models_Other_top[,2:8])
RNA_subsets_models_Other_top <- RNA_subsets_models_Other_top[order(RNA_subsets_models_Other_top$rowmean,decreasing = T),]
RNA_subsets_models_Other_top <- RNA_subsets_models_Other_top[1:25,]
RNA_subsets_models_Other_top[2:8] <- log10(RNA_subsets_models_Other_top[2:8])
RNA_subsets_models_Other_top <- do.call(data.frame,lapply(RNA_subsets_models_Other_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_subsets_models_Other_top)


## Heatmap time ##
col_fun_HM = colorRamp2(c(6.5,7.5,9.5), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(RNA_subsets_models_UTR5_top[2:8], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_UTR5_top[2:8]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_subsets_models_UTR5_top$ID, cellwidth = 8, main = "5'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 4)

pheatmap(RNA_subsets_models_CDS_top[2:8], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_CDS_top[2:8]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_subsets_models_CDS_top$ID, cellwidth = 8, main = "CDS",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 4)

pheatmap(RNA_subsets_models_UTR3_top[2:8], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_UTR3_top[2:8]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_subsets_models_UTR3_top$ID, cellwidth = 8, main = "3'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 4)

pheatmap(RNA_subsets_models_Other_top[2:8], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_Other_top[2:8]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_subsets_models_Other$ID, cellwidth = 8, cellheight = 4, main = "general",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 4)


```


```{r importance per region}

RNA_subsets_models_for_imp_plot <- RNA_subsets_models
# RNA_subsets_models_for_imp_plot[RNA_subsets_models_for_imp_plot==min(RNA_subsets_models_for_imp_plot[,2:8])]=NA
# RNA_subsets_models_for_imp_plot[2:8] <- RNA_subsets_models_for_imp_plot[2:8] - min(RNA_subsets_models_for_imp_plot[2:8])
RNA_subsets_models_for_imp_plot <- melt(RNA_subsets_models_for_imp_plot)

# 
# ggplot(RNA_subsets_models_for_imp_plot,aes(x=region,y=value,group=variable))+
#   geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
#   #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
#   stat_summary(fun = median, fun.min = median, fun.max = median,
#                geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
#   stat_summary(fun = mean, fun.min = mean, fun.max = mean,
#                geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
#   theme_minimal()+
#   scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
#   theme(aspect.ratio = 1)


## average feature imp ##
RNA_subsets_models_avg_imp <- data.frame("ID"=RNA_subsets_models$ID,"rowmean"=rowMeans(RNA_subsets_models[,2:8]),"region"=RNA_subsets_models$region)
RNA_subsets_models_avg_imp <- melt(RNA_subsets_models_avg_imp)


ggplot(RNA_subsets_models_avg_imp,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = region), position = position_jitterdodge(jitter.width = 0.7, dodge.width = 0.8), size = 0.75, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.7, color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.7, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 3)


```

```{r capturing the variance of regions}

RNA_subsets_models_var <- data.frame(RNA_subsets_models)
RNA_subsets_models_var[2:8] <- 10^(RNA_subsets_models_var[2:8])

RNA_subsets_models_var[is.na(RNA_subsets_models_var)]=0
RNA_subsets_models_var$row_mean <- rowMeans(as.matrix(RNA_subsets_models_var[2:8]))

RNA_subsets_models_var$row_sd <- matrixStats::rowSds(as.matrix(RNA_subsets_models_var[2:8]))
RNA_subsets_models_var$corrected_sd <- RNA_subsets_models_var$row_sd/RNA_subsets_models_var$row_mean
RNA_subsets_models_var$row_mean <- log10(RNA_subsets_models_var$row_mean)
RNA_subsets_models_var[2:8] <- log10(RNA_subsets_models_var[2:8])

# plot(RNA_subsets_models_var$row_sd,RNA_subsets_models_var$row_mean)
# plot(RNA_subsets_models_var$corrected_sd,RNA_subsets_models_var$row_mean)

# ggplot(RNA_subsets_models_var,aes(x=corrected_sd,y=row_mean))+
#   geom_point(alpha=0.1)+
#   theme_minimal()+
#   theme(aspect.ratio = 1)


```

```{r full variablity assessment}

RNA_subsets_models_var$row_min <- do.call(pmin, as.data.frame(RNA_subsets_models_var[2:8]))
table(RNA_subsets_models_var$row_min==0)

dim(RNA_subsets_models_var[(RNA_subsets_models_var$corrected_sd < median(RNA_subsets_models_var$corrected_sd,na.rm = T)) & (RNA_subsets_models_var$row_min>0),])

dim(RNA_subsets_models_var[(RNA_subsets_models_var$row_min>0),])

core_features_RNA <- data.frame(
  table(RNA_subsets_models_var[(RNA_subsets_models_var$corrected_sd < median(RNA_subsets_models_var$corrected_sd,na.rm = T)) & (RNA_subsets_models_var$row_min>0),]$region)/table(RNA_subsets_models_var$region)*100)
core_features_RNA$group <- "C_core"

HV_core_features_RNA <- data.frame(
  table(RNA_subsets_models_var[(RNA_subsets_models_var$corrected_sd < median(RNA_subsets_models_var$corrected_sd,na.rm = T)) & (RNA_subsets_models_var$row_min==0),]$region)/table(RNA_subsets_models_var$region)*100)
HV_core_features_RNA$group <- "A_hyper_variable"

core_features_RNA <- rbind(core_features_RNA, HV_core_features_RNA, data.frame("Var1"=c("CDS","Other","UTR3","UTR5"),"Freq"= 100-(core_features_RNA$Freq + HV_core_features_RNA$Freq),"group"="B_variable"))


ggplot(core_features_RNA,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  scale_y_continuous(expand = c(0,0))+
  theme_minimal()+
  theme(aspect.ratio = 2)
  

# ggplot(RNA_subsets_models_var,aes(x=corrected_sd,y=row_mean))+
#   geom_point(alpha=0.1)+
#   theme(aspect.ratio = 1)


ggplot(RNA_subsets_models_var[RNA_subsets_models_var$row_min>0,],aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme(aspect.ratio = 1)


```

```{r capturing the variance of regions}
non_zero_RNA_subsets_models <- RNA_subsets_models_var

non_zero_RNA_subsets_models <- subset(non_zero_RNA_subsets_models,non_zero_RNA_subsets_models$row_min>0)
dim(non_zero_RNA_subsets_models)

ggplot(non_zero_RNA_subsets_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(non_zero_RNA_subsets_models,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

non_zero_RNA_subsets_models$region <- factor(non_zero_RNA_subsets_models$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(non_zero_RNA_subsets_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(non_zero_RNA_subsets_models$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)






```






```{r k-means clustering CD8 only}
# RNA_subsets_models[is.na(RNA_subsets_models)]=0
dim(RNA_subsets_models)

RNA_subsets_models_tops <- RNA_subsets_models
RNA_subsets_models_tops[6:9] <- NULL

RNA_subsets_models_tops[2:5] <- 10^(RNA_subsets_models_tops[2:5])
RNA_subsets_models_tops[is.na(RNA_subsets_models_tops)]=0
RNA_subsets_models_tops$row_mean <- rowMeans(as.matrix(RNA_subsets_models_tops[2:5]))

RNA_subsets_models_tops$row_sd <- matrixStats::rowSds(as.matrix(RNA_subsets_models_tops[2:5]))
RNA_subsets_models_tops$corrected_sd <- RNA_subsets_models_tops$row_sd/RNA_subsets_models_tops$row_mean
RNA_subsets_models_tops$row_mean <- log10(RNA_subsets_models_tops$row_mean)
RNA_subsets_models_tops$corrected_sd <- log10(RNA_subsets_models_tops$corrected_sd)
RNA_subsets_models_tops[2:5] <- log10(RNA_subsets_models_tops[2:5])
RNA_subsets_models_tops$row_min <- do.call(pmin, as.data.frame(RNA_subsets_models_tops[2:5]))


RNA_subsets_models_tops <- subset(RNA_subsets_models_tops,RNA_subsets_models_tops$corrected_sd>median(RNA_subsets_models_tops$corrected_sd))
RNA_subsets_models_tops <- subset(RNA_subsets_models_tops,RNA_subsets_models_tops$row_min > 0)

dim(RNA_subsets_models_tops)

RNA_subsets_models_scaled <- data.frame(RNA_subsets_models_tops[1:1],t(scale(t(RNA_subsets_models_tops[2:5]))))

set.seed(1234)
RNA_subsets_models_kmeans <- pheatmap(RNA_subsets_models_scaled[2:5], col_fun(seq(-5,5,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_subsets_models_scaled[2:5]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_subsets_models_scaled$ID,cellwidth = 20, cellheight = 20,kmeans_k = 8,clustering_method = "centroid")


#RNA_subsets_models_kmeans$kmeans$cluster

RNA_subsets_models_scaled$cluster <- RNA_subsets_models_kmeans$kmeans$cluster

RNA_subsets_models_scaled <- RNA_subsets_models_scaled[order(RNA_subsets_models_scaled$cluster,decreasing = F),]

pheatmap(RNA_subsets_models_scaled[2:5], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_subsets_models_scaled[2:5]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_subsets_models_scaled$ID,cellwidth = 20)

RNA_subsets_models_scaled <- data.frame(RNA_subsets_models_scaled)
RNA_subsets_models_scaled$rowvar <- matrixStats::rowVars(as.matrix(RNA_subsets_models_scaled[2:5]))
RNA_subsets_models_scaled <- subset(RNA_subsets_models_scaled,RNA_subsets_models_scaled$rowvar>0.1)
RNA_subsets_models_scaled$rowvar <- NULL
RNA_subsets_models_scaled <- data.table::data.table(RNA_subsets_models_scaled)

RNA_subsets_models_scaled_melted <- data.table::melt.data.table(RNA_subsets_models_scaled,id.vars = c("ID","cluster"))

RNA_subsets_models_scaled_melted$value <- as.numeric(RNA_subsets_models_scaled_melted$value)

RNA_subsets_models_scaled_melted <- RNA_subsets_models_scaled_melted[RNA_subsets_models_scaled_melted$variable!="region"]

write.table(RNA_subsets_models_scaled,"/home/ben/Analysis/RF_human/T_cell_subsets/kmers_CD8_RNA_models_with_features.csv", sep=";",dec = ",",row.names = F)

mean_data <- dplyr::group_by(RNA_subsets_models_scaled_melted,variable, cluster) %>%
             summarise(value = mean(value, na.rm = TRUE))

ggplot(RNA_subsets_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  geom_line(data=mean_data, aes(group = cluster), alpha=1, color="red")+
  facet_wrap(. ~ cluster, ncol = 4)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


```


```{r prep data for GO}
## Here we make a list of genes that are expressed in T cells to limit the GO search to genes found in cluster ##

# importing counts of T cell subsets #
CD4_Tcm_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD4_Tcm_TPM_log10.csv", sep = ";",dec = ",")
CD4_Tcm_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD4_Tcm_TPM_log10.csv", sep = ";",dec = ",")
CD4_Tcm_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD4_Tcm_TPM_log10.csv", sep = ";",dec = ",")

CD8_Tn_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD8_Tn_TPM_log10.csv", sep = ";",dec = ",")
CD8_Tcm_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD8_Tcm_TPM_log10.csv", sep = ";",dec = ",")
CD8_Tem_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD8_Tem_TPM_log10.csv", sep = ";",dec = ",")
CD8_Teff_TPM_log10 <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/RNA/CD8_Teff_TPM_log10.csv", sep = ";",dec = ",")


CD8_TPM_log10 <- rbind(CD8_Tn_TPM_log10,CD8_Tcm_TPM_log10,CD8_Tem_TPM_log10,CD8_Teff_TPM_log10)
CD8_TPM_log10 <- CD8_TPM_log10$ID %>% unique() %>% unique(fromLast=T)
CD8_TPM_log10 <- data.frame("tx.id"=CD8_TPM_log10)
length(CD8_TPM_log10)

tx2gene <- data.frame("tx.id"=Sequence_parameters$tx.id, "gene_name"= Sequence_parameters$gene_name)

CD8_genes_expressed <- merge(CD8_TPM_log10, tx2gene, by="tx.id")
CD8_genes_expressed <- CD8_genes_expressed$gene_name %>% unique() %>% unique(fromLast=T)
length(CD8_genes_expressed)

```


```{r getting Gos per cluster for CD8+}

RNA_CD8_subsets_models_scaled <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/kmers_CD8_RNA_models_with_features.csv", sep=";",dec = ",")

RNA_CD8_subsets_models_cluster1 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==1,]
RNA_CD8_subsets_models_cluster2 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==2,]
RNA_CD8_subsets_models_cluster3 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==3,]
RNA_CD8_subsets_models_cluster4 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==4,]
RNA_CD8_subsets_models_cluster5 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==5,]
RNA_CD8_subsets_models_cluster6 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==6,]
RNA_CD8_subsets_models_cluster7 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==7,]
RNA_CD8_subsets_models_cluster8 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==8,]
RNA_CD8_subsets_models_cluster9 <- RNA_CD8_subsets_models_scaled[RNA_CD8_subsets_models_scaled$cluster==9,]

dim(RNA_CD8_subsets_models_cluster1)
dim(RNA_CD8_subsets_models_cluster2)
dim(RNA_CD8_subsets_models_cluster3)
dim(RNA_CD8_subsets_models_cluster4)
dim(RNA_CD8_subsets_models_cluster5)
dim(RNA_CD8_subsets_models_cluster6)
dim(RNA_CD8_subsets_models_cluster7)
dim(RNA_CD8_subsets_models_cluster8)
dim(RNA_CD8_subsets_models_cluster9)







RNA_CD8_subsets_models_cluster1 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster1$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster2 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster2$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster3 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster3$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster4 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster4$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster5 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster5$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster6 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster6$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster7 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster7$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster8 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster8$ID],"gene_name"=Sequence_parameters$gene_name)
RNA_CD8_subsets_models_cluster9 <- cbind(Sequence_parameters[,RNA_CD8_subsets_models_cluster9$ID],"gene_name"=Sequence_parameters$gene_name)



dim(RNA_CD8_subsets_models_cluster1)
dim(RNA_CD8_subsets_models_cluster2)
dim(RNA_CD8_subsets_models_cluster3)
dim(RNA_CD8_subsets_models_cluster4)
dim(RNA_CD8_subsets_models_cluster5)
dim(RNA_CD8_subsets_models_cluster6)
dim(RNA_CD8_subsets_models_cluster7)
dim(RNA_CD8_subsets_models_cluster8)
dim(RNA_CD8_subsets_models_cluster9)



RNA_CD8_subsets_models_cluster1$rowSums <- rowSums(RNA_CD8_subsets_models_cluster1[1:dim(RNA_CD8_subsets_models_cluster1)[2]-1])
RNA_CD8_subsets_models_cluster2$rowSums <- rowSums(RNA_CD8_subsets_models_cluster2[1:dim(RNA_CD8_subsets_models_cluster2)[2]-1])
RNA_CD8_subsets_models_cluster3$rowSums <- rowSums(RNA_CD8_subsets_models_cluster3[1:dim(RNA_CD8_subsets_models_cluster3)[2]-1])
RNA_CD8_subsets_models_cluster4$rowSums <- rowSums(RNA_CD8_subsets_models_cluster4[1:dim(RNA_CD8_subsets_models_cluster4)[2]-1])
RNA_CD8_subsets_models_cluster5$rowSums <- rowSums(RNA_CD8_subsets_models_cluster5[1:dim(RNA_CD8_subsets_models_cluster5)[2]-1])
RNA_CD8_subsets_models_cluster6$rowSums <- rowSums(RNA_CD8_subsets_models_cluster6[1:dim(RNA_CD8_subsets_models_cluster6)[2]-1])
RNA_CD8_subsets_models_cluster7$rowSums <- rowSums(RNA_CD8_subsets_models_cluster7[1:dim(RNA_CD8_subsets_models_cluster7)[2]-1])
RNA_CD8_subsets_models_cluster8$rowSums <- rowSums(RNA_CD8_subsets_models_cluster8[1:dim(RNA_CD8_subsets_models_cluster8)[2]-1])
RNA_CD8_subsets_models_cluster9$rowSums <- rowSums(RNA_CD8_subsets_models_cluster9[1:dim(RNA_CD8_subsets_models_cluster9)[2]-1])




RNA_CD8_subsets_models_cluster1 <- subset(RNA_CD8_subsets_models_cluster1, RNA_CD8_subsets_models_cluster1$rowSums>100)
RNA_CD8_subsets_models_cluster2 <- subset(RNA_CD8_subsets_models_cluster2, RNA_CD8_subsets_models_cluster2$rowSums>1)
RNA_CD8_subsets_models_cluster3 <- subset(RNA_CD8_subsets_models_cluster3, RNA_CD8_subsets_models_cluster3$rowSums>1)
RNA_CD8_subsets_models_cluster4 <- subset(RNA_CD8_subsets_models_cluster4, RNA_CD8_subsets_models_cluster4$rowSums>1)
RNA_CD8_subsets_models_cluster5 <- subset(RNA_CD8_subsets_models_cluster5, RNA_CD8_subsets_models_cluster5$rowSums>1)
RNA_CD8_subsets_models_cluster6 <- subset(RNA_CD8_subsets_models_cluster6, RNA_CD8_subsets_models_cluster6$rowSums>1)
RNA_CD8_subsets_models_cluster7 <- subset(RNA_CD8_subsets_models_cluster7, RNA_CD8_subsets_models_cluster7$rowSums>1)
RNA_CD8_subsets_models_cluster8 <- subset(RNA_CD8_subsets_models_cluster8, RNA_CD8_subsets_models_cluster8$rowSums>1)
RNA_CD8_subsets_models_cluster9 <- subset(RNA_CD8_subsets_models_cluster9, RNA_CD8_subsets_models_cluster9$rowSums>1)


gene_name_cluster1 <- RNA_CD8_subsets_models_cluster1$gene_name
gene_name_cluster2 <- RNA_CD8_subsets_models_cluster2$gene_name
gene_name_cluster3 <- RNA_CD8_subsets_models_cluster3$gene_name
gene_name_cluster4 <- RNA_CD8_subsets_models_cluster4$gene_name
gene_name_cluster5 <- RNA_CD8_subsets_models_cluster5$gene_name
gene_name_cluster6 <- RNA_CD8_subsets_models_cluster6$gene_name
gene_name_cluster7 <- RNA_CD8_subsets_models_cluster7$gene_name
gene_name_cluster8 <- RNA_CD8_subsets_models_cluster8$gene_name
gene_name_cluster9 <- RNA_CD8_subsets_models_cluster9$gene_name



length(gene_name_cluster1)
length(gene_name_cluster2)
length(gene_name_cluster3)
length(gene_name_cluster4)
length(gene_name_cluster5)
length(gene_name_cluster6)
length(gene_name_cluster7)
length(gene_name_cluster8)
length(gene_name_cluster9)


gene_name_cluster1 <- gene_name_cluster1 %>% unique() %>% unique(fromLast = T)
gene_name_cluster2 <- gene_name_cluster2 %>% unique() %>% unique(fromLast = T)
gene_name_cluster3 <- gene_name_cluster3 %>% unique() %>% unique(fromLast = T)
gene_name_cluster4 <- gene_name_cluster4 %>% unique() %>% unique(fromLast = T)
gene_name_cluster5 <- gene_name_cluster5 %>% unique() %>% unique(fromLast = T)
gene_name_cluster6 <- gene_name_cluster6 %>% unique() %>% unique(fromLast = T)
gene_name_cluster7 <- gene_name_cluster7 %>% unique() %>% unique(fromLast = T)
gene_name_cluster8 <- gene_name_cluster8 %>% unique() %>% unique(fromLast = T)
gene_name_cluster9 <- gene_name_cluster9 %>% unique() %>% unique(fromLast = T)


gene_name_cluster1 <- gene_name_cluster1[gene_name_cluster1 %in% CD8_genes_expressed]
gene_name_cluster2 <- gene_name_cluster2[gene_name_cluster2 %in% CD8_genes_expressed]
gene_name_cluster3 <- gene_name_cluster3[gene_name_cluster3 %in% CD8_genes_expressed]
gene_name_cluster4 <- gene_name_cluster4[gene_name_cluster4 %in% CD8_genes_expressed]
gene_name_cluster5 <- gene_name_cluster5[gene_name_cluster5 %in% CD8_genes_expressed]
gene_name_cluster6 <- gene_name_cluster6[gene_name_cluster6 %in% CD8_genes_expressed]
gene_name_cluster7 <- gene_name_cluster7[gene_name_cluster7 %in% CD8_genes_expressed]
gene_name_cluster8 <- gene_name_cluster8[gene_name_cluster8 %in% CD8_genes_expressed]
gene_name_cluster9 <- gene_name_cluster9[gene_name_cluster9 %in% CD8_genes_expressed]


write.table(gene_name_cluster1,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster1.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster2,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster2.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster3,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster3.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster4,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster4.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster5,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster5.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster6,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster6.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster7,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster7.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster8,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster8.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster9,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/gene_name_cluster9.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)



table(gene_name_cluster1==gene_name_cluster7)


```






```{r k-means clustering CD4 only}
# RNA_subsets_models[is.na(RNA_subsets_models)]=0
dim(RNA_subsets_models)

RNA_subsets_models_tops_CD4 <- RNA_subsets_models
RNA_subsets_models_tops_CD4[2:5] <- NULL

RNA_subsets_models_tops_CD4[2:4] <- 10^(RNA_subsets_models_tops_CD4[2:4])
RNA_subsets_models_tops_CD4[is.na(RNA_subsets_models_tops_CD4)]=0
RNA_subsets_models_tops_CD4$row_mean <- rowMeans(as.matrix(RNA_subsets_models_tops_CD4[2:4]))

RNA_subsets_models_tops_CD4$row_sd <- matrixStats::rowSds(as.matrix(RNA_subsets_models_tops_CD4[2:4]))
RNA_subsets_models_tops_CD4$corrected_sd <- RNA_subsets_models_tops_CD4$row_sd/RNA_subsets_models_tops_CD4$row_mean
RNA_subsets_models_tops_CD4$row_mean <- log10(RNA_subsets_models_tops_CD4$row_mean)
RNA_subsets_models_tops_CD4$corrected_sd <- log10(RNA_subsets_models_tops_CD4$corrected_sd)
RNA_subsets_models_tops_CD4[2:4] <- log10(RNA_subsets_models_tops_CD4[2:4])
RNA_subsets_models_tops_CD4$row_min <- do.call(pmin, as.data.frame(RNA_subsets_models_tops_CD4[2:4]))


RNA_subsets_models_tops_CD4 <- subset(RNA_subsets_models_tops_CD4,RNA_subsets_models_tops_CD4$corrected_sd>median(RNA_subsets_models_tops_CD4$corrected_sd))
RNA_subsets_models_tops_CD4 <- subset(RNA_subsets_models_tops_CD4,RNA_subsets_models_tops_CD4$row_min > 0)

dim(RNA_subsets_models_tops_CD4)

RNA_subsets_models_scaled <- data.frame(RNA_subsets_models_tops_CD4[1:1],t(scale(t(RNA_subsets_models_tops_CD4[2:4]))))

set.seed(1234)
RNA_subsets_models_kmeans <- pheatmap(RNA_subsets_models_scaled[2:4], col_fun(seq(-5,5,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_subsets_models_scaled[2:4]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_subsets_models_scaled$ID,cellwidth = 20, cellheight = 20,kmeans_k = 4,clustering_method = "centroid")


#RNA_subsets_models_kmeans$kmeans$cluster

RNA_subsets_models_scaled$cluster <- RNA_subsets_models_kmeans$kmeans$cluster

RNA_subsets_models_scaled <- RNA_subsets_models_scaled[order(RNA_subsets_models_scaled$cluster,decreasing = F),]

pheatmap(RNA_subsets_models_scaled[2:4], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_subsets_models_scaled[2:4]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_subsets_models_scaled$ID,cellwidth = 20)

RNA_subsets_models_scaled <- data.frame(RNA_subsets_models_scaled)
RNA_subsets_models_scaled$rowvar <- matrixStats::rowVars(as.matrix(RNA_subsets_models_scaled[2:4]))
RNA_subsets_models_scaled <- subset(RNA_subsets_models_scaled,RNA_subsets_models_scaled$rowvar>0.1)
RNA_subsets_models_scaled$rowvar <- NULL
RNA_subsets_models_scaled <- data.table::data.table(RNA_subsets_models_scaled)

RNA_subsets_models_scaled_melted <- data.table::melt.data.table(RNA_subsets_models_scaled,id.vars = c("ID","cluster"))

RNA_subsets_models_scaled_melted$value <- as.numeric(RNA_subsets_models_scaled_melted$value)

RNA_subsets_models_scaled_melted <- RNA_subsets_models_scaled_melted[RNA_subsets_models_scaled_melted$variable!="region"]

write.table(RNA_subsets_models_scaled,"/home/ben/Analysis/RF_human/T_cell_subsets/kmers_CD4_RNA_models_with_features.csv", sep=";",dec = ",",row.names = F)

mean_data <- dplyr::group_by(RNA_subsets_models_scaled_melted,variable, cluster) %>%
             summarise(value = mean(value, na.rm = TRUE))

ggplot(RNA_subsets_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  geom_line(data=mean_data, aes(group = cluster), alpha=1, color="red")+
  facet_wrap(. ~ cluster, ncol = 4)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


```











```{r }
RNA_subsets_models_codons <- RNA_subsets_models
RNA_subsets_models_codons$region[grep("codon", RNA_subsets_models_codons$ID)] <- "codon"

RNA_subsets_models_codons <- subset(RNA_subsets_models_codons, RNA_subsets_models_codons$region=="codon")

dim(RNA_subsets_models_codons)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(RNA_subsets_models_codons[2:8], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_codons[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_subsets_models_codons$ID, gaps_col = 4,cellwidth = 20, cellheight = 6)


```





```{r AGAA}


RNA_subsets_models_AGAA <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")

RNA_subsets_models_AGAA <- RNA_subsets_models_AGAA[grep("AGAA",RNA_subsets_models_AGAA$ID),]

RNA_subsets_models_AGAA <- subset(RNA_subsets_models_AGAA,
                     RNA_subsets_models_AGAA$CD8_Tn_RNA>-2|
                     RNA_subsets_models_AGAA$CD8_Tcm_RNA>-2 |  
                     RNA_subsets_models_AGAA$CD8_Tem_RNA>-2 |
                     RNA_subsets_models_AGAA$CD8_Teff_RNA>-2 |
                     RNA_subsets_models_AGAA$CD4_Tn_RNA>-2 |
                     RNA_subsets_models_AGAA$CD4_Tcm_RNA>-2 |  
                     RNA_subsets_models_AGAA$CD4_Tem_RNA>-2)
dim(RNA_subsets_models_AGAA)

#154360

col_fun_blue = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#2980b9","#000000")) 

pheatmap(RNA_subsets_models_AGAA[2:8], col_fun_blue(seq(-3,4,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_AGAA[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RNA_subsets_models_AGAA$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```




```{r CTTT}


RNA_subsets_models_CTTT <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")

RNA_subsets_models_CTTT <- RNA_subsets_models_CTTT[grep("CTTTC",RNA_subsets_models_CTTT$ID),]

RNA_subsets_models_CTTT <- subset(RNA_subsets_models_CTTT,
                     RNA_subsets_models_CTTT$CD8_Tn_RNA>-2|
                     RNA_subsets_models_CTTT$CD8_Tcm_RNA>-2 |  
                     RNA_subsets_models_CTTT$CD8_Tem_RNA>-2 |
                     RNA_subsets_models_CTTT$CD8_Teff_RNA>-2 |
                     RNA_subsets_models_CTTT$CD4_Tn_RNA>-2 |
                     RNA_subsets_models_CTTT$CD4_Tcm_RNA>-2 |  
                     RNA_subsets_models_CTTT$CD4_Tem_RNA>-2)
dim(RNA_subsets_models_CTTT)

col_fun_red = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#c0392b","#000000")) 

pheatmap(RNA_subsets_models_CTTT[2:8], col_fun_red(seq(-5,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_CTTT[2:8]), fontsize_row=5,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RNA_subsets_models_CTTT$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)

```




```{r ATTTA}


RNA_subsets_models_ATTTA <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")

RNA_subsets_models_ATTTA <- RNA_subsets_models_ATTTA[grep("ATTTA",RNA_subsets_models_ATTTA$ID),]

RNA_subsets_models_ATTTA <- subset(RNA_subsets_models_ATTTA,
                     RNA_subsets_models_ATTTA$CD8_Tn_RNA>-2|
                     RNA_subsets_models_ATTTA$CD8_Tcm_RNA>-2 |  
                     RNA_subsets_models_ATTTA$CD8_Tem_RNA>-2 |
                     RNA_subsets_models_ATTTA$CD8_Teff_RNA>-2 |
                     RNA_subsets_models_ATTTA$CD4_Tn_RNA>-2 |
                     RNA_subsets_models_ATTTA$CD4_Tcm_RNA>-2 |  
                     RNA_subsets_models_ATTTA$CD4_Tem_RNA>-2)
dim(RNA_subsets_models_ATTTA)

col_fun_purple = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#8e44ad","#000000")) 

pheatmap(RNA_subsets_models_ATTTA[2:8], col_fun_purple(seq(-6,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_ATTTA[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RNA_subsets_models_ATTTA$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```




```{r CTCAGG}


RNA_subsets_models_CTCAGG <- subset(RNA_subsets_models, RNA_subsets_models$region=="UTR3")

RNA_subsets_models_CTCAGG <- RNA_subsets_models_CTCAGG[grep("CTCAGG",RNA_subsets_models_CTCAGG$ID),]

# RNA_subsets_models_CTCAGG <- subset(RNA_subsets_models_CTCAGG,
                     # RNA_subsets_models_CTCAGG$CD8_Tn_RNA>-2|
                     # RNA_subsets_models_CTCAGG$CD8_Tcm_RNA>-2 |  
                     # RNA_subsets_models_CTCAGG$CD8_Tem_RNA>-2 |
                     # RNA_subsets_models_CTCAGG$CD8_Teff_RNA>-2 |
                     # RNA_subsets_models_CTCAGG$CD4_Tn_RNA>-2 |
                     # RNA_subsets_models_CTCAGG$CD4_Tcm_RNA>-2 |  
                     # RNA_subsets_models_CTCAGG$CD4_Tem_RNA>-2)
dim(RNA_subsets_models_CTCAGG)

col_fun_orange = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#f39c12","#000000")) 

pheatmap(RNA_subsets_models_CTCAGG[2:8], col_fun_orange(seq(-6,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_subsets_models_CTCAGG[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RNA_subsets_models_CTCAGG$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```






```{r importing RNA feature for CD8 Tcm}

CD8_Tcm_train <- read.delim("~/Analysis/RF_human/T_cell_subsets/test-train-setsTPM/TPM_CD8_Tcm_train", sep=";")
CD8_Tcm_test <- read.delim("~/Analysis/RF_human/T_cell_subsets/test-train-setsTPM/TPM_CD8_Tcm_test", sep=";")

CD8_Tcm <- rbind.data.frame(CD8_Tcm_train,CD8_Tcm_test)
```



```{r RNA feature examples}
ggplot(CD8_Tcm, aes(x=(total_m7G),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F,position = "jitter")+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm', color= "red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,65))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$total_m7G, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$total_m7G and CD8_Tcm$TPM
# t = 36.074, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.2718579 0.3017181
# sample estimates:
#       cor 
# 0.2868577 
# R^2 : 0.08228734


ggplot(CD8_Tcm, aes(x=(total_m1A),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,70))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$total_m1A, CD8_Tcm$TPM,method = "pearson")

# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$total_m1A and CD8_Tcm$TPM
# t = 24.238, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.1815554 0.2128268
# sample estimates:
#       cor 
# 0.1972413 


ggplot(CD8_Tcm, aes(x=log10(total_m6A),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  theme(aspect.ratio = 1)

CD8_Tcm$total_m6A_log10 <- log10(CD8_Tcm$total_m6A)
CD8_Tcm <- do.call(data.frame,lapply(CD8_Tcm,function(x) replace(x, is.infinite(x),NA)))
cor.test(CD8_Tcm$total_m6A_log10, CD8_Tcm$TPM,method = "pearson")

# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$total_m6A_log10 and CD8_Tcm$TPM
# t = 9.8448, df = 13270, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.06823701 0.10201685
# sample estimates:
#       cor 
# 0.0851514 




ggplot(CD8_Tcm, aes(x=log10(total_m5C),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  theme(aspect.ratio = 1)

CD8_Tcm$total_m5C_log10 <- log10(CD8_Tcm$total_m5C)

CD8_Tcm <- do.call(data.frame,lapply(CD8_Tcm,function(x) replace(x, is.infinite(x),NA)))
cor.test(CD8_Tcm$total_m5C_log10, CD8_Tcm$TPM,method = "pearson",na.action = na.omit)

# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$total_m5C_log10 and CD8_Tcm$TPM
# t = -6.3016, df = 3693, p-value = 3.292e-10
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1349399 -0.0711346
# sample estimates:
#        cor 
# -0.1031433 


ggplot(CD8_Tcm, aes(x=log10(total_AtoI),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  theme(aspect.ratio = 1)



cor.test(CD8_Tcm$total_AtoI, CD8_Tcm$TPM,method = "pearson")


```


```{r m7G in regions}


ggplot(CD8_Tcm, aes(x=(CDS_m7G),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm', color='red')+
  geom_smooth(method='gam')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,70))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$CDS_m7G, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$CDS_m7G and CD8_Tcm$TPM
# t = 31.071, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.2344244 0.2649325
# sample estimates:
#       cor 
# 0.2497404 



ggplot(CD8_Tcm, aes(x=(UTR3_m7G),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm', color='red')+
  geom_smooth(method='gam')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,70))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$UTR3_m7G, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$UTR3_m7G and CD8_Tcm$TPM
# t = 17.532, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.1280452 0.1599075
# sample estimates:
#       cor 
# 0.1440137


```


```{r m6A in regions}

ggplot(CD8_Tcm, aes(x=(UTR5_m6a),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm', color='red')+
  geom_smooth(method='gam')+
  theme_minimal()+
  #scale_x_continuous(limits = c(0,70))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$UTR5_m6a, CD8_Tcm$TPM,method = "pearson")



ggplot(CD8_Tcm, aes(x=(CDS_m6a),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm', color='red')+
  geom_smooth(method='gam')+
  theme_minimal()+
  scale_x_continuous(limits = c(0,500))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$CDS_m6a, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$CDS_m7G and CD8_Tcm$TPM
# t = 31.071, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.2344244 0.2649325
# sample estimates:
#       cor 
# 0.2497404 



ggplot(CD8_Tcm, aes(x=(UTR3_m6a),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='lm', color='red')+
  geom_smooth(method='gam')+
  theme_minimal()+
  #scale_x_continuous(limits = c(0,70))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$UTR3_m6a, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$UTR3_m7G and CD8_Tcm$TPM
# t = 17.532, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  0.1280452 0.1599075
# sample estimates:
#       cor 
# 0.1440137


```



```{r 5UTR mRNA feature examples}


ggplot(CD8_Tcm, aes(x=(GC_percentage_UTR5),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(10,100))+
  theme(aspect.ratio = 1)



cor.test(CD8_Tcm$GC_percentage_UTR5, CD8_Tcm$TPM,method = "pearson")

# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$GC_percentage_UTR5 and CD8_Tcm$TPM
# t = -18.616, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1685640 -0.1367858
# sample estimates:
#        cor 
# -0.1527144 



ggplot(CD8_Tcm, aes(x=log10(UTR5_length_UTR5),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  #scale_x_continuous(limits = c(10,100))+
  theme(aspect.ratio = 1)



cor.test(CD8_Tcm$UTR5_length_UTR5, CD8_Tcm$TPM,method = "pearson")

# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$UTR5_length_UTR5 and CD8_Tcm$TPM
# t = -14.167, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1328054 -0.1007121
# sample estimates:
#        cor 
# -0.1167892 






ggplot(CD8_Tcm, aes(x=(AGGGA_UTR5),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  #scale_x_continuous(limits = c(10,100))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$AGGGA_UTR5, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$AGGGA_UTR5 and CD8_Tcm$TPM
# t = -11.868, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.11412660 -0.08190236
# sample estimates:
#         cor 
# -0.09804018 


ggplot(CD8_Tcm, aes(x=(CACA_UTR5),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$CACA_UTR5, CD8_Tcm$TPM,method = "pearson")

# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$CACA_UTR5 and CD8_Tcm$TPM
# t = -14.813, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1380318 -0.1059794
# sample estimates:
#        cor 
# -0.1220374 



ggplot(CD8_Tcm, aes(x=(CGCG_UTR5),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,50))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$CGCG_UTR5, CD8_Tcm$TPM,method = "pearson")

# Pearson's product-moment correlation
# 
# data:  CD8_Tcm$CGCG_UTR5 and CD8_Tcm$TPM
# t = -10.234, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.10077819 -0.06847438
# sample estimates:
#         cor 
# -0.08464853 


```



```{r CDS mRNA feature examples}


ggplot(CD8_Tcm, aes(x=(GC_percentage_CDS),y=TPM))+
  geom_pointdensity(stroke=0,size=2,show.legend = F)+
  scale_color_viridis(option = "G")+
  theme_minimal()+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  scale_x_continuous(limits = c(10,100))+
  theme(aspect.ratio = 1)



cor.test(CD8_Tcm$GC_percentage_CDS, CD8_Tcm$TPM,method = "pearson")

# Pearson's product-moment correlation
# 
# data:  CD8_Tcm$GC_percentage_CDS and CD8_Tcm$TPM
# t = -25.645, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.2237204 -0.1925936
# sample estimates:
#        cor 
# -0.2082097 

ggplot(CD8_Tcm, aes(x=log10(CDS_length_CDS),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)



cor.test(CD8_Tcm$CDS_length_CDS, CD8_Tcm$TPM,method = "pearson")
# Pearson's product-moment correlation
# 
# data:  CD8_Tcm$CDS_length_CDS and CD8_Tcm$TPM
# t = -18.201, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1652514 -0.1334404
# sample estimates:
#        cor 
# -0.1493845 



ggplot(CD8_Tcm, aes(x=log10(CTA_codon_CDS),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$CTA_codon_CDS, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$CTA_codon_CDS and CD8_Tcm$TPM
# t = 0.64879, df = 14513, p-value = 0.5165
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.01088401  0.02165194
# sample estimates:
#        cor 
# 0.00538539 



ggplot(CD8_Tcm, aes(x=log10(AGG_codon_CDS),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(-2,3))+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$AGG_codon_CDS, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$AGG_codon_CDS and CD8_Tcm$TPM
# t = -14.066, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.13198885 -0.09988943
# sample estimates:
#        cor 
# -0.1159694 





```


```{r 3UTR mRNA feature examples}


ggplot(CD8_Tcm, aes(x=(GC_percentage_UTR3),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  theme_minimal()+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  scale_x_continuous(limits = c(10,100))+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)



cor.test(CD8_Tcm$GC_percentage_UTR3, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$GC_percentage_UTR3 and CD8_Tcm$TPM
# t = -37.795, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.3140800 -0.2844579
# sample estimates:
#        cor 
# -0.2993411 



ggplot(CD8_Tcm, aes(x=(UTR3_length_UTR3),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

ggplot(CD8_Tcm, aes(x=log10(UTR3_length_UTR3),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)


cor.test(CD8_Tcm$UTR3_length_UTR3, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$UTR3_length_UTR3 and CD8_Tcm$TPM
# t = -25.14, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.2198201 -0.1886407
# sample estimates:
#        cor 
# -0.2042822 




ggplot(CD8_Tcm, aes(x=(AATAA_UTR3),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$AATAA_UTR3, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$AATAA_UTR3 and CD8_Tcm$TPM
# t = -20.073, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1801438 -0.1484856
# sample estimates:
#        cor 
# -0.1643571 


ggplot(CD8_Tcm, aes(x=(ATTTA_UTR3),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$ATTTA_UTR3, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$ATTTA_UTR3 and CD8_Tcm$TPM
# t = -17.672, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1610286 -0.1291769
# sample estimates:
#        cor 
# -0.1451403 




ggplot(CD8_Tcm, aes(x=(TTATTT_UTR3),y=TPM))+
  geom_pointdensity(stroke=0,size=1,show.legend = F)+
  scale_color_viridis(option = "G")+
  geom_smooth(method='gam')+
  geom_smooth(method='lm',color="red")+
  theme_minimal()+
  scale_y_continuous(limits = c(-1.5,4.5))+
  theme(aspect.ratio = 1)

cor.test(CD8_Tcm$TTATTT_UTR3, CD8_Tcm$TPM,method = "pearson")
# 	Pearson's product-moment correlation
# 
# data:  CD8_Tcm$TTATTT_UTR3 and CD8_Tcm$TPM
# t = -15.846, df = 14513, p-value < 2.2e-16
# alternative hypothesis: true correlation is not equal to 0
# 95 percent confidence interval:
#  -0.1463658 -0.1143821
# sample estimates:
#        cor 
# -0.1304079 




```



## CN Models ##

```{r importing CN models}

CD8_Tn_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD8_Tn_18_10_2021.RDS")
CD8_Tcm_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD8_Tcm_18_10_2021.RDS")
CD8_Tem_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD8_Tem_18_10_2021.RDS")
CD8_Teff_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD8_Teff_18_10_2021.RDS")

CD4_Tn_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD4_Tn_18_10_2021 (1).RDS")
CD4_Tcm_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD4_Tcm_18_10_2021.RDS")
CD4_Tem_CN <- readRDS("/home/ben/Analysis/RF_human/T_cell_subsets/models/xgb_CN_CD4_Tem_18_10_2021.RDS")


CD8_Tn_CN<- data.frame(varImp(CD8_Tn_CN)$importance)
CD8_Tcm_CN<- data.frame(varImp(CD8_Tcm_CN)$importance)
CD8_Tem_CN<- data.frame(varImp(CD8_Tem_CN)$importance)
CD8_Teff_CN<- data.frame(varImp(CD8_Teff_CN)$importance)
CD4_Tn_CN<- data.frame(varImp(CD4_Tn_CN)$importance)
CD4_Tcm_CN<- data.frame(varImp(CD4_Tcm_CN)$importance)
CD4_Tem_CN<- data.frame(varImp(CD4_Tem_CN)$importance)

CD8_Tn_CN$ID <- rownames(CD8_Tn_CN)
CD8_Tcm_CN$ID <- rownames(CD8_Tcm_CN)
CD8_Tem_CN$ID <- rownames(CD8_Tem_CN)
CD8_Teff_CN$ID <- rownames(CD8_Teff_CN)
CD4_Tn_CN$ID <- rownames(CD4_Tn_CN)
CD4_Tcm_CN$ID <- rownames(CD4_Tcm_CN)
CD4_Tem_CN$ID <- rownames(CD4_Tem_CN)


colnames(CD8_Tn_CN)[1] <- "CD8_Tn_CN"
colnames(CD8_Tcm_CN)[1] <- "CD8_Tcm_CN"
colnames(CD8_Tem_CN)[1] <- "CD8_Tem_CN"
colnames(CD8_Teff_CN)[1] <- "CD8_Teff_CN"
colnames(CD4_Tn_CN)[1] <- "CD4_Tn_CN"
colnames(CD4_Tcm_CN)[1] <- "CD4_Tcm_CN"
colnames(CD4_Tem_CN)[1] <- "CD4_Tem_CN"


CN_subsets_models <- list(CD8_Tn_CN,CD8_Tcm_CN,CD8_Tem_CN,CD8_Teff_CN,CD4_Tn_CN,CD4_Tcm_CN,CD4_Tem_CN)

CN_subsets_models <- join_all(CN_subsets_models,by = "ID")
CN_subsets_models[1:2] <- CN_subsets_models[2:1]
colnames(CN_subsets_models)[1:2] <- colnames(CN_subsets_models)[2:1]
dim(CN_subsets_models)

CN_subsets_models[2:8] <- log2(CN_subsets_models[2:8])

CN_subsets_models <- do.call(data.frame,lapply(CN_subsets_models,function(x) replace(x, is.infinite(x),NA)))


dim(CN_subsets_models)
CN_subsets_models$region <- "Other"
CN_subsets_models$region[grep("CDS", CN_subsets_models$ID)] <- "CDS"
CN_subsets_models$region[grep("UTR5", CN_subsets_models$ID)] <- "UTR5"
CN_subsets_models$region[grep("UTR3", CN_subsets_models$ID)] <- "UTR3"
#CN_subsets_models$region[grep("codon", CN_subsets_models$ID)] <- "codon"

dim(CN_subsets_models)







CN_subsets_models_CDS <- subset(CN_subsets_models, CN_subsets_models$region=="CDS")
CN_subsets_models_CDS <- subset(CN_subsets_models_CDS,
                     CN_subsets_models_CDS$CD8_Tn_CN>0.43|
                     CN_subsets_models_CDS$CD8_Tcm_CN>0.43 |  
                     CN_subsets_models_CDS$CD8_Tem_CN>0.43 |
                     CN_subsets_models_CDS$CD8_Teff_CN>0.43|
                     CN_subsets_models_CDS$CD4_Tn_CN>0.43 |
                     CN_subsets_models_CDS$CD4_Tcm_CN>0.43 |  
                     CN_subsets_models_CDS$CD4_Tem_CN>0.43 )
dim(CN_subsets_models_CDS)

CN_subsets_models_UTR5 <- subset(CN_subsets_models, CN_subsets_models$region=="UTR5")
CN_subsets_models_UTR5 <- subset(CN_subsets_models_UTR5,
                     CN_subsets_models_UTR5$CD8_Tn_CN>-1.2 |
                     CN_subsets_models_UTR5$CD8_Tcm_CN>-1.2 |  
                     CN_subsets_models_UTR5$CD8_Tem_CN>-1.2 |
                     CN_subsets_models_UTR5$CD8_Teff_CN>-1.2|
                     CN_subsets_models_UTR5$CD4_Tn_CN>-1.2 |
                     CN_subsets_models_UTR5$CD4_Tcm_CN>-1.2 |  
                     CN_subsets_models_UTR5$CD4_Tem_CN>-1.2)
dim(CN_subsets_models_UTR5)

CN_subsets_models_UTR3 <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")
CN_subsets_models_UTR3 <- subset(CN_subsets_models_UTR3,
                     CN_subsets_models_UTR3$CD8_Tn_CN>-1.3|
                     CN_subsets_models_UTR3$CD8_Tcm_CN>-1.3 |  
                     CN_subsets_models_UTR3$CD8_Tem_CN>-1.3 |
                     CN_subsets_models_UTR3$CD8_Teff_CN>-1.3 |
                     CN_subsets_models_UTR3$CD4_Tn_CN>-1.3 |
                     CN_subsets_models_UTR3$CD4_Tcm_CN>-1.3 |  
                     CN_subsets_models_UTR3$CD4_Tem_CN>-1.3)
dim(CN_subsets_models_UTR3)

CN_subsets_models_Other <- subset(CN_subsets_models, CN_subsets_models$region=="Other")
CN_subsets_models_Other <- subset(CN_subsets_models_Other,
                     CN_subsets_models_Other$CD8_Tn_CN>-1|
                     CN_subsets_models_Other$CD8_Tcm_CN>-1 |  
                     CN_subsets_models_Other$CD8_Tem_CN>-1 |
                     CN_subsets_models_Other$CD8_Teff_CN>-1 |
                     CN_subsets_models_Other$CD4_Tn_CN>-1|
                     CN_subsets_models_Other$CD4_Tcm_CN>-1 |  
                     CN_subsets_models_Other$CD4_Tem_CN>-1)
dim(CN_subsets_models_Other)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(CN_subsets_models_Other[2:8], col_fun(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_Other[2:8]), fontsize_row=5,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =CN_subsets_models_Other$ID, gaps_col = 4,cellwidth = 20)


pheatmap(CN_subsets_models_CDS[2:8], col_fun(seq(-5,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_CDS[2:8]), fontsize_row=3 ,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_subsets_models_CDS$ID, gaps_col = 4,cellwidth = 20)


pheatmap(CN_subsets_models_UTR5[2:8], col_fun(seq(-15,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_UTR5[2:8]), fontsize_row=3,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_subsets_models_UTR5$ID, gaps_col = 4,cellwidth = 20)


pheatmap(CN_subsets_models_UTR3[2:8], col_fun(seq(-15,2,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_UTR3[2:8]), fontsize_row=3,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_subsets_models_UTR3$ID, gaps_col = 4,cellwidth = 20)

```


```{r importance per region}

CN_subsets_models_for_imp_plot <- CN_subsets_models
CN_subsets_models_for_imp_plot <- melt(CN_subsets_models_for_imp_plot)


ggplot(CN_subsets_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
  scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```



```{r capturing the variance of regions}
CN_subsets_models_var <- data.frame(CN_subsets_models)
CN_subsets_models_var[2:8] <- 2^CN_subsets_models_var[2:8]
CN_subsets_models_var$row_var <- matrixStats::rowVars(as.matrix(CN_subsets_models_var[2:8]))
CN_subsets_models_var$row_mean <- rowMeans(as.matrix(CN_subsets_models_var[2:8]))

CN_subsets_models_var$row_sd <- matrixStats::rowSds(as.matrix(CN_subsets_models_var[2:8]))
CN_subsets_models_var$corrected_sd <- CN_subsets_models_var$row_sd/CN_subsets_models_var$row_mean
CN_subsets_models_var$row_mean <- log10(CN_subsets_models_var$row_mean)
CN_subsets_models_var$row_var <- log10(CN_subsets_models_var$row_var)

plot(CN_subsets_models_var$row_var,CN_subsets_models_var$row_mean)
plot(CN_subsets_models_var$row_sd,CN_subsets_models_var$row_mean)
plot(CN_subsets_models_var$corrected_sd,CN_subsets_models_var$row_mean)

ggplot(CN_subsets_models_var,aes(x=corrected_sd, fill=NULL, color=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(CN_subsets_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  theme(aspect.ratio = 0.5)

ggplot(CN_subsets_models_var,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)


ggplot(CN_subsets_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)



mean(CN_subsets_models_var[CN_subsets_models_var$region=="UTR5",]$corrected_sd,na.rm = T) # = -2.38509
mean(CN_subsets_models_var[CN_subsets_models_var$region=="CDS",]$corrected_sd,na.rm = T) # = -2.031128
mean(CN_subsets_models_var[CN_subsets_models_var$region=="UTR3",]$corrected_sd,na.rm = T) # = -2.061535
mean(CN_subsets_models_var[CN_subsets_models_var$region=="Other",]$corrected_sd,na.rm = T) # = -2.781053


table(CN_subsets_models_var[CN_subsets_models_var$corrected_sd>1.5,]$region)/table(CN_subsets_models_var$region)*100



```



```{r k-means clustering CD8 only}
# CN_subsets_models[is.na(CN_subsets_models)]=0
CN_subsets_models_tops <- CN_subsets_models
CN_subsets_models_tops[6:8] <- NULL
CN_subsets_models_tops$rowmean <- rowMeans(CN_subsets_models_tops[2:5])
#CN_subsets_models_tops$rowvar <- rowVars(as.matrix(CN_subsets_models_tops[2:8]))
CN_subsets_models_tops <- subset(CN_subsets_models_tops,CN_subsets_models_tops$rowmean>-10)

CN_subsets_models_scaled <- data.frame(CN_subsets_models_tops[1:1],scale(CN_subsets_models_tops[2:5]),CN_subsets_models_tops[6:6])

set.seed(1234)
CN_subsets_models_kmeans <- pheatmap(CN_subsets_models_scaled[2:5], col_fun(seq(-5,5,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(CN_subsets_models_scaled[2:5]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_subsets_models_scaled$ID,cellwidth = 20, cellheight = 20,kmeans_k = 7,clustering_method = "centroid")


CN_subsets_models_kmeans$kmeans$cluster

CN_subsets_models_scaled$cluster <- CN_subsets_models_kmeans$kmeans$cluster

CN_subsets_models_scaled <- CN_subsets_models_scaled[order(CN_subsets_models_scaled$cluster,decreasing = F),]

pheatmap(CN_subsets_models_scaled[2:5], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(CN_subsets_models_scaled[2:5]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_subsets_models_scaled$ID,cellwidth = 20)

CN_subsets_models_scaled <- data.frame(CN_subsets_models_scaled)
CN_subsets_models_scaled$rowvar <- matrixStats::rowVars(as.matrix(CN_subsets_models_scaled[2:5]))
CN_subsets_models_scaled <- subset(CN_subsets_models_scaled,CN_subsets_models_scaled$rowvar>0.1)
CN_subsets_models_scaled$rowvar <- NULL
CN_subsets_models_scaled <- data.table::data.table(CN_subsets_models_scaled)

CN_subsets_models_scaled_melted <- data.table::melt.data.table(CN_subsets_models_scaled,id.vars = c("ID","cluster"))

CN_subsets_models_scaled_melted$value <- as.numeric(CN_subsets_models_scaled_melted$value)

CN_subsets_models_scaled_melted <- CN_subsets_models_scaled_melted[CN_subsets_models_scaled_melted$variable!="region"]


ggplot(CN_subsets_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  facet_wrap(. ~ cluster, ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

#write.table(CN_subsets_models_scaled,"/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Protein_naive_t_models_integrated_with_kmers.csv", sep=";",dec = ",",row.names = F)





## frequencies in clusters ##

region_freq_in_clusters_Prot <- rbind(data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==1,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==1])[1]*100, "cluster"=1),
data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==2,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==2])[1]*100, "cluster"=2),
data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==3,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==3])[1]*100, "cluster"=3),
data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==4,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==4])[1]*100, "cluster"=4),
data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==5,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==5])[1]*100, "cluster"=5),
data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==6,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==6])[1]*100, "cluster"=6),
data.frame(table(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==7,]$region)/dim(CN_subsets_models_scaled[CN_subsets_models_scaled$cluster==7])[1]*100, "cluster"=7)

)

region_freq_in_clusters_Prot


ggplot(region_freq_in_clusters_Prot,aes(y=Freq, x=Var1, fill=Var1))+
  geom_bar(stat = 'identity')+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  facet_wrap(. ~ cluster, ncol = 2)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))





```






```{r k-means clustering CD4 only}
# CN_subsets_models[is.na(CN_subsets_models)]=0
CN_CD4_subsets_models_tops <- CN_subsets_models
CN_CD4_subsets_models_tops[2:5] <- NULL
CN_CD4_subsets_models_tops[2:4] <- 2^CN_CD4_subsets_models_tops[2:4]
CN_CD4_subsets_models_tops$rowmean <- rowMeans(CN_CD4_subsets_models_tops[2:4])
CN_CD4_subsets_models_tops$row_sd <- matrixStats::rowSds(as.matrix(CN_CD4_subsets_models_tops[2:4]))
CN_CD4_subsets_models_tops$corrected_sd <- CN_CD4_subsets_models_tops$row_sd/CN_CD4_subsets_models_tops$rowmean
CN_CD4_subsets_models_tops$rowmean <- log10(CN_CD4_subsets_models_tops$rowmean)
CN_CD4_subsets_models_tops <- subset(CN_CD4_subsets_models_tops,CN_CD4_subsets_models_tops$rowmean>-3)
plot(CN_CD4_subsets_models_tops$corrected_sd,CN_CD4_subsets_models_tops$rowmean)
CN_CD4_subsets_models_tops[2:4] <- log2(CN_CD4_subsets_models_tops[2:4])


CN_CD4_subsets_models_scaled <- data.frame(CN_CD4_subsets_models_tops[1:1],scale(CN_CD4_subsets_models_tops[2:4]),CN_CD4_subsets_models_tops[5:5])

set.seed(1234)
CN_subsets_models_kmeans <- pheatmap(CN_CD4_subsets_models_scaled[2:4], col_fun(seq(-5,5,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(CN_CD4_subsets_models_scaled[2:4]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_CD4_subsets_models_scaled$ID,cellwidth = 20, cellheight = 20,kmeans_k = 4,clustering_method = "centroid")


# CN_subsets_models_kmeans$kmeans$cluster

CN_CD4_subsets_models_scaled$cluster <- CN_subsets_models_kmeans$kmeans$cluster

CN_CD4_subsets_models_scaled <- CN_CD4_subsets_models_scaled[order(CN_CD4_subsets_models_scaled$cluster,decreasing = F),]

pheatmap(CN_CD4_subsets_models_scaled[2:4], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(CN_CD4_subsets_models_scaled[2:4]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_CD4_subsets_models_scaled$ID,cellwidth = 20)

CN_CD4_subsets_models_scaled <- data.frame(CN_CD4_subsets_models_scaled)
CN_CD4_subsets_models_scaled$rowvar <- matrixStats::rowVars(as.matrix(CN_CD4_subsets_models_scaled[2:4]))
CN_CD4_subsets_models_scaled <- subset(CN_CD4_subsets_models_scaled,CN_CD4_subsets_models_scaled$rowvar>0.1)
CN_CD4_subsets_models_scaled$rowvar <- NULL
CN_CD4_subsets_models_scaled <- data.table::data.table(CN_CD4_subsets_models_scaled)

CN_CD4_subsets_models_scaled_melted <- data.table::melt.data.table(CN_CD4_subsets_models_scaled,id.vars = c("ID","cluster"))

CN_CD4_subsets_models_scaled_melted$value <- as.numeric(CN_CD4_subsets_models_scaled_melted$value)

CN_CD4_subsets_models_scaled_melted <- CN_CD4_subsets_models_scaled_melted[CN_CD4_subsets_models_scaled_melted$variable!="region"]


ggplot(CN_CD4_subsets_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  facet_wrap(. ~ cluster, ncol = 1)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

#write.table(CN_CD4_subsets_models_scaled,"/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Protein_naive_t_models_integrated_with_kmers.csv", sep=";",dec = ",",row.names = F)





## frequencies in clusters ##

region_freq_in_clusters_Prot <- rbind(data.frame(table(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==1,]$region)/dim(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==1])[1]*100, "cluster"=1),
data.frame(table(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==2,]$region)/dim(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==2])[1]*100, "cluster"=2),
data.frame(table(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==3,]$region)/dim(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==3])[1]*100, "cluster"=3),
data.frame(table(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==4,]$region)/dim(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==4])[1]*100, "cluster"=4)
#data.frame(table(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==5,]$region)/dim(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==5])[1]*100, "cluster"=5),
#data.frame(table(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==6,]$region)/dim(CN_CD4_subsets_models_scaled[CN_CD4_subsets_models_scaled$cluster==6])[1]*100, "cluster"=6)
)

region_freq_in_clusters_Prot


ggplot(region_freq_in_clusters_Prot,aes(y=Freq, x=Var1, fill=Var1))+
  geom_bar(stat = 'identity')+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  facet_wrap(. ~ cluster, ncol = 1)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))





```








```{r }
CN_subsets_models_codons <- CN_subsets_models
CN_subsets_models_codons$region[grep("codon", CN_subsets_models_codons$ID)] <- "codon"

CN_subsets_models_codons <- subset(CN_subsets_models_codons, CN_subsets_models_codons$region=="codon")

dim(CN_subsets_models_codons)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(CN_subsets_models_codons[2:8], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_codons[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =CN_subsets_models_codons$ID, gaps_col = 4,cellwidth = 20, cellheight = 6)


```



```{r }
CN_subsets_models_aminos <- CN_subsets_models
CN_subsets_models_aminos$region[grep("amino", CN_subsets_models_aminos$ID)] <- "amino"

CN_subsets_models_aminos <- subset(CN_subsets_models_aminos, CN_subsets_models_aminos$region=="amino")

dim(CN_subsets_models_aminos)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(CN_subsets_models_aminos[2:8], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_aminos[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =CN_subsets_models_aminos$ID, gaps_col = 4,cellwidth = 20, cellheight = 6)


```



```{r AGAA}


CN_subsets_models_AGAA <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")

CN_subsets_models_AGAA <- CN_subsets_models_AGAA[grep("AGAA",CN_subsets_models_AGAA$ID),]

CN_subsets_models_AGAA <- subset(CN_subsets_models_AGAA,
                     CN_subsets_models_AGAA$CD8_Tn_CN>-2.5|
                     CN_subsets_models_AGAA$CD8_Tcm_CN>-2.5 |  
                     CN_subsets_models_AGAA$CD8_Tem_CN>-2.5 |
                     CN_subsets_models_AGAA$CD8_Teff_CN>-2.5 |
                     CN_subsets_models_AGAA$CD4_Tn_CN>-2.5 |
                     CN_subsets_models_AGAA$CD4_Tcm_CN>-2.5 |  
                     CN_subsets_models_AGAA$CD4_Tem_CN>-2.5)
dim(CN_subsets_models_AGAA)

#154360

col_fun_blue = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#2980b9","#000000")) 

pheatmap(CN_subsets_models_AGAA[2:8], col_fun_blue(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_AGAA[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =CN_subsets_models_AGAA$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```




```{r CTTT}


CN_subsets_models_CTTT <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")

CN_subsets_models_CTTT <- CN_subsets_models_CTTT[grep("CTTT",CN_subsets_models_CTTT$ID),]

CN_subsets_models_CTTT <- subset(CN_subsets_models_CTTT,
                     CN_subsets_models_CTTT$CD8_Tn_CN>-2.5|
                     CN_subsets_models_CTTT$CD8_Tcm_CN>-2.5 |  
                     CN_subsets_models_CTTT$CD8_Tem_CN>-2.5 |
                     CN_subsets_models_CTTT$CD8_Teff_CN>-2.5 |
                     CN_subsets_models_CTTT$CD4_Tn_CN>-2.5 |
                     CN_subsets_models_CTTT$CD4_Tcm_CN>-2.5 |  
                     CN_subsets_models_CTTT$CD4_Tem_CN>-2.5)
dim(CN_subsets_models_CTTT)

col_fun_red = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#c0392b","#000000")) 

pheatmap(CN_subsets_models_CTTT[2:8], col_fun_red(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_CTTT[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =CN_subsets_models_CTTT$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)

```




```{r ATTTA}


CN_subsets_models_ATTTA <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")

CN_subsets_models_ATTTA <- CN_subsets_models_ATTTA[grep("ATTTA",CN_subsets_models_ATTTA$ID),]

CN_subsets_models_ATTTA <- subset(CN_subsets_models_ATTTA,
                     CN_subsets_models_ATTTA$CD8_Tn_CN>-2|
                     CN_subsets_models_ATTTA$CD8_Tcm_CN>-2 |  
                     CN_subsets_models_ATTTA$CD8_Tem_CN>-2 |
                     CN_subsets_models_ATTTA$CD8_Teff_CN>-2 |
                     CN_subsets_models_ATTTA$CD4_Tn_CN>-2 |
                     CN_subsets_models_ATTTA$CD4_Tcm_CN>-2 |  
                     CN_subsets_models_ATTTA$CD4_Tem_CN>-2)
dim(CN_subsets_models_ATTTA)

col_fun_purple = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#8e44ad","#000000")) 

pheatmap(CN_subsets_models_ATTTA[2:8], col_fun_purple(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_ATTTA[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =CN_subsets_models_ATTTA$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```




```{r CTCAGG}


CN_subsets_models_CTCAGG <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")

CN_subsets_models_CTCAGG <- CN_subsets_models_CTCAGG[grep("CTCAGG",CN_subsets_models_CTCAGG$ID),]

# CN_subsets_models_CTCAGG <- subset(CN_subsets_models_CTCAGG,
                     # CN_subsets_models_CTCAGG$CD8_Tn_CN>-2|
                     # CN_subsets_models_CTCAGG$CD8_Tcm_CN>-2 |  
                     # CN_subsets_models_CTCAGG$CD8_Tem_CN>-2 |
                     # CN_subsets_models_CTCAGG$CD8_Teff_CN>-2 |
                     # CN_subsets_models_CTCAGG$CD4_Tn_CN>-2 |
                     # CN_subsets_models_CTCAGG$CD4_Tcm_CN>-2 |  
                     # CN_subsets_models_CTCAGG$CD4_Tem_CN>-2)
dim(CN_subsets_models_CTCAGG)

col_fun_orange = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#f39c12","#000000")) 

pheatmap(CN_subsets_models_CTCAGG[2:8], col_fun_orange(seq(-8,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_CTCAGG[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =CN_subsets_models_CTCAGG$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```


## exporting CN versus features ##

```{r importing CN data}

CN_CD8_Tcm_test <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/CN_CD8_Tcm_test", sep=";")
CN_CD8_Tcm_train <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/CN_CD8_Tcm_train", sep=";")

CN_CD8_Tcm <- rbind.data.frame(CN_CD8_Tcm_test,CN_CD8_Tcm_train)


```



```{r RNA modif prot}

ggplot(CN_CD8_Tcm, aes(x=log10(total_m1A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m5C),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m6A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m7G),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_AtoI),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```


```{r PTM modif CN models}

ggplot(CN_CD8_Tcm, aes(x=log10(Acetylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Ubiquitination_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Malonylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Methylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=(drerio_homolog_perc_id_r1),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```




```{r m6a regions}


ggplot(CN_CD8_Tcm, aes(x=log10(UTR5_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(CDS_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(UTR3_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m6A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

```


```{r region m7G}


ggplot(CN_CD8_Tcm, aes(x=UTR5_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=CDS_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=UTR3_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```


```{r 5UTR}

ggplot(CN_CD8_Tcm, aes(x=log10(UTR5_length_UTR5),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



ggplot(CN_CD8_Tcm, aes(x=GC_percentage_UTR5,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=GCGC_UTR5,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)





```



```{r CDS Prot}

ggplot(CN_CD8_Tcm, aes(x=log10(CTTC_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(AGAA_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(S_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(C_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(D_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=(R_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=GC_percentage_CDS,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r condon usage}

ggplot(CN_CD8_Tcm, aes(x=(AAG_codon_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,30))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(TGC_codon_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,10))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```



```{r 3UTR prot}

ggplot(CN_CD8_Tcm, aes(x=GC_percentage_UTR3,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(TTTT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(ATTTA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(AATAAA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



```




```{r selected motifs}

ggplot(CN_CD8_Tcm, aes(x=log10(TATTTA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(AGAAGA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(CTTTCTT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(CTCAGGT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```








## Translation efficiency estimates 



```{r importing TE models}

TE_rest <- readRDS("/home/ben/Analysis/RF_human/xgb_TE_rest_08-11-2021.RDS")
TE_PMA <- readRDS("/home/ben/Analysis/RF_human/xgb_TE_PMA_08-11-2021.RDS")


TE_rest<- data.frame(varImp(TE_rest)$importance)
TE_PMA<- data.frame(varImp(TE_PMA)$importance)

TE_rest$ID <- rownames(TE_rest)
TE_PMA$ID <- rownames(TE_PMA)


colnames(TE_rest)[1] <- "TE_rest"
colnames(TE_PMA)[1] <- "TE_PMA"

TE_models <- list(TE_rest,TE_PMA)

TE_models <- join_all(TE_models,by = "ID")
TE_models[1:2] <- TE_models[2:1]
colnames(TE_models)[1:2] <- colnames(TE_models)[2:1]
dim(TE_models)

TE_models[2:3] <- log2(TE_models[2:3])

TE_models <- do.call(data.frame,lapply(TE_models,function(x) replace(x, is.infinite(x),NA)))


dim(TE_models)
TE_models$region <- "Other"
TE_models$region[grep("CDS", TE_models$ID)] <- "CDS"
TE_models$region[grep("UTR5", TE_models$ID)] <- "UTR5"
TE_models$region[grep("UTR3", TE_models$ID)] <- "UTR3"
#TE_models$region[grep("codon", TE_models$ID)] <- "codon"

dim(TE_models)

TE_models_NAs <- TE_models

TE_models[is.na(TE_models)]=-5





TE_models_CDS <- subset(TE_models, TE_models$region=="CDS")
TE_models_CDS <- subset(TE_models_CDS,
                     TE_models_CDS$TE_rest>2.63|
                     TE_models_CDS$TE_PMA>2.63 )
dim(TE_models_CDS)

TE_models_UTR5 <- subset(TE_models, TE_models$region=="UTR5")
TE_models_UTR5 <- subset(TE_models_UTR5,
                     TE_models_UTR5$TE_rest>1.34 |
                     TE_models_UTR5$TE_PMA>1.34 )
dim(TE_models_UTR5)

TE_models_UTR3 <- subset(TE_models, TE_models$region=="UTR3")
TE_models_UTR3 <- subset(TE_models_UTR3,
                     TE_models_UTR3$TE_rest>1.01|
                     TE_models_UTR3$TE_PMA>1.01 )
dim(TE_models_UTR3)

TE_models_Other <- subset(TE_models, TE_models$region=="Other")
TE_models_Other <- subset(TE_models_Other,
                     TE_models_Other$TE_rest>1|
                     TE_models_Other$TE_PMA>1)
dim(TE_models_Other)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(TE_models_Other[2:3], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_Other[2:3]), fontsize_row=5,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =TE_models_Other$ID,cellwidth = 20)


pheatmap(TE_models_CDS[2:3], col_fun(seq(-5,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_CDS[2:3]), fontsize_row=4 ,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_CDS$ID,cellwidth = 20)


pheatmap(TE_models_UTR5[2:3], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_UTR5[2:3]), fontsize_row=4,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_UTR5$ID, cellwidth = 20)


pheatmap(TE_models_UTR3[2:3], col_fun(seq(-4,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_UTR3[2:3]), fontsize_row=4,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_UTR3$ID,cellwidth = 20)


```


```{r importance per region}

TE_models_for_imp_plot <- TE_models_NAs
TE_models_for_imp_plot <- melt(TE_models_for_imp_plot)


ggplot(TE_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.7), size = 0.75, stroke=0)+
  scale_color_manual(values = c("#3498db","#C70039"))+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.6, color="red",position = position_dodge(preserve = "total",width = 0.7))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.6, color="blue",position = position_dodge(preserve = "total",width = 0.7))+
  theme_minimal()+
  theme(aspect.ratio = 2)

TE_models_for_imp_plot$miRNA <- "Not miRNA"
TE_models_for_imp_plot$miRNA[grep("hsa", TE_models$ID)] <- "miRNA"


ggplot(TE_models_for_imp_plot,aes(x=miRNA,y=value,color=miRNA))+
  geom_point(position = "jitter")+
  # geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.7), size = 0.75, stroke=0)+
  scale_color_manual(values = c("#C70039","#3498db"))+
  scale_x_discrete(limits = c("Not miRNA","miRNA"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.6, color="red",position = position_dodge(preserve = "total",width = 0.7))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.6, color="blue",position = position_dodge(preserve = "total",width = 0.7))+
  theme_minimal()+
  theme(aspect.ratio = 2)

```




```{r CDS}
TE_models_CDS_dotplot <-TE_models_NAs[TE_models_NAs$region=="CDS",]

TE_models_CDS_dotplot$LFC_rest_PMA <- TE_models_CDS_dotplot$TE_PMA - TE_models_CDS_dotplot$TE_rest

TE_models_CDS_dotplot <- do.call(data.frame,lapply(TE_models_CDS_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_CDS_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[abs(TE_models_CDS_dotplot$LFC_rest_PMA)>1 & (TE_models_CDS_dotplot$TE_rest>2 | TE_models_CDS_dotplot$TE_PMA>2),], aes(label=ID),size=2.5)+
  theme_minimal()+
  theme(aspect.ratio = 1)




TE_models_CDS_dotplot$rank_TE_rest <- rank(-TE_models_CDS_dotplot$TE_rest)

ggplot(TE_models_CDS_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[TE_models_CDS_dotplot$TE_rest>4.3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1500))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




TE_models_CDS_dotplot$rank_TE_PMA <- rank(-TE_models_CDS_dotplot$TE_PMA)

ggplot(TE_models_CDS_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[TE_models_CDS_dotplot$TE_PMA>4.3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1400))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


```




```{r 3UTR}
TE_models_UTR3_dotplot <-TE_models_NAs[TE_models_NAs$region=="UTR3",]

TE_models_UTR3_dotplot$LFC_rest_PMA <- TE_models_UTR3_dotplot$TE_PMA - TE_models_UTR3_dotplot$TE_rest

TE_models_UTR3_dotplot <- do.call(data.frame,lapply(TE_models_UTR3_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_UTR3_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[abs(TE_models_UTR3_dotplot$LFC_rest_PMA)>1 & (TE_models_UTR3_dotplot$TE_rest>1.5 | TE_models_UTR3_dotplot$TE_PMA>1.5),], aes(label=ID),size=2)+
  theme_minimal()+
  theme(aspect.ratio = 1)



TE_models_UTR3_dotplot$rank_TE_rest <- rank(-TE_models_UTR3_dotplot$TE_rest)

ggplot(TE_models_UTR3_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[TE_models_UTR3_dotplot$TE_rest>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1200))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




TE_models_UTR3_dotplot$rank_TE_PMA <- rank(-TE_models_UTR3_dotplot$TE_PMA)

ggplot(TE_models_UTR3_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[TE_models_UTR3_dotplot$TE_PMA>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1200))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




```






```{r 5UTR}
TE_models_UTR5_dotplot <-TE_models_NAs[TE_models_NAs$region=="UTR5",]

TE_models_UTR5_dotplot$LFC_rest_PMA <- TE_models_UTR5_dotplot$TE_PMA - TE_models_UTR5_dotplot$TE_rest

TE_models_UTR5_dotplot <- do.call(data.frame,lapply(TE_models_UTR5_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_UTR5_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[abs(TE_models_UTR5_dotplot$LFC_rest_PMA)>1 & (TE_models_UTR5_dotplot$TE_rest>1.5 | TE_models_UTR5_dotplot$TE_PMA>1.5),], aes(label=ID),size=2)+
  theme_minimal()+
  theme(aspect.ratio = 1)



TE_models_UTR5_dotplot$rank_TE_rest <- rank(-TE_models_UTR5_dotplot$TE_rest)

ggplot(TE_models_UTR5_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[TE_models_UTR5_dotplot$TE_rest>3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,800))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)






TE_models_UTR5_dotplot$rank_TE_PMA <- rank(-TE_models_UTR5_dotplot$TE_PMA)

ggplot(TE_models_UTR5_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[TE_models_UTR5_dotplot$TE_PMA>3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,700))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




```



```{r Other}
TE_models_Other_dotplot <-TE_models_NAs[TE_models_NAs$region=="Other",]

TE_models_Other_dotplot$LFC_rest_PMA <- TE_models_Other_dotplot$TE_PMA - TE_models_Other_dotplot$TE_rest

TE_models_Other_dotplot <- do.call(data.frame,lapply(TE_models_Other_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_Other_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[abs(TE_models_Other_dotplot$LFC_rest_PMA)>1 & (TE_models_Other_dotplot$TE_rest>0 | TE_models_Other_dotplot$TE_PMA>0),], aes(label=ID),size=2.5)+
  theme_minimal()+
  theme(aspect.ratio = 1)





TE_models_Other_dotplot$rank_TE_rest <- rank(-TE_models_Other_dotplot$TE_rest)

ggplot(TE_models_Other_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[TE_models_Other_dotplot$TE_rest>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 10,force = 1)+
  scale_x_continuous(limits = c(0,80))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)






TE_models_Other_dotplot$rank_TE_PMA <- rank(-TE_models_Other_dotplot$TE_PMA)

ggplot(TE_models_Other_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[TE_models_Other_dotplot$TE_PMA>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 10,force = 1)+
  scale_x_continuous(limits = c(0,80))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)

```


