---
title: "combining protein models for visualization"
author: "Benoit Nicolet"
date: "25/05/2020"
output: html_document
---

```{r setup, include=FALSE}

#install.packages("pheatmap")

library(plyr)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(circlize)
library(pheatmap)
library(caret)
library(data.table)
library(MatrixGenerics)

# BiocManager::install("MatrixGenerics")

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set("/home/ben/Analysis/RF_human/T_cell_subsets/")
setwd("/home/ben/Analysis/RF_human/T_cell_subsets/")

```

```{r importing RNA library}

Sequence_parameters <- read.delim("/home/ben/Analysis/RF_human/Library_V2_Aug2021/Protein_per_gene_library_v2_RBP_GC_length_codon_AA_m6A_m5C_AtoI_m1A_m7G_CD8miRDB_PTM.csv",sep = "\t", dec=".")

```

```{r importing CN models}

Tn_0h_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_CN_Tn_0h_10k_15_10_2021.RDS")
Tn_6h_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_CN_Tn_6h_10k_15_10_2021.RDS")
Tn_12h_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_CN_Tn_12h_10k_15_10_2021.RDS")
Tn_24h_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_CN_Tn_24h_10k_15_10_2021.RDS")
Tn_48h_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_CN_Tn_48h_10k_15_10_2021.RDS")

# Tn_0h_model$finalModel # 42.49%
# Tn_6h_model$finalModel # 45.35%
# Tn_12h_model$finalModel # 47.1%
# Tn_24h_model$finalModel # 46.76%
# Tn_48h_model$finalModel #

Tn_0h_model<- data.frame(varImp(Tn_0h_model,scale=F)$importance)
Tn_6h_model<- data.frame(varImp(Tn_6h_model,scale=F)$importance)
Tn_12h_model<- data.frame(varImp(Tn_12h_model,scale=F)$importance)
Tn_24h_model<- data.frame(varImp(Tn_24h_model,scale=F)$importance)
Tn_48h_model<- data.frame(varImp(Tn_48h_model,scale=F)$importance)

Tn_0h_model$ID <- rownames(Tn_0h_model)
Tn_6h_model$ID <- rownames(Tn_6h_model)
Tn_12h_model$ID <- rownames(Tn_12h_model)
Tn_24h_model$ID <- rownames(Tn_24h_model)
Tn_48h_model$ID <- rownames(Tn_48h_model)

colnames(Tn_0h_model)[1] <- "Tn_0h_model"
colnames(Tn_6h_model)[1] <- "Tn_6h_model"
colnames(Tn_12h_model)[1] <- "Tn_12h_model"
colnames(Tn_24h_model)[1] <- "Tn_24h_model"
colnames(Tn_48h_model)[1] <- "Tn_48h_model"

Prot_Tn_models <- list(Tn_0h_model,Tn_6h_model,Tn_12h_model,Tn_24h_model,Tn_48h_model)

Prot_Tn_models <- join_all(Prot_Tn_models,by = "ID")
Prot_Tn_models[1:2] <- Prot_Tn_models[2:1]
colnames(Prot_Tn_models)[1:2] <- colnames(Prot_Tn_models)[2:1]
dim(Prot_Tn_models)


# Min correction #
Prot_Tn_models[2:6] <- log10(Prot_Tn_models[2:6])
Prot_Tn_models <- do.call(data.frame,lapply(Prot_Tn_models,function(x) replace(x, is.infinite(x),NA)))
Prot_Tn_models[is.na(Prot_Tn_models)]=min(Prot_Tn_models[2:6],na.rm = T)
Prot_Tn_models[2:6] <- Prot_Tn_models[2:6] - min(Prot_Tn_models[2:6])

dim(Prot_Tn_models)


Prot_Tn_models$region <- "Other"
Prot_Tn_models$region[grep("CDS", Prot_Tn_models$ID)] <- "CDS"
Prot_Tn_models$region[grep("UTR5", Prot_Tn_models$ID)] <- "UTR5"
Prot_Tn_models$region[grep("UTR3", Prot_Tn_models$ID)] <- "UTR3"

Prot_Tn_models_UTR5 <- subset(Prot_Tn_models, Prot_Tn_models$region=="UTR5")
Prot_Tn_models_CDS <- subset(Prot_Tn_models, Prot_Tn_models$region=="CDS")
Prot_Tn_models_UTR3 <- subset(Prot_Tn_models, Prot_Tn_models$region=="UTR3")
Prot_Tn_models_Other <- subset(Prot_Tn_models, Prot_Tn_models$region=="Other")

dim(Prot_Tn_models_UTR5)
dim(Prot_Tn_models_CDS)
dim(Prot_Tn_models_UTR3)
dim(Prot_Tn_models_Other)


```

```{r corplot}

col_fun_cor= colorRamp2(c(-3, -1, 0, 0.5, 2), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))


corrplot::corrplot(cor(Prot_Tn_models[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)),diag = T,order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "full lib",mar = c(0,0,1,0), type = "lower")

corrplot::corrplot(cor(Prot_Tn_models_UTR5[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)), diag = T, order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "5'UTR",mar = c(0,0,1,0), type = "lower")

corrplot::corrplot(cor(Prot_Tn_models_CDS[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)),diag = T,order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "CDS",mar = c(0,0,1,0), type = "lower")


corrplot::corrplot(cor(Prot_Tn_models_UTR3[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)), diag = T, order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "3'UTR",mar = c(0,0,1,0), type = "lower")

corrplot::corrplot(cor(Prot_Tn_models_Other[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)), diag = T, order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "other",mar = c(0,0,1,0), type = "lower")

# cor(Prot_Tn_models[,2:8],method = "pearson",use = "pairwise.complete.obs")

```

```{r corplot and heatmap on top features}



Prot_Tn_models_CDS_top <- Prot_Tn_models_CDS
Prot_Tn_models_CDS_top[2:6] <- 10^(Prot_Tn_models_CDS_top[2:6])
Prot_Tn_models_CDS_top$rowmean <- rowMeans(Prot_Tn_models_CDS_top[,2:6])
Prot_Tn_models_CDS_top <- Prot_Tn_models_CDS_top[order(Prot_Tn_models_CDS_top$rowmean,decreasing = T),]
Prot_Tn_models_CDS_top <- Prot_Tn_models_CDS_top[1:100,]
Prot_Tn_models_CDS_top[2:6] <- log10(Prot_Tn_models_CDS_top[2:6])
Prot_Tn_models_CDS_top <- do.call(data.frame,lapply(Prot_Tn_models_CDS_top,function(x) replace(x, is.infinite(x),NA)))
dim(Prot_Tn_models_CDS_top)
#Prot_Tn_models_CDS_top$rowmean <- NULL

Prot_Tn_models_UTR5_top <- Prot_Tn_models_UTR5
Prot_Tn_models_UTR5_top[2:6] <- 10^(Prot_Tn_models_UTR5_top[2:6])
Prot_Tn_models_UTR5_top$rowmean <- rowMeans(Prot_Tn_models_UTR5_top[,2:6])
Prot_Tn_models_UTR5_top <- Prot_Tn_models_UTR5_top[order(Prot_Tn_models_UTR5_top$rowmean,decreasing = T),]
Prot_Tn_models_UTR5_top <- Prot_Tn_models_UTR5_top[1:100,]
Prot_Tn_models_UTR5_top[2:6] <- log10(Prot_Tn_models_UTR5_top[2:6])
Prot_Tn_models_UTR5_top <- do.call(data.frame,lapply(Prot_Tn_models_UTR5_top,function(x) replace(x, is.infinite(x),NA)))
dim(Prot_Tn_models_UTR5_top)


Prot_Tn_models_UTR3_top <- Prot_Tn_models_UTR3
Prot_Tn_models_UTR3_top[2:6] <- 10^(Prot_Tn_models_UTR3_top[2:6])
Prot_Tn_models_UTR3_top$rowmean <- rowMeans(Prot_Tn_models_UTR3_top[,2:6])
Prot_Tn_models_UTR3_top <- Prot_Tn_models_UTR3_top[order(Prot_Tn_models_UTR3_top$rowmean,decreasing = T),]
Prot_Tn_models_UTR3_top <- Prot_Tn_models_UTR3_top[1:100,]
Prot_Tn_models_UTR3_top[2:6] <- log10(Prot_Tn_models_UTR3_top[2:6])
Prot_Tn_models_UTR3_top <- do.call(data.frame,lapply(Prot_Tn_models_UTR3_top,function(x) replace(x, is.infinite(x),NA)))
dim(Prot_Tn_models_UTR3_top)


Prot_Tn_models_Other_top <- Prot_Tn_models_Other
Prot_Tn_models_Other_top[2:6] <- 10^(Prot_Tn_models_Other_top[2:6])
Prot_Tn_models_Other_top$rowmean <- rowMeans(Prot_Tn_models_Other_top[,2:6])
Prot_Tn_models_Other_top <- Prot_Tn_models_Other_top[order(Prot_Tn_models_Other_top$rowmean,decreasing = T),]
Prot_Tn_models_Other_top <- Prot_Tn_models_Other_top[1:25,]
Prot_Tn_models_Other_top[2:6] <- log10(Prot_Tn_models_Other_top[2:6])
Prot_Tn_models_Other_top <- do.call(data.frame,lapply(Prot_Tn_models_Other_top,function(x) replace(x, is.infinite(x),NA)))
dim(Prot_Tn_models_Other_top)


## Heatmap time ##
col_fun_HM = colorRamp2(c(6.5,7.5,9.5), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(Prot_Tn_models_UTR5_top[2:6], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(Prot_Tn_models_UTR5_top[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =Prot_Tn_models_UTR5_top$ID, cellwidth = 8, main = "5'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

pheatmap(Prot_Tn_models_CDS_top[2:6], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(Prot_Tn_models_CDS_top[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =Prot_Tn_models_CDS_top$ID, cellwidth = 8, main = "CDS",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

pheatmap(Prot_Tn_models_UTR3_top[2:6], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(Prot_Tn_models_UTR3_top[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =Prot_Tn_models_UTR3_top$ID, cellwidth = 8, main = "3'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

pheatmap(Prot_Tn_models_Other_top[2:6], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(Prot_Tn_models_Other_top[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =Prot_Tn_models_Other$ID, cellwidth = 8, cellheight = 4, main = "general",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


```

```{r k-means clustering}
# Prot_Tn_models[is.na(Prot_Tn_models)]=0
Prot_Tn_models_tops <- Prot_Tn_models
Prot_Tn_models_tops$row_min <- do.call(pmin, as.data.frame(Prot_Tn_models_tops[2:6]))
Prot_Tn_models_tops <- subset(Prot_Tn_models_tops,Prot_Tn_models_tops$row_min>1)


Prot_Tn_models_scaled <- data.frame(Prot_Tn_models_tops[1:1],t(scale(t(Prot_Tn_models_tops[2:6]))),Prot_Tn_models_tops[7:7])

set.seed(1234)
col_fun= colorRamp2(c(-3, -1.5, 0, 1.5, 3), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))
Prot_Tn_models_kmeans <- pheatmap(Prot_Tn_models_scaled[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(Prot_Tn_models_scaled[2:6]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =Prot_Tn_models_scaled$ID,cellwidth = 20, cellheight = 20,kmeans_k = 8,clustering_method = "centroid")


Prot_Tn_models_kmeans$kmeans$cluster

Prot_Tn_models_scaled$cluster <- Prot_Tn_models_kmeans$kmeans$cluster

Prot_Tn_models_scaled <- Prot_Tn_models_scaled[order(Prot_Tn_models_scaled$cluster,decreasing = F),]

pheatmap(Prot_Tn_models_scaled[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(Prot_Tn_models_scaled[2:6]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =Prot_Tn_models_scaled$ID,cellwidth = 20)

Prot_Tn_models_scaled <- data.frame(Prot_Tn_models_scaled)
Prot_Tn_models_scaled$rowvar <- rowVars(as.matrix(Prot_Tn_models_scaled[2:6]))
Prot_Tn_models_scaled <- subset(Prot_Tn_models_scaled,Prot_Tn_models_scaled$rowvar>0.1)
Prot_Tn_models_scaled$rowvar <- NULL
Prot_Tn_models_scaled <- data.table(Prot_Tn_models_scaled)

Prot_Tn_models_scaled_melted <- melt.data.table(Prot_Tn_models_scaled,id.vars = c("ID","cluster"))

Prot_Tn_models_scaled_melted$value <- as.numeric(Prot_Tn_models_scaled_melted$value)

Prot_Tn_models_scaled_melted <- Prot_Tn_models_scaled_melted[Prot_Tn_models_scaled_melted$variable!="region"]

mean_data <- dplyr::group_by(Prot_Tn_models_scaled_melted,variable, cluster) %>%
             summarise(value = mean(value, na.rm = TRUE))

ggplot(Prot_Tn_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  geom_line(data=mean_data, aes(group = cluster), alpha=1, color="red")+
  facet_wrap(. ~ cluster, ncol = 4)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


write.table(Prot_Tn_models_scaled,"/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Protein_naive_t_models_integrated_with_kmers.csv", sep=";",dec = ",",row.names = F)





# ## frequencies in clusters ##
# 
# region_freq_in_clusters_Prot <- rbind(data.frame(table(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==1,]$region)/dim(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==1])[1]*100, "cluster"=1),
# data.frame(table(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==2,]$region)/dim(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==2])[1]*100, "cluster"=2),
# data.frame(table(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==3,]$region)/dim(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==3])[1]*100, "cluster"=3),
# data.frame(table(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==4,]$region)/dim(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==4])[1]*100, "cluster"=4),
# data.frame(table(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==5,]$region)/dim(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==5])[1]*100, "cluster"=5),
# data.frame(table(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==6,]$region)/dim(Prot_Tn_models_scaled[Prot_Tn_models_scaled$cluster==6])[1]*100, "cluster"=6)
# )
# 
# region_freq_in_clusters_Prot
# 
# 
# ggplot(region_freq_in_clusters_Prot,aes(y=Freq, x=Var1, fill=Var1))+
#   geom_bar(stat = 'identity')+
#   scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
#   facet_wrap(. ~ cluster, ncol = 1)+
#   theme_minimal()+
#   theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r importance per region}

Prot_Tn_models_for_imp_plot <- Prot_Tn_models
# Prot_Tn_models_for_imp_plot[Prot_Tn_models_for_imp_plot==min(Prot_Tn_models_for_imp_plot[,2:6])]=NA
# Prot_Tn_models_for_imp_plot[2:6] <- Prot_Tn_models_for_imp_plot[2:6] - min(Prot_Tn_models_for_imp_plot[2:6])
Prot_Tn_models_for_imp_plot <- melt(Prot_Tn_models_for_imp_plot)


ggplot(Prot_Tn_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 1)


## average feature imp ##
Prot_Tn_models_avg_imp <- data.frame("ID"=Prot_Tn_models$ID,"rowmean"=rowMeans(Prot_Tn_models[,2:6]),"region"=Prot_Tn_models$region)
Prot_Tn_models_avg_imp <- melt(Prot_Tn_models_avg_imp)


ggplot(Prot_Tn_models_avg_imp,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = region), position = position_jitterdodge(jitter.width = 0.7, dodge.width = 0.8), size = 0.75, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.7, color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.7, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 3)


```

```{r capturing the variance of regions}

Prot_Tn_models_var <- data.frame(Prot_Tn_models)
Prot_Tn_models_var[2:6] <- 10^(Prot_Tn_models_var[2:6])

Prot_Tn_models_var[is.na(Prot_Tn_models_var)]=0
Prot_Tn_models_var$row_mean <- rowMeans(as.matrix(Prot_Tn_models_var[2:6]))

Prot_Tn_models_var$row_sd <- matrixStats::rowSds(as.matrix(Prot_Tn_models_var[2:6]))
Prot_Tn_models_var$corrected_sd <- Prot_Tn_models_var$row_sd/Prot_Tn_models_var$row_mean
Prot_Tn_models_var$row_mean <- log10(Prot_Tn_models_var$row_mean)
Prot_Tn_models_var[2:6] <- log10(Prot_Tn_models_var[2:6])

plot(Prot_Tn_models_var$row_sd,Prot_Tn_models_var$row_mean)
plot(Prot_Tn_models_var$corrected_sd,Prot_Tn_models_var$row_mean)

ggplot(Prot_Tn_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r full variablity assessment}

Prot_Tn_models_var$row_min <- do.call(pmin, as.data.frame(Prot_Tn_models_var[2:6]))
table(Prot_Tn_models_var$row_min==0)

dim(Prot_Tn_models_var[(Prot_Tn_models_var$corrected_sd < median(Prot_Tn_models_var$corrected_sd,na.rm = T)) & (Prot_Tn_models_var$row_min>0),])

dim(Prot_Tn_models_var[(Prot_Tn_models_var$row_min>0),])

core_features_RNA <- data.frame(
  table(Prot_Tn_models_var[(Prot_Tn_models_var$corrected_sd < median(Prot_Tn_models_var$corrected_sd,na.rm = T)) & (Prot_Tn_models_var$row_min>0),]$region)/table(Prot_Tn_models_var$region)*100)
core_features_RNA$group <- "C_core"

HV_core_features_RNA <- data.frame(
  table(Prot_Tn_models_var[(Prot_Tn_models_var$corrected_sd < median(Prot_Tn_models_var$corrected_sd,na.rm = T)) & (Prot_Tn_models_var$row_min==0),]$region)/table(Prot_Tn_models_var$region)*100)
HV_core_features_RNA$group <- "A_hyper_variable"

core_features_RNA <- rbind(core_features_RNA, HV_core_features_RNA, data.frame("Var1"=c("UTR5","CDS","UTR3","Other"),"Freq"= 100-(core_features_RNA$Freq + HV_core_features_RNA$Freq),"group"="B_variable"))


ggplot(core_features_RNA,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  # scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)
  

ggplot(Prot_Tn_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme(aspect.ratio = 1)


ggplot(Prot_Tn_models_var[Prot_Tn_models_var$row_min>0,],aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme(aspect.ratio = 1)


```

```{r capturing the variance of regions}
non_zero_Prot_Tn_models <- Prot_Tn_models_var

non_zero_Prot_Tn_models <- subset(non_zero_Prot_Tn_models,non_zero_Prot_Tn_models$row_min>0)
dim(non_zero_Prot_Tn_models)

ggplot(non_zero_Prot_Tn_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(non_zero_Prot_Tn_models,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

non_zero_Prot_Tn_models$region <- factor(non_zero_Prot_Tn_models$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(non_zero_Prot_Tn_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(non_zero_Prot_Tn_models$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)






```

```{r }
Prot_Tn_models_codons <- Prot_Tn_models
Prot_Tn_models_codons$region[grep("codon", Prot_Tn_models_codons$ID)] <- "codon"

Prot_Tn_models_codons <- subset(Prot_Tn_models_codons, Prot_Tn_models_codons$region=="codon")

dim(Prot_Tn_models_codons)


#col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green
col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

Prot_Tn_models_codons <- Prot_Tn_models_codons[order(Prot_Tn_models_codons$ID),]


pheatmap(Prot_Tn_models_codons[2:6], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(Prot_Tn_models_codons[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =Prot_Tn_models_codons$ID, cellwidth = 8,main = "codons",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

```

```{r }
Prot_Tn_models_aminos <- Prot_Tn_models
Prot_Tn_models_aminos$region[grep("amino", Prot_Tn_models_aminos$ID)] <- "amino"

Prot_Tn_models_aminos <- subset(Prot_Tn_models_aminos, Prot_Tn_models_aminos$region=="amino")

dim(Prot_Tn_models_aminos)


col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

Prot_Tn_models_aminos <- Prot_Tn_models_aminos[order(Prot_Tn_models_aminos$ID),]


pheatmap(Prot_Tn_models_aminos[2:6], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(Prot_Tn_models_aminos[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =Prot_Tn_models_aminos$ID, cellwidth = 8,main = "amino acids",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)



```

```{r prep data for GO}
## Here we make a list of genes that are expressed in T cells to limit the GO search to genes found in cluster ##

# importing counts of T cell subsets #
Tn_0h_CN <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/Protein/CD4_Tn_0h_CN_log10.csv", sep = ";",dec = ",")
Tn_6h_CN <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/Protein/CD4_Tn_6h_CN_log10.csv", sep = ";",dec = ",")
Tn_12h_CN <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/Protein/CD4_Tn_12h_CN_log10.csv", sep = ";",dec = ",")
Tn_24h_CN <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/Protein/CD4_Tn_24h_CN_log10.csv", sep = ";",dec = ",")
Tn_48h_CN <- read.delim("/home/ben/Analysis/RF_human/T_cell_subsets/Protein/CD4_Tn_48h_CN_log10.csv", sep = ";",dec = ",")

Tn_activation_CN <- rbind(Tn_0h_CN,Tn_6h_CN,Tn_12h_CN,Tn_24h_CN,Tn_48h_CN)
Tn_activation_CN <- Tn_activation_CN$ID %>% unique() %>% unique(fromLast=T)
Tn_activation_CN <- data.frame("ID"= Tn_activation_CN)
length(Tn_activation_CN)

# tx2gene <- data.frame("tx.id"=Sequence_parameters$tx.id, "gene_name"= Sequence_parameters$gene_name)

# CD8_genes_expressed <- merge(CD8_TPM_log10, tx2gene, by="tx.id")
Tn_activation_CN <- Tn_activation_CN$ID %>% unique() %>% unique(fromLast=T)
length(Tn_activation_CN) # 7392

```

```{r getting Gos per cluster for CD8+}

Tn_activation_CN_models_scaled <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Protein_naive_t_models_integrated_with_kmers.csv", sep=";",dec = ",")

Tn_activation_CN_models_cluster1 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==1,]
Tn_activation_CN_models_cluster2 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==2,]
Tn_activation_CN_models_cluster3 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==3,]
Tn_activation_CN_models_cluster4 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==4,]
Tn_activation_CN_models_cluster5 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==5,]
Tn_activation_CN_models_cluster6 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==6,]
Tn_activation_CN_models_cluster7 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==7,]
Tn_activation_CN_models_cluster8 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==8,]
# Tn_activation_CN_models_cluster9 <- Tn_activation_CN_models_scaled[Tn_activation_CN_models_scaled$cluster==9,]

dim(Tn_activation_CN_models_cluster1)
dim(Tn_activation_CN_models_cluster2)
dim(Tn_activation_CN_models_cluster3)
dim(Tn_activation_CN_models_cluster4)
dim(Tn_activation_CN_models_cluster5)
dim(Tn_activation_CN_models_cluster6)
dim(Tn_activation_CN_models_cluster7)
dim(Tn_activation_CN_models_cluster8)
# dim(Tn_activation_CN_models_cluster9)







Tn_activation_CN_models_cluster1 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster1$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster2 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster2$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster3 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster3$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster4 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster4$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster5 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster5$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster6 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster6$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster7 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster7$ID],"gene_name"=Sequence_parameters$gene_name)
Tn_activation_CN_models_cluster8 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster8$ID],"gene_name"=Sequence_parameters$gene_name)
# Tn_activation_CN_models_cluster9 <- cbind(Sequence_parameters[,Tn_activation_CN_models_cluster9$ID],"gene_name"=Sequence_parameters$gene_name)



dim(Tn_activation_CN_models_cluster1)
dim(Tn_activation_CN_models_cluster2)
dim(Tn_activation_CN_models_cluster3)
dim(Tn_activation_CN_models_cluster4)
dim(Tn_activation_CN_models_cluster5)
dim(Tn_activation_CN_models_cluster6)
dim(Tn_activation_CN_models_cluster7)
dim(Tn_activation_CN_models_cluster8)
# dim(Tn_activation_CN_models_cluster9)



Tn_activation_CN_models_cluster1$rowSums <- rowSums(Tn_activation_CN_models_cluster1[1:dim(Tn_activation_CN_models_cluster1)[2]-1])
Tn_activation_CN_models_cluster2$rowSums <- rowSums(Tn_activation_CN_models_cluster2[1:dim(Tn_activation_CN_models_cluster2)[2]-1])
Tn_activation_CN_models_cluster3$rowSums <- rowSums(Tn_activation_CN_models_cluster3[1:dim(Tn_activation_CN_models_cluster3)[2]-1])
Tn_activation_CN_models_cluster4$rowSums <- rowSums(Tn_activation_CN_models_cluster4[1:dim(Tn_activation_CN_models_cluster4)[2]-1])
Tn_activation_CN_models_cluster5$rowSums <- rowSums(Tn_activation_CN_models_cluster5[1:dim(Tn_activation_CN_models_cluster5)[2]-1])
Tn_activation_CN_models_cluster6$rowSums <- rowSums(Tn_activation_CN_models_cluster6[1:dim(Tn_activation_CN_models_cluster6)[2]-1])
Tn_activation_CN_models_cluster7$rowSums <- rowSums(Tn_activation_CN_models_cluster7[1:dim(Tn_activation_CN_models_cluster7)[2]-1])
Tn_activation_CN_models_cluster8$rowSums <- rowSums(Tn_activation_CN_models_cluster8[1:dim(Tn_activation_CN_models_cluster8)[2]-1])
# Tn_activation_CN_models_cluster9$rowSums <- rowSums(Tn_activation_CN_models_cluster9[1:dim(Tn_activation_CN_models_cluster9)[2]-1])




Tn_activation_CN_models_cluster1 <- subset(Tn_activation_CN_models_cluster1, Tn_activation_CN_models_cluster1$rowSums>100)
Tn_activation_CN_models_cluster2 <- subset(Tn_activation_CN_models_cluster2, Tn_activation_CN_models_cluster2$rowSums>100)
Tn_activation_CN_models_cluster3 <- subset(Tn_activation_CN_models_cluster3, Tn_activation_CN_models_cluster3$rowSums>100)
Tn_activation_CN_models_cluster4 <- subset(Tn_activation_CN_models_cluster4, Tn_activation_CN_models_cluster4$rowSums>100)
Tn_activation_CN_models_cluster5 <- subset(Tn_activation_CN_models_cluster5, Tn_activation_CN_models_cluster5$rowSums>100)
Tn_activation_CN_models_cluster6 <- subset(Tn_activation_CN_models_cluster6, Tn_activation_CN_models_cluster6$rowSums>100)
Tn_activation_CN_models_cluster7 <- subset(Tn_activation_CN_models_cluster7, Tn_activation_CN_models_cluster7$rowSums>100)
Tn_activation_CN_models_cluster8 <- subset(Tn_activation_CN_models_cluster8, Tn_activation_CN_models_cluster8$rowSums>100)
# Tn_activation_CN_models_cluster9 <- subset(Tn_activation_CN_models_cluster9, Tn_activation_CN_models_cluster9$rowSums>1)


gene_name_cluster1 <- Tn_activation_CN_models_cluster1$gene_name
gene_name_cluster2 <- Tn_activation_CN_models_cluster2$gene_name
gene_name_cluster3 <- Tn_activation_CN_models_cluster3$gene_name
gene_name_cluster4 <- Tn_activation_CN_models_cluster4$gene_name
gene_name_cluster5 <- Tn_activation_CN_models_cluster5$gene_name
gene_name_cluster6 <- Tn_activation_CN_models_cluster6$gene_name
gene_name_cluster7 <- Tn_activation_CN_models_cluster7$gene_name
gene_name_cluster8 <- Tn_activation_CN_models_cluster8$gene_name
# gene_name_cluster9 <- Tn_activation_CN_models_cluster9$gene_name



length(gene_name_cluster1)
length(gene_name_cluster2)
length(gene_name_cluster3)
length(gene_name_cluster4)
length(gene_name_cluster5)
length(gene_name_cluster6)
length(gene_name_cluster7)
length(gene_name_cluster8)
# length(gene_name_cluster9)


gene_name_cluster1 <- gene_name_cluster1 %>% unique() %>% unique(fromLast = T)
gene_name_cluster2 <- gene_name_cluster2 %>% unique() %>% unique(fromLast = T)
gene_name_cluster3 <- gene_name_cluster3 %>% unique() %>% unique(fromLast = T)
gene_name_cluster4 <- gene_name_cluster4 %>% unique() %>% unique(fromLast = T)
gene_name_cluster5 <- gene_name_cluster5 %>% unique() %>% unique(fromLast = T)
gene_name_cluster6 <- gene_name_cluster6 %>% unique() %>% unique(fromLast = T)
gene_name_cluster7 <- gene_name_cluster7 %>% unique() %>% unique(fromLast = T)
gene_name_cluster8 <- gene_name_cluster8 %>% unique() %>% unique(fromLast = T)
# gene_name_cluster9 <- gene_name_cluster9 %>% unique() %>% unique(fromLast = T)


gene_name_cluster1 <- gene_name_cluster1[gene_name_cluster1 %in% Tn_activation_CN]
gene_name_cluster2 <- gene_name_cluster2[gene_name_cluster2 %in% Tn_activation_CN]
gene_name_cluster3 <- gene_name_cluster3[gene_name_cluster3 %in% Tn_activation_CN]
gene_name_cluster4 <- gene_name_cluster4[gene_name_cluster4 %in% Tn_activation_CN]
gene_name_cluster5 <- gene_name_cluster5[gene_name_cluster5 %in% Tn_activation_CN]
gene_name_cluster6 <- gene_name_cluster6[gene_name_cluster6 %in% Tn_activation_CN]
gene_name_cluster7 <- gene_name_cluster7[gene_name_cluster7 %in% Tn_activation_CN]
gene_name_cluster8 <- gene_name_cluster8[gene_name_cluster8 %in% Tn_activation_CN]
# gene_name_cluster9 <- gene_name_cluster9[gene_name_cluster9 %in% Tn_activation_CN]




length(gene_name_cluster1)
length(gene_name_cluster2)
length(gene_name_cluster3)
length(gene_name_cluster4)
length(gene_name_cluster5)
length(gene_name_cluster6)
length(gene_name_cluster7)
length(gene_name_cluster8)
# length(gene_name_cluster9)

write.table(gene_name_cluster1,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster1.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster2,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster2.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster3,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster3.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster4,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster4.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster5,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster5.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster6,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster6.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster7,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster7.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)
write.table(gene_name_cluster8,"/home/ben/Analysis/RF_human/kmeans_clusters_GO/Tn_activation_gene_name_cluster8.csv",sep=";", dec = ",",row.names = F,col.names = F,quote = F)

table(gene_name_cluster3 %in% gene_name_cluster8)

```

```{r importing RNA models}

Tn_0h_RNA_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_1000nrounds_TPM_Tn_0h_15_10_2021.RDS")
Tn_6h_RNA_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_1000nrounds_TPM_Tn_6h_15_10_2021.RDS")
Tn_24h_RNA_model <- readRDS("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/models/xgb_1000nrounds_TPM_Tn_24h_15_10_2021.RDS")


Tn_0h_RNA_model<- data.frame(varImp(Tn_0h_RNA_model,scale=F)$importance)
Tn_6h_RNA_model<- data.frame(varImp(Tn_6h_RNA_model,scale=F)$importance)
Tn_24h_RNA_model<- data.frame(varImp(Tn_24h_RNA_model,scale=F)$importance)

Tn_0h_RNA_model$ID <- rownames(Tn_0h_RNA_model)
Tn_6h_RNA_model$ID <- rownames(Tn_6h_RNA_model)
Tn_24h_RNA_model$ID <- rownames(Tn_24h_RNA_model)

colnames(Tn_0h_RNA_model)[1] <- "Tn_0h_RNA_model"
colnames(Tn_6h_RNA_model)[1] <- "Tn_6h_RNA_model"
colnames(Tn_24h_RNA_model)[1] <- "Tn_24h_RNA_model"

RNA_Tn_models <- list(Tn_0h_RNA_model,Tn_6h_RNA_model,Tn_24h_RNA_model)

RNA_Tn_models <- join_all(RNA_Tn_models,by = "ID")
RNA_Tn_models[2:1] <- RNA_Tn_models[1:2]
colnames(RNA_Tn_models)[1:2] <- colnames(RNA_Tn_models)[2:1]
dim(RNA_Tn_models)




# Min correction # ,scale=F
RNA_Tn_models[2:4] <- log10(RNA_Tn_models[2:4])
RNA_Tn_models <- do.call(data.frame,lapply(RNA_Tn_models,function(x) replace(x, is.infinite(x),NA)))
RNA_Tn_models[is.na(RNA_Tn_models)]=min(RNA_Tn_models[2:4],na.rm = T)
RNA_Tn_models[2:4] <- RNA_Tn_models[2:4] - min(RNA_Tn_models[2:4])

dim(RNA_Tn_models)


# RNA_Tn_models[2:4] <- log2(RNA_Tn_models[2:4])

dim(RNA_Tn_models)
RNA_Tn_models$region <- "Other"
RNA_Tn_models$region[grep("CDS", RNA_Tn_models$ID)] <- "CDS"
RNA_Tn_models$region[grep("UTR5", RNA_Tn_models$ID)] <- "UTR5"
RNA_Tn_models$region[grep("UTR3", RNA_Tn_models$ID)] <- "UTR3"

dim(RNA_Tn_models)



RNA_Tn_models_UTR5 <- subset(RNA_Tn_models, RNA_Tn_models$region=="UTR5")
RNA_Tn_models_CDS <- subset(RNA_Tn_models, RNA_Tn_models$region=="CDS")
RNA_Tn_models_UTR3 <- subset(RNA_Tn_models, RNA_Tn_models$region=="UTR3")
RNA_Tn_models_Other <- subset(RNA_Tn_models, RNA_Tn_models$region=="Other")


dim(RNA_Tn_models_UTR5)
dim(RNA_Tn_models_CDS)
dim(RNA_Tn_models_UTR3)
dim(RNA_Tn_models_Other)


```

```{r corplot}

col_fun_cor= colorRamp2(c(-3, -1, 0, 0.5, 2), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))


corrplot::corrplot(cor(RNA_Tn_models[,2:4],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)),diag = T,order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "full lib",mar = c(0,0,1,0), type = "lower")

corrplot::corrplot(cor(RNA_Tn_models_UTR5[,2:4],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)), diag = T, order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "5'UTR",mar = c(0,0,1,0), type = "lower")

corrplot::corrplot(cor(RNA_Tn_models_CDS[,2:4],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)),diag = T,order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "CDS",mar = c(0,0,1,0), type = "lower")


corrplot::corrplot(cor(RNA_Tn_models_UTR3[,2:4],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)), diag = T, order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "3'UTR",mar = c(0,0,1,0), type = "lower")

corrplot::corrplot(cor(RNA_Tn_models_Other[,2:4],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun_cor(seq(-9,3,by=0.01)), diag = T, order = "original",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "other",mar = c(0,0,1,0), type = "lower")

# cor(RNA_Tn_models[,2:8],method = "pearson",use = "pairwise.complete.obs")

```

```{r corplot and heatmap on top features}



RNA_Tn_models_CDS_top <- RNA_Tn_models_CDS
RNA_Tn_models_CDS_top[2:4] <- 10^(RNA_Tn_models_CDS_top[2:4])
RNA_Tn_models_CDS_top$rowmean <- rowMeans(RNA_Tn_models_CDS_top[,2:4])
RNA_Tn_models_CDS_top <- RNA_Tn_models_CDS_top[order(RNA_Tn_models_CDS_top$rowmean,decreasing = T),]
RNA_Tn_models_CDS_top <- RNA_Tn_models_CDS_top[1:100,]
RNA_Tn_models_CDS_top[2:4] <- log10(RNA_Tn_models_CDS_top[2:4])
RNA_Tn_models_CDS_top <- do.call(data.frame,lapply(RNA_Tn_models_CDS_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_Tn_models_CDS_top)
#RNA_Tn_models_CDS_top$rowmean <- NULL

RNA_Tn_models_UTR5_top <- RNA_Tn_models_UTR5
RNA_Tn_models_UTR5_top[2:4] <- 10^(RNA_Tn_models_UTR5_top[2:4])
RNA_Tn_models_UTR5_top$rowmean <- rowMeans(RNA_Tn_models_UTR5_top[,2:4])
RNA_Tn_models_UTR5_top <- RNA_Tn_models_UTR5_top[order(RNA_Tn_models_UTR5_top$rowmean,decreasing = T),]
RNA_Tn_models_UTR5_top <- RNA_Tn_models_UTR5_top[1:100,]
RNA_Tn_models_UTR5_top[2:4] <- log10(RNA_Tn_models_UTR5_top[2:4])
RNA_Tn_models_UTR5_top <- do.call(data.frame,lapply(RNA_Tn_models_UTR5_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_Tn_models_UTR5_top)


RNA_Tn_models_UTR3_top <- RNA_Tn_models_UTR3
RNA_Tn_models_UTR3_top[2:4] <- 10^(RNA_Tn_models_UTR3_top[2:4])
RNA_Tn_models_UTR3_top$rowmean <- rowMeans(RNA_Tn_models_UTR3_top[,2:4])
RNA_Tn_models_UTR3_top <- RNA_Tn_models_UTR3_top[order(RNA_Tn_models_UTR3_top$rowmean,decreasing = T),]
RNA_Tn_models_UTR3_top <- RNA_Tn_models_UTR3_top[1:100,]
RNA_Tn_models_UTR3_top[2:4] <- log10(RNA_Tn_models_UTR3_top[2:4])
RNA_Tn_models_UTR3_top <- do.call(data.frame,lapply(RNA_Tn_models_UTR3_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_Tn_models_UTR3_top)


RNA_Tn_models_Other_top <- RNA_Tn_models_Other
RNA_Tn_models_Other_top[2:4] <- 10^(RNA_Tn_models_Other_top[2:4])
RNA_Tn_models_Other_top$rowmean <- rowMeans(RNA_Tn_models_Other_top[,2:4])
RNA_Tn_models_Other_top <- RNA_Tn_models_Other_top[order(RNA_Tn_models_Other_top$rowmean,decreasing = T),]
RNA_Tn_models_Other_top <- RNA_Tn_models_Other_top[1:25,]
RNA_Tn_models_Other_top[2:4] <- log10(RNA_Tn_models_Other_top[2:4])
RNA_Tn_models_Other_top <- do.call(data.frame,lapply(RNA_Tn_models_Other_top,function(x) replace(x, is.infinite(x),NA)))
dim(RNA_Tn_models_Other_top)


## Heatmap time ##
col_fun_HM = colorRamp2(c(6.5,7.5,9.5), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(RNA_Tn_models_UTR5_top[2:4], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_Tn_models_UTR5_top[2:4]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_Tn_models_UTR5_top$ID, cellwidth = 8, main = "5'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

pheatmap(RNA_Tn_models_CDS_top[2:4], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_Tn_models_CDS_top[2:4]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_Tn_models_CDS_top$ID, cellwidth = 8, main = "CDS",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

pheatmap(RNA_Tn_models_UTR3_top[2:4], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_Tn_models_UTR3_top[2:4]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_Tn_models_UTR3_top$ID, cellwidth = 8, main = "3'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

pheatmap(RNA_Tn_models_Other_top[2:4], col_fun_HM(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_Tn_models_Other_top[2:4]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =RNA_Tn_models_Other$ID, cellwidth = 8, cellheight = 4, main = "general",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


```

```{r k-means clustering}
# RNA_Tn_models[is.na(RNA_Tn_models)]=0
RNA_Tn_models_tops <- RNA_Tn_models
RNA_Tn_models_tops$row_min <- do.call(pmin, as.data.frame(RNA_Tn_models_tops[2:4]))
RNA_Tn_models_tops <- subset(RNA_Tn_models_tops,RNA_Tn_models_tops$row_min>1)


RNA_Tn_models_scaled <- data.frame(RNA_Tn_models_tops[1:1],t(scale(t(RNA_Tn_models_tops[2:4]))),RNA_Tn_models_tops[5:5])

set.seed(1234)
col_fun= colorRamp2(c(-3, -1.5, 0, 1.5, 3), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))
RNA_Tn_models_kmeans <- pheatmap(RNA_Tn_models_scaled[2:4], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RNA_Tn_models_scaled[2:4]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_Tn_models_scaled$ID,cellwidth = 20, cellheight = 20,kmeans_k = 4,clustering_method = "centroid")


RNA_Tn_models_kmeans$kmeans$cluster

RNA_Tn_models_scaled$cluster <- RNA_Tn_models_kmeans$kmeans$cluster

RNA_Tn_models_scaled <- RNA_Tn_models_scaled[order(RNA_Tn_models_scaled$cluster,decreasing = F),]

pheatmap(RNA_Tn_models_scaled[2:4], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_Tn_models_scaled[2:4]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =RNA_Tn_models_scaled$ID,cellwidth = 20)

RNA_Tn_models_scaled <- data.frame(RNA_Tn_models_scaled)
RNA_Tn_models_scaled$rowvar <- rowVars(as.matrix(RNA_Tn_models_scaled[2:4]))
RNA_Tn_models_scaled <- subset(RNA_Tn_models_scaled,RNA_Tn_models_scaled$rowvar>0.1)
RNA_Tn_models_scaled$rowvar <- NULL
RNA_Tn_models_scaled <- data.table(RNA_Tn_models_scaled)

RNA_Tn_models_scaled_melted <- melt.data.table(RNA_Tn_models_scaled,id.vars = c("ID","cluster"))

RNA_Tn_models_scaled_melted$value <- as.numeric(RNA_Tn_models_scaled_melted$value)

RNA_Tn_models_scaled_melted <- RNA_Tn_models_scaled_melted[RNA_Tn_models_scaled_melted$variable!="region"]

mean_data <- dplyr::group_by(RNA_Tn_models_scaled_melted,variable, cluster) %>%
             summarise(value = mean(value, na.rm = TRUE))

ggplot(RNA_Tn_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  geom_line(data=mean_data, aes(group = cluster), alpha=1, color="red")+
  facet_wrap(. ~ cluster, ncol = 4)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


write.table(RNA_Tn_models_scaled,"/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/RNA_models_Tn_models_integrated_with_kmers.csv", sep=";",dec = ",",row.names = F)

RNA_Tn_models_scaled <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/RNA_models_Tn_models_integrated_with_kmers.csv", sep=";",dec = ",")




# ## frequencies in clusters ##
# 
region_freq_in_clusters_Prot <- rbind(data.frame(table(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==1,]$region)/dim(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==1])[1]*100, "cluster"=1),
data.frame(table(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==2,]$region)/dim(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==2])[1]*100, "cluster"=2),
data.frame(table(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==3,]$region)/dim(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==3])[1]*100, "cluster"=3),
data.frame(table(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==4,]$region)/dim(RNA_Tn_models_scaled[RNA_Tn_models_scaled$cluster==4])[1]*100, "cluster"=4) )

region_freq_in_clusters_Prot


ggplot(region_freq_in_clusters_Prot,aes(y=Freq, x=Var1, fill=Var1))+
  geom_bar(stat = 'identity')+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  facet_wrap(. ~ cluster, ncol = 1)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

```{r importance per region}

RNA_Tn_models_for_imp_plot <- RNA_Tn_models
# RNA_Tn_models_for_imp_plot[RNA_Tn_models_for_imp_plot==min(RNA_Tn_models_for_imp_plot[,2:4])]=NA
# RNA_Tn_models_for_imp_plot[2:4] <- RNA_Tn_models_for_imp_plot[2:4] - min(RNA_Tn_models_for_imp_plot[2:4])
RNA_Tn_models_for_imp_plot <- melt(RNA_Tn_models_for_imp_plot)


ggplot(RNA_Tn_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 1)


## average feature imp ##
RNA_Tn_models_avg_imp <- data.frame("ID"=RNA_Tn_models$ID,"rowmean"=rowMeans(RNA_Tn_models[,2:4]),"region"=RNA_Tn_models$region)
RNA_Tn_models_avg_imp <- melt(RNA_Tn_models_avg_imp)


ggplot(RNA_Tn_models_avg_imp,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = region), position = position_jitterdodge(jitter.width = 0.7, dodge.width = 0.8), size = 0.75, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.7, color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.7, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 3)


```

```{r capturing the variance of regions}

RNA_Tn_models_var <- data.frame(RNA_Tn_models)
RNA_Tn_models_var[2:4] <- 10^(RNA_Tn_models_var[2:4])

RNA_Tn_models_var[is.na(RNA_Tn_models_var)]=0
RNA_Tn_models_var$row_mean <- rowMeans(as.matrix(RNA_Tn_models_var[2:4]))

RNA_Tn_models_var$row_sd <- matrixStats::rowSds(as.matrix(RNA_Tn_models_var[2:4]))
RNA_Tn_models_var$corrected_sd <- RNA_Tn_models_var$row_sd/RNA_Tn_models_var$row_mean
RNA_Tn_models_var$row_mean <- log10(RNA_Tn_models_var$row_mean)
RNA_Tn_models_var[2:4] <- log10(RNA_Tn_models_var[2:4])

plot(RNA_Tn_models_var$row_sd,RNA_Tn_models_var$row_mean)
plot(RNA_Tn_models_var$corrected_sd,RNA_Tn_models_var$row_mean)

ggplot(RNA_Tn_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r full variablity assessment}

RNA_Tn_models_var$row_min <- do.call(pmin, as.data.frame(RNA_Tn_models_var[2:4]))
table(RNA_Tn_models_var$row_min==0)

dim(RNA_Tn_models_var[(RNA_Tn_models_var$corrected_sd < median(RNA_Tn_models_var$corrected_sd,na.rm = T)) & (RNA_Tn_models_var$row_min>0),])

dim(RNA_Tn_models_var[(RNA_Tn_models_var$row_min>0),])

core_features_RNA <- data.frame(
  table(RNA_Tn_models_var[(RNA_Tn_models_var$corrected_sd < median(RNA_Tn_models_var$corrected_sd,na.rm = T)) & (RNA_Tn_models_var$row_min>0),]$region)/table(RNA_Tn_models_var$region)*100)
core_features_RNA$group <- "C_core"

HV_core_features_RNA <- data.frame(
  table(RNA_Tn_models_var[(RNA_Tn_models_var$corrected_sd < median(RNA_Tn_models_var$corrected_sd,na.rm = T)) & (RNA_Tn_models_var$row_min==0),]$region)/table(RNA_Tn_models_var$region)*100)
HV_core_features_RNA$group <- "A_hyper_variable"

core_features_RNA <- rbind(core_features_RNA, HV_core_features_RNA, data.frame("Var1"=c("CDS","Other","UTR3","UTR5"),"Freq"= 100-(core_features_RNA$Freq + HV_core_features_RNA$Freq),"group"="B_variable"))


ggplot(core_features_RNA,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  scale_y_continuous(expand = c(0,0))+
  theme_classic()+
  theme(aspect.ratio = 2)
  

ggplot(RNA_Tn_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme(aspect.ratio = 1)


ggplot(RNA_Tn_models_var[RNA_Tn_models_var$row_min>0,],aes(x=corrected_sd,y=row_mean))+
  geom_point(alpha=0.1)+
  theme(aspect.ratio = 1)


```

```{r capturing the variance of regions}
non_zero_RNA_Tn_models <- RNA_Tn_models_var

non_zero_RNA_Tn_models <- subset(non_zero_RNA_Tn_models,non_zero_RNA_Tn_models$row_min>0)
dim(non_zero_RNA_Tn_models)

ggplot(non_zero_RNA_Tn_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(non_zero_RNA_Tn_models,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

non_zero_RNA_Tn_models$region <- factor(non_zero_RNA_Tn_models$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(non_zero_RNA_Tn_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(non_zero_RNA_Tn_models$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)






```

```{r }
RNA_Tn_models_codons <- RNA_Tn_models
RNA_Tn_models_codons$region[grep("codon", RNA_Tn_models_codons$ID)] <- "codon"

RNA_Tn_models_codons <- subset(RNA_Tn_models_codons, RNA_Tn_models_codons$region=="codon")

dim(RNA_Tn_models_codons)


#col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green
col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

RNA_Tn_models_codons <- RNA_Tn_models_codons[order(RNA_Tn_models_codons$ID),]


pheatmap(RNA_Tn_models_codons[2:4], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_Tn_models_codons[2:4]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_Tn_models_codons$ID, cellwidth = 8,main = "codons",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)

```

```{r }
RNA_Tn_models_aminos <- RNA_Tn_models
RNA_Tn_models_aminos$region[grep("amino", RNA_Tn_models_aminos$ID)] <- "amino"

RNA_Tn_models_aminos <- subset(RNA_Tn_models_aminos, RNA_Tn_models_aminos$region=="amino")

dim(RNA_Tn_models_aminos)


col_fun = colorRamp2(c(7,8.5,10), c("#FFFFFF","#1E8449","#000000")) 

RNA_Tn_models_aminos <- RNA_Tn_models_aminos[order(RNA_Tn_models_aminos$ID),]


pheatmap(RNA_Tn_models_aminos[2:4], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RNA_Tn_models_aminos[2:4]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RNA_Tn_models_aminos$ID, cellwidth = 8,main = "amino acids",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10)



```

```{r importing RNA data for plots}


CD4_RNA_0h_train <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/split_sets/TPM_0h_train",sep = ";", dec=",")
CD4_RNA_0h_test <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/split_sets/TPM_0h_test",sep = ";", dec=",")

CD4_RNA_6h_train <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/split_sets/TPM_6h_train",sep = ";", dec=",")
CD4_RNA_6h_test <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/split_sets/TPM_6h_test",sep = ";", dec=",")

CD4_RNA_24h_train <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/split_sets/TPM_24h_train",sep = ";", dec=",")
CD4_RNA_24h_test <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/split_sets/TPM_24h_test",sep = ";", dec=",")


CD4_RNA_0h <- rbind(CD4_RNA_0h_train,CD4_RNA_0h_test)
CD4_RNA_6h <- rbind(CD4_RNA_6h_train,CD4_RNA_6h_test)
CD4_RNA_24h <- rbind(CD4_RNA_24h_train,CD4_RNA_24h_test)

dim(CD4_RNA_0h)
dim(CD4_RNA_6h)
dim(CD4_RNA_24h)


CD4_RNA_0h[2:dim(CD4_RNA_0h)[2]-1] <- lapply(CD4_RNA_0h[2:dim(CD4_RNA_0h)[2]-1], function(x) {
  if(is.character(x)) as.numeric(as.character(x)) else x
})

CD4_RNA_6h[2:dim(CD4_RNA_6h)[2]-1] <- lapply(CD4_RNA_6h[2:dim(CD4_RNA_6h)[2]-1], function(x) {
  if(is.character(x)) as.numeric(as.character(x)) else x
})

CD4_RNA_24h[2:dim(CD4_RNA_24h)[2]-1] <- lapply(CD4_RNA_24h[2:dim(CD4_RNA_24h)[2]-1], function(x) {
  if(is.character(x)) as.numeric(as.character(x)) else x
})

CD4_RNA_0h$group <- "0h"
CD4_RNA_6h$group <- "6h"
CD4_RNA_24h$group <- "24h"

```

```{r plotting interesting codons}

# GAT

ggplot(CD4_RNA_0h, aes(x=(GAT_codon_CDS), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)

lm(CD4_RNA_0h$TPM ~ CD4_RNA_0h$GAT_codon_CDS,na.action = na.omit) # 0.01974
lm(CD4_RNA_6h$TPM ~ CD4_RNA_6h$GAT_codon_CDS,na.action = na.omit) # 0.04239
lm(CD4_RNA_24h$TPM ~ CD4_RNA_24h$GAT_codon_CDS,na.action = na.omit) # 0.03869

lm_GAT <- data.frame("condition"=c("0h","6h","24h"),
                     "lm"=c(0.01974,0.04239,0.03869),
                     "XGB"=c(7.906459,8.234213,8.400316))

ggplot(lm_GAT,aes(x=condition, y=(lm)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_GAT,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)



# CGA

ggplot(CD4_RNA_0h, aes(x=(CGA_codon_CDS), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)

lm(CD4_RNA_0h$TPM ~ CD4_RNA_0h$CGA_codon_CDS,na.action = na.omit) # 0.06478
lm(CD4_RNA_6h$TPM ~ CD4_RNA_6h$CGA_codon_CDS,na.action = na.omit) # 0.1132
lm(CD4_RNA_24h$TPM ~ CD4_RNA_24h$CGA_codon_CDS,na.action = na.omit) # 0.09329

lm_CGA <- data.frame("condition"=c("0h","6h","24h"),
                     "lm"=c(0.06478,0.1132,0.09329),
                     "XGB"=c(8.362780,8.566990,8.129615))

ggplot(lm_CGA,aes(x=condition, y=(lm)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_CGA,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)


# AAG

ggplot(CD4_RNA_0h, aes(x=log10(AAG_codon_CDS), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)

lm(CD4_RNA_0h$TPM ~ CD4_RNA_0h$AAG_codon_CDS,na.action = na.omit) # 0.01862
lm(CD4_RNA_6h$TPM ~ CD4_RNA_6h$AAG_codon_CDS,na.action = na.omit) # 0.0265
lm(CD4_RNA_24h$TPM ~ CD4_RNA_24h$AAG_codon_CDS,na.action = na.omit) # 0.02886


lm_AAG <- data.frame("condition"=c("0h","6h","24h"),
                     "lm"=c(0.01862,0.0265,0.02886),
                     "XGB"=c(8.607527,8.641046,8.453814))

ggplot(lm_AAG,aes(x=condition, y=abs(lm)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_AAG,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

# TGG

ggplot(CD4_RNA_0h, aes(x=(TGG_codon_CDS), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red", fullrange= T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


lm(CD4_RNA_0h$TPM ~ CD4_RNA_0h$TGG_codon_CDS,na.action = na.omit) # -0.1052
lm(CD4_RNA_6h$TPM ~ CD4_RNA_6h$TGG_codon_CDS,na.action = na.omit) # -0.1040
lm(CD4_RNA_24h$TPM ~ CD4_RNA_24h$TGG_codon_CDS,na.action = na.omit) # -0.08757 

lm_TGG <- data.frame("condition"=c("0h","6h","24h"),
                     "lm"=c(-0.1052,-0.1040,-0.08757),
                     "XGB"=c(7.805555,7.502656,7.601395))

ggplot(lm_TGG,aes(x=condition, y=abs(lm)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_TGG,aes(x=condition, y=abs(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)




## tests ##

ggplot(CD4_RNA_0h, aes(x=(CGT_codon_CDS), y=TPM, group=group))+
  geom_pointdensity(alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  theme_minimal()+
  theme(aspect.ratio = 1)



ggplot(CD4_RNA_0h, aes(x=(GAT_codon_CDS), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red", fullrange= T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r codon correlation through activation}

## combine plot ##
CD4_RNA_0h_codons <- melt(data.frame("TPM"=CD4_RNA_0h$TPM ,
                                "TGG_codon_CDS"=CD4_RNA_0h$TGG_codon_CDS,
                                "AAG_codon_CDS"=CD4_RNA_0h$AAG_codon_CDS,
                                "CAG_codon_CDS"=CD4_RNA_0h$CAG_codon_CDS,
                                "CGT_codon_CDS"=CD4_RNA_0h$CGT_codon_CDS),measure.vars = c("TGG_codon_CDS","AAG_codon_CDS","CAG_codon_CDS","CGT_codon_CDS") )


ggplot(CD4_RNA_0h_codons, aes(x=value, y=TPM))+
  geom_pointdensity(alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(method = "lm", color="red", fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(variable,levels = c("AAG_codon_CDS","CGT_codon_CDS","CAG_codon_CDS","TGG_codon_CDS")), ncol = 4)+
  scale_y_continuous(limits = c(-1,4.5), expand = c(0,0))+
  theme_minimal()+
  theme(aspect.ratio = 1)


## calculating the signedness of codons ##
CD4_RNA_0h_codons <- data.frame("TPM"=CD4_RNA_0h$TPM, CD4_RNA_0h[,grepl("codon",names(CD4_RNA_0h))])
CD4_RNA_6h_codons <- data.frame("TPM"=CD4_RNA_6h$TPM, CD4_RNA_6h[,grepl("codon",names(CD4_RNA_6h))])
CD4_RNA_24h_codons <- data.frame("TPM"=CD4_RNA_24h$TPM, CD4_RNA_24h[,grepl("codon",names(CD4_RNA_24h))])

for (i in 2:dim(CD4_RNA_0h_codons)[2]) {
  df <- lm(CD4_RNA_0h_codons$TPM ~ CD4_RNA_0h_codons[,i])
  datalist[[i]] <- df
}

for (i in 2:dim(CD4_RNA_6h_codons)[2]) {
  df <- lm(CD4_RNA_6h_codons$TPM ~ CD4_RNA_6h_codons[,i])
  datalist[[i]] <- df
}

for (i in 2:dim(CD4_RNA_24h_codons)[2]) {
  df <- lm(CD4_RNA_24h_codons$TPM ~ CD4_RNA_24h_codons[,i])
  datalist[[i]] <- df
}

CD4_RNA_0h_codons <- data.frame("cor"=cor(CD4_RNA_0h_codons)[1,2:65],"ID"=colnames(CD4_RNA_0h_codons[-1]),"group"="0h")
CD4_RNA_6h_codons <- data.frame("cor"=cor(CD4_RNA_6h_codons)[1,2:65],"ID"=colnames(CD4_RNA_6h_codons[-1]),"group"="6h")
CD4_RNA_24h_codons <- data.frame("cor"=cor(CD4_RNA_24h_codons)[1,2:65],"ID"=colnames(CD4_RNA_24h_codons[-1]),"group"="24h")

CD4_RNA_cor_codons <- rbind(CD4_RNA_0h_codons,CD4_RNA_6h_codons,CD4_RNA_24h_codons)

CD4_RNA_cor_codons$ID <- gsub("_codon_CDS","",CD4_RNA_cor_codons$ID)

ggplot(CD4_RNA_cor_codons, aes(x=ID,y=cor, fill=cor))+
  geom_bar(stat = "identity",color="black")+
  theme_minimal()+
  scale_fill_gradient2(low = "#154360", high = "#ff0000", mid = "#ffffff",space = "Lab", guide = "colourbar", aesthetics = "fill")+
  scale_y_continuous(limits = c(-0.2,0.1),expand = c(0,0))+
  facet_wrap(.~group,ncol=1)+
  theme(aspect.ratio = 0.1,axis.text.x = element_text(angle = 90))


```

```{r plotting interesting 5UTR features}

# 

ggplot(CD4_RNA_0h, aes(x=(AGCAC_UTR5), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



# 

ggplot(CD4_RNA_0h, aes(x=log10(UTR5_length_UTR5), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



# 

ggplot(CD4_RNA_0h, aes(x=log10(CACA_UTR5), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


# 

ggplot(CD4_RNA_0h, aes(x=(GGGG_UTR5), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



# 

ggplot(CD4_RNA_0h, aes(x=(GGCG_UTR5), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)






```

```{r plotting interesting CDS features}

# 

ggplot(CD4_RNA_0h, aes(x=log10(CDS_m7G), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



# 

ggplot(CD4_RNA_0h, aes(x=log10(CDS_m1A), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



# 

ggplot(CD4_RNA_0h, aes(x=log10(CDS_m6a), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


# 

ggplot(CD4_RNA_0h, aes(x=log10(CCTC_CDS), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "gam")+
  geom_smooth(data = CD4_RNA_6h,method = "gam")+
  geom_smooth(data = CD4_RNA_24h,method = "gam")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red")+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red")+
  geom_hline(yintercept = 0, linetype="dotted")+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



# 

ggplot(CD4_RNA_0h, aes(x=log10(UTR5_length_UTR5), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red", fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red", fullrange=T)+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  scale_x_continuous(limits = c(0.5,4))+
  theme_minimal()+
  theme(aspect.ratio = 1)



CD4_RNA_0h$log10_UTR5_length_UTR5 <- log10(CD4_RNA_0h$UTR5_length_UTR5)
CD4_RNA_6h$log10_UTR5_length_UTR5 <- log10(CD4_RNA_6h$UTR5_length_UTR5)
CD4_RNA_24h$log10_UTR5_length_UTR5 <- log10(CD4_RNA_24h$UTR5_length_UTR5)

CD4_RNA_0h$log10_UTR5_length_UTR5 <- as.numeric(gsub("-Inf","NA",CD4_RNA_0h$log10_UTR5_length_UTR5))
CD4_RNA_6h$log10_UTR5_length_UTR5 <- as.numeric(gsub("-Inf","NA",CD4_RNA_6h$log10_UTR5_length_UTR5))
CD4_RNA_24h$log10_UTR5_length_UTR5 <- as.numeric(gsub("-Inf","NA",CD4_RNA_24h$log10_UTR5_length_UTR5))

cor(CD4_RNA_0h$TPM, CD4_RNA_0h$log10_UTR5_length_UTR5,use = "pairwise.complete.obs") # 0.01409145
cor(CD4_RNA_6h$TPM, CD4_RNA_6h$log10_UTR5_length_UTR5,use = "pairwise.complete.obs") # -0.02656101
cor(CD4_RNA_24h$TPM, CD4_RNA_24h$log10_UTR5_length_UTR5,use = "pairwise.complete.obs") # -0.05351587


lm_UTR5_length <- data.frame("condition"=c("0h","6h","24h"),
                     "cor"=c(0.01409145,-0.02656101,-0.05351587),
                     "XGB"=c(8.568977,8.399713,8.352577))

ggplot(lm_UTR5_length,aes(x=condition, y=(cor)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_UTR5_length,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)





```

```{r m7G in regions throughout activation}

# m7G

ggplot(CD4_RNA_0h, aes(x=log10(total_m7G), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



CD4_RNA_0h$log10_total_m7G <- log10(CD4_RNA_0h$total_m7G)
CD4_RNA_6h$log10_total_m7G <- log10(CD4_RNA_6h$total_m7G)
CD4_RNA_24h$log10_total_m7G <- log10(CD4_RNA_24h$total_m7G)

CD4_RNA_0h$log10_total_m7G <- as.numeric(gsub("-Inf","NA",CD4_RNA_0h$log10_total_m7G))
CD4_RNA_6h$log10_total_m7G <- as.numeric(gsub("-Inf","NA",CD4_RNA_6h$log10_total_m7G))
CD4_RNA_24h$log10_total_m7G <- as.numeric(gsub("-Inf","NA",CD4_RNA_24h$log10_total_m7G))

cor(CD4_RNA_0h$TPM, CD4_RNA_0h$log10_total_m7G,use = "pairwise.complete.obs") # 0.2420157
cor(CD4_RNA_6h$TPM, CD4_RNA_6h$log10_total_m7G,use = "pairwise.complete.obs") # 0.2913818
cor(CD4_RNA_24h$TPM, CD4_RNA_24h$log10_total_m7G,use = "pairwise.complete.obs") # 0.2967893


lm_total_m7G <- data.frame("condition"=c("0h","6h","24h"),
                     "cor"=c(0.2420157,0.2913818,0.2967893),
                     "XGB"=c(9.327762,9.725776,9.668995))

ggplot(lm_total_m7G ,aes(x=condition, y=(cor)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_total_m7G ,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)









# 
# ggplot(CD4_RNA_0h, aes(x=log10(UTR5_m7G), y=TPM, group=group))+
#   geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
#   geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
#   geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
#   scale_color_viridis(option = "H")+
#   geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
#   geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
#   geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
#   geom_hline(yintercept = 0, linetype="dotted")+
#   scale_x_continuous(limits = c(0,3))+
#   facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
#   theme_minimal()+
#   theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(CDS_m7G), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(UTR3_m7G), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



```

```{r m6A in regions through activation}

# m6A

ggplot(CD4_RNA_0h, aes(x=log10(total_m6A), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(UTR5_m6a), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(CDS_m6a), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)




CD4_RNA_0h$log10_CDS_m6a <- log10(CD4_RNA_0h$CDS_m6a)
CD4_RNA_6h$log10_CDS_m6a <- log10(CD4_RNA_6h$CDS_m6a)
CD4_RNA_24h$log10_CDS_m6a <- log10(CD4_RNA_24h$CDS_m6a)

CD4_RNA_0h$log10_CDS_m6a <- as.numeric(gsub("-Inf","NA",CD4_RNA_0h$log10_CDS_m6a))
CD4_RNA_6h$log10_CDS_m6a <- as.numeric(gsub("-Inf","NA",CD4_RNA_6h$log10_CDS_m6a))
CD4_RNA_24h$log10_CDS_m6a <- as.numeric(gsub("-Inf","NA",CD4_RNA_24h$log10_CDS_m6a))


cor(CD4_RNA_0h$TPM, CD4_RNA_0h$log10_CDS_m6a,use = "pairwise.complete.obs") # 0.03858643
cor(CD4_RNA_6h$TPM, CD4_RNA_6h$log10_CDS_m6a,use = "pairwise.complete.obs") # 0.006029353
cor(CD4_RNA_24h$TPM, CD4_RNA_24h$log10_CDS_m6a,use = "pairwise.complete.obs") # -0.0355017


lm_CDS_m6a <- data.frame("condition"=c("0h","6h","24h"),
                     "lm"=c(0.03858643,0.006029353,-0.0355017),
                     "XGB"=c(8.460099,8.599940,8.820129))

ggplot(lm_CDS_m6a,aes(x=condition, y=(lm)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_CDS_m6a,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)






ggplot(CD4_RNA_0h, aes(x=log10(UTR3_m6a), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)




```

```{r m1A in regions through activation}

# m1A

ggplot(CD4_RNA_0h, aes(x=log10(total_m1A), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(UTR5_m1A), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(CDS_m1A), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)





ggplot(CD4_RNA_0h, aes(x=(UTR3_m1A), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,20))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)
# 
# CD4_RNA_0h$log10_UTR3_m1A <- log10(CD4_RNA_0h$UTR3_m1A)
# CD4_RNA_6h$log10_UTR3_m1A <- log10(CD4_RNA_6h$UTR3_m1A)
# CD4_RNA_24h$log10_UTR3_m1A <- log10(CD4_RNA_24h$UTR3_m1A)
# 
# CD4_RNA_0h$log10_UTR3_m1A <- as.numeric(gsub("-Inf","NA",CD4_RNA_0h$log10_UTR3_m1A))
# CD4_RNA_6h$log10_UTR3_m1A <- as.numeric(gsub("-Inf","NA",CD4_RNA_6h$log10_UTR3_m1A))
# CD4_RNA_24h$log10_UTR3_m1A <- as.numeric(gsub("-Inf","NA",CD4_RNA_24h$log10_UTR3_m1A))
# 
# cor(CD4_RNA_0h$TPM, CD4_RNA_0h$UTR3_m1A,use = "pairwise.complete.obs") # 0.1050803
# cor(CD4_RNA_6h$TPM, CD4_RNA_6h$UTR3_m1A,use = "pairwise.complete.obs") # 0.09776499
# cor(CD4_RNA_24h$TPM, CD4_RNA_24h$UTR3_m1A,use = "pairwise.complete.obs") #  0.1139342
# 
# 
# lm_UTR3_m1A <- data.frame("condition"=c("0h","6h","24h"),
#                      "lm"=c(0.1050803,0.09776499,0.1139342),
#                      "XGB"=c(7.088472,6.969945,7.557673))
# 
# ggplot(lm_UTR3_m1A,aes(x=condition, y=abs(lm)))+
#   geom_bar(stat="identity")+
#   scale_x_discrete(limit = c("0h","6h","24h"))+
#   theme_minimal()+
#   theme(aspect.ratio = 2)
# 
# ggplot(lm_UTR3_m1A,aes(x=condition, y=10^(XGB)))+
#   geom_bar(stat="identity")+
#   scale_x_discrete(limit = c("0h","6h","24h"))+
#   theme_minimal()+
#   theme(aspect.ratio = 2)



```

```{r m5C in regions through activation}

# m5C

ggplot(CD4_RNA_0h, aes(x=log10(total_m5C), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,4))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)



CD4_RNA_0h$log10_total_m5C <- log10(CD4_RNA_0h$total_m5C)
CD4_RNA_6h$log10_total_m5C <- log10(CD4_RNA_6h$total_m5C)
CD4_RNA_24h$log10_total_m5C <- log10(CD4_RNA_24h$total_m5C)

CD4_RNA_0h$log10_total_m5C <- as.numeric(gsub("-Inf","NA",CD4_RNA_0h$log10_total_m5C))
CD4_RNA_6h$log10_total_m5C <- as.numeric(gsub("-Inf","NA",CD4_RNA_6h$log10_total_m5C))
CD4_RNA_24h$log10_total_m5C <- as.numeric(gsub("-Inf","NA",CD4_RNA_24h$log10_total_m5C))

cor(CD4_RNA_0h$TPM, CD4_RNA_0h$log10_total_m5C,use = "pairwise.complete.obs") # 0.01047208
cor(CD4_RNA_6h$TPM, CD4_RNA_6h$log10_total_m5C,use = "pairwise.complete.obs") # -0.02549405
cor(CD4_RNA_24h$TPM, CD4_RNA_24h$log10_total_m5C,use = "pairwise.complete.obs") # -0.03979533


lm_total_m5C <- data.frame("condition"=c("0h","6h","24h"),
                     "cor"=c(0.01047208,-0.02549405,-0.03979533),
                     "XGB"=c(8.995858,8.913800,8.902540))

ggplot(lm_total_m5C,aes(x=condition, y=(cor)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)

ggplot(lm_total_m5C,aes(x=condition, y=10^(XGB)))+
  geom_bar(stat="identity")+
  scale_x_discrete(limit = c("0h","6h","24h"))+
  theme_minimal()+
  theme(aspect.ratio = 2)










ggplot(CD4_RNA_0h, aes(x=log10(UTR5_m5C), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(CDS_m5C), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CD4_RNA_0h, aes(x=log10(UTR3_m5C), y=TPM, group=group))+
  geom_pointdensity(data = CD4_RNA_0h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_6h,alpha=0.1,show.legend = F, stroke = 0)+
  geom_pointdensity(data = CD4_RNA_24h,alpha=0.1,show.legend = F, stroke = 0)+
  scale_color_viridis(option = "H")+
  geom_smooth(data = CD4_RNA_0h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_6h,method = "lm", color="red",fullrange=T)+
  geom_smooth(data = CD4_RNA_24h,method = "lm", color="red",fullrange=T)+
  geom_hline(yintercept = 0, linetype="dotted")+
  scale_x_continuous(limits = c(0,3))+
  facet_wrap(.~ factor(group, levels=c('0h','6h','24h')), ncol = 3)+
  theme_minimal()+
  theme(aspect.ratio = 1)





```
