---
title: "Visualizing XGB output in RBP-KO"
author: "Benoit Nicolet"
date: "09/05/2022"
output: html_document
---


```{r setup, include=FALSE}

#install.packages("pheatmap")

library(plyr)
library(dplyr)
library(ggplot2)
library(ggfortify)
library(tidyverse)
library(circlize)
library(pheatmap)
library(caret)
library(ggpointdensity)
library(viridis)
library(reshape)
library(GOexpress)
library(data.table)

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set("/home/ben/Analysis/RF_human/immune_system/")
setwd("/home/ben/Analysis/RF_human/immune_system/")

```





## RBP Models ##
```{r importing RBP models}

NT <- readRDS("/home/ben/Analysis/RF_human/RBP_validation/models/xgb_TPM_NT_03_05_2022.RDS")
FXR2 <- readRDS("/home/ben/Analysis/RF_human/RBP_validation/models/xgb_TPM_FXR2_03_05_2022.RDS")
FUBP3 <- readRDS("/home/ben/Analysis/RF_human/RBP_validation/models/xgb_TPM_FUBP3_03_05_2022.RDS")
ELAVL1 <- readRDS("/home/ben/Analysis/RF_human/RBP_validation/models/xgb_TPM_ELAVL1_03_05_2022.RDS")
PDCD4 <- readRDS("/home/ben/Analysis/RF_human/RBP_validation/models/xgb_TPM_PDCD4_03_05_2022.RDS")

NT <- data.frame(caret::varImp(NT)$importance)
FXR2<- data.frame(caret::varImp(FXR2)$importance)
FUBP3<- data.frame(caret::varImp(FUBP3)$importance)
ELAVL1<- data.frame(caret::varImp(ELAVL1)$importance)
PDCD4<- data.frame(caret::varImp(PDCD4)$importance)


NT$ID <- rownames(NT)
FXR2$ID <- rownames(FXR2)
FUBP3$ID <- rownames(FUBP3)
ELAVL1$ID <- rownames(ELAVL1)
PDCD4$ID <- rownames(PDCD4)



colnames(NT)[1] <- "NT"
colnames(FXR2)[1] <- "FXR2"
colnames(FUBP3)[1] <- "FUBP3"
colnames(ELAVL1)[1] <- "ELAVL1"
colnames(PDCD4)[1] <- "PDCD4"

RBP_KO_models <- list(NT,FXR2,FUBP3,ELAVL1,PDCD4)

RBP_KO_models <- plyr::join_all(RBP_KO_models,by = "ID")
RBP_KO_models[1:2] <- RBP_KO_models[2:1]
colnames(RBP_KO_models)[1:2] <- colnames(RBP_KO_models)[2:1]
dim(RBP_KO_models)



## median correction ##
#RBP_KO_models <- data.frame(RBP_KO_models[1:1],scale(RBP_KO_models[2:23],center = F))

RBP_KO_models[2:6] <- log2(RBP_KO_models[2:6])

RBP_KO_models <- do.call(data.frame,lapply(RBP_KO_models,function(x) replace(x, is.infinite(x),NA)))

# RBP_KO_models_median_cor <- RBP_KO_models
# RBP_KO_models_median_cor[,2:23] <- 2^RBP_KO_models_median_cor[,2:23]
# RBP_KO_models_median_cor[is.na(RBP_KO_models_median_cor)]=0
# RBP_KO_models_median_cor[,2:23] <- RBP_KO_models_median_cor[,2:23] - mapply(median,RBP_KO_models_median_cor[,2:23])
# RBP_KO_models_median_cor[,2:23] <- log2(RBP_KO_models_median_cor[,2:23])
# RBP_KO_models_median_cor <- do.call(data.frame,lapply(RBP_KO_models_median_cor,function(x) replace(x, is.infinite(x),NA)))
# RBP_KO_models_median_cor[is.na(RBP_KO_models_median_cor)]=min(RBP_KO_models_median_cor[,2:23],na.rm = T)
# RBP_KO_models <- RBP_KO_models_median_cor

#RBP_KO_models[is.na(RBP_KO_models)]=min(RBP_KO_models[2:23],na.rm = T)

dim(RBP_KO_models)
RBP_KO_models$region <- "Other"
RBP_KO_models$region[grep("CDS", RBP_KO_models$ID)] <- "CDS"
RBP_KO_models$region[grep("UTR5", RBP_KO_models$ID)] <- "UTR5"
RBP_KO_models$region[grep("UTR3", RBP_KO_models$ID)] <- "UTR3"
#RBP_KO_models$region[grep("codon", RBP_KO_models$ID)] <- "codon"

dim(RBP_KO_models)


RBP_KO_models_CDS <- subset(RBP_KO_models, RBP_KO_models$region=="CDS")
dim(RBP_KO_models_CDS)

RBP_KO_models_UTR5 <- subset(RBP_KO_models, RBP_KO_models$region=="UTR5")
dim(RBP_KO_models_UTR5)

RBP_KO_models_UTR3 <- subset(RBP_KO_models, RBP_KO_models$region=="UTR3")
dim(RBP_KO_models_UTR3)

RBP_KO_models_Other <- subset(RBP_KO_models, RBP_KO_models$region=="Other")
dim(RBP_KO_models_Other)


```




```{r corplot}
# RBP_KO_models_no_NA <- RBP_KO_models
# RBP_KO_models_no_NA[is.na(RBP_KO_models_no_NA)]=min(RBP_KO_models_no_NA[2:6],na.rm = T)

# col_fun= colorRamp2(c(-3, -1.5, 0, 1.5, 3), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))
col_fun= colorRamp2(c(-2, -0.5, 0, 0.5, 2), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))
# pheatmap((cor(RBP_KO_models[,2:6],method = "spearman",use = "pairwise.complete.obs")),
#          col=col_fun(seq(-0,8,by=0.01)), clustering_method = "complete", scale = "none",cellwidth = 7, cellheight = 7,
#          treeheight_row = 10, treeheight_col = 10,fontsize_row = 7, fontsize_col = 7,main = "full lib")
# 
# pheatmap((cor(RBP_KO_models_UTR5[,2:6],method = "spearman",use = "pairwise.complete.obs")),
#          col=col_fun(seq(-0,8,by=0.01)), clustering_method = "complete", scale = "none",cellwidth = 7, cellheight = 7,
#          treeheight_row = 10, treeheight_col = 10,fontsize_row = 7, fontsize_col = 7,main = "5'UTR lib")
# 
# pheatmap((cor(RBP_KO_models_CDS[,2:6],method = "spearman",use = "pairwise.complete.obs")),
#          col=col_fun(seq(-0,8,by=0.01)), clustering_method = "complete", scale = "none",cellwidth = 7, cellheight = 7,
#          treeheight_row = 10, treeheight_col = 10,fontsize_row = 7, fontsize_col = 7,main = "CDS lib")
# 
# pheatmap((cor(RBP_KO_models_UTR3[,2:6],method = "spearman",use = "pairwise.complete.obs")),
#          col=col_fun(seq(0,8,by=0.01)), clustering_method = "complete", scale = "none",cellwidth = 7, cellheight = 7,
#          treeheight_row = 10, treeheight_col = 10,fontsize_row = 7, fontsize_col = 7,main = "3'UTR lib")


corrplot::corrplot(cor(RBP_KO_models[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-8.5,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "full lib",mar = c(0,0,1,0))

corrplot::corrplot(cor(RBP_KO_models_UTR5[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-8.5,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "5'UTR",mar = c(0,0,1,0))

corrplot::corrplot(cor(RBP_KO_models_CDS[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-8.5,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "CDS",mar = c(0,0,1,0))


corrplot::corrplot(cor(RBP_KO_models_UTR3[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-8.5,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "3'UTR",mar = c(0,0,1,0))

corrplot::corrplot(cor(RBP_KO_models_Other[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-8.5,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "other",mar = c(0,0,1,0))

# cor(RBP_KO_models[,2:8],method = "pearson",use = "pairwise.complete.obs")

```



```{r corplot and heatmap on top features}



RBP_KO_models_CDS_top <- RBP_KO_models_CDS
RBP_KO_models_CDS_top[2:6] <- 2^(RBP_KO_models_CDS_top[2:6])
RBP_KO_models_CDS_top$rowmean <- rowMeans(RBP_KO_models_CDS_top[,2:6])
RBP_KO_models_CDS_top <- RBP_KO_models_CDS_top[order(RBP_KO_models_CDS_top$rowmean,decreasing = T),]
RBP_KO_models_CDS_top <- RBP_KO_models_CDS_top[1:100,]
RBP_KO_models_CDS_top[2:6] <- log2(RBP_KO_models_CDS_top[2:6])
RBP_KO_models_CDS_top <- do.call(data.frame,lapply(RBP_KO_models_CDS_top,function(x) replace(x, is.infinite(x),NA)))
dim(RBP_KO_models_CDS_top)
#RBP_KO_models_CDS_top$rowmean <- NULL

RBP_KO_models_UTR5_top <- RBP_KO_models_UTR5
RBP_KO_models_UTR5_top[2:6] <- 2^(RBP_KO_models_UTR5_top[2:6])
RBP_KO_models_UTR5_top$rowmean <- rowMeans(RBP_KO_models_UTR5_top[,2:6])
RBP_KO_models_UTR5_top <- RBP_KO_models_UTR5_top[order(RBP_KO_models_UTR5_top$rowmean,decreasing = T),]
RBP_KO_models_UTR5_top <- RBP_KO_models_UTR5_top[1:100,]
RBP_KO_models_UTR5_top[2:6] <- log2(RBP_KO_models_UTR5_top[2:6])
RBP_KO_models_UTR5_top <- do.call(data.frame,lapply(RBP_KO_models_UTR5_top,function(x) replace(x, is.infinite(x),NA)))
dim(RBP_KO_models_UTR5_top)


RBP_KO_models_UTR3_top <- RBP_KO_models_UTR3
RBP_KO_models_UTR3_top[2:6] <- 2^(RBP_KO_models_UTR3_top[2:6])
RBP_KO_models_UTR3_top$rowmean <- rowMeans(RBP_KO_models_UTR3_top[,2:6])
RBP_KO_models_UTR3_top <- RBP_KO_models_UTR3_top[order(RBP_KO_models_UTR3_top$rowmean,decreasing = T),]
RBP_KO_models_UTR3_top <- RBP_KO_models_UTR3_top[1:100,]
RBP_KO_models_UTR3_top[2:6] <- log2(RBP_KO_models_UTR3_top[2:6])
RBP_KO_models_UTR3_top <- do.call(data.frame,lapply(RBP_KO_models_UTR3_top,function(x) replace(x, is.infinite(x),NA)))
dim(RBP_KO_models_UTR3_top)


RBP_KO_models_Other_top <- RBP_KO_models_Other
RBP_KO_models_Other_top[2:6] <- 2^(RBP_KO_models_Other_top[2:6])
RBP_KO_models_Other_top$rowmean <- rowMeans(RBP_KO_models_Other_top[,2:6])
RBP_KO_models_Other_top <- RBP_KO_models_Other_top[order(RBP_KO_models_Other_top$rowmean,decreasing = T),]
RBP_KO_models_Other_top <- RBP_KO_models_Other_top[1:25,]
RBP_KO_models_Other_top[2:6] <- log2(RBP_KO_models_Other_top[2:6])
RBP_KO_models_Other_top <- do.call(data.frame,lapply(RBP_KO_models_Other_top,function(x) replace(x, is.infinite(x),NA)))
dim(RBP_KO_models_Other_top)


col_fun = colorRamp2(c(-1, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(RBP_KO_models_Other_top[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "column",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_Other_top[2:6]), fontsize_row=4,na_col = "#515a5a",
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_Other$ID, cellwidth = 8, main = "general",
         breaks = seq(-12,5,by=0.04),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RBP_KO_models_CDS_top[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "column",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_CDS_top[2:6]), fontsize_row=3 ,na_col = "#515a5a",
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_CDS$ID,cellwidth = 8,main = "CDS",
         breaks = seq(-12,5,by=0.04),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RBP_KO_models_UTR5_top[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "column",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_UTR5_top[2:6]), fontsize_row=3,na_col = "#515a5a",
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_UTR5$ID,cellwidth = 8,main = "5'UTR",
         breaks = seq(-12,5,by=0.04),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RBP_KO_models_UTR3_top[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "column",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_UTR3_top[2:6]), fontsize_row=3,na_col = "#515a5a",
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_UTR3$ID, cellwidth = 8,main = "3'UTR",
         breaks = seq(-12,5,by=0.04),treeheight_row = 10 ,treeheight_col = 10)






col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(RBP_KO_models_Other_top[2:6], col_fun(seq(-5,8,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_Other_top[2:6]), fontsize_row=4,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_Other$ID, cellwidth = 8, main = "general",
         breaks = seq(-7,6,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RBP_KO_models_CDS_top[2:6], col_fun(seq(-5,8,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_CDS_top[2:6]), fontsize_row=3 ,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_CDS$ID,cellwidth = 8,main = "CDS",
         breaks = seq(-7,6,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RBP_KO_models_UTR5_top[2:6], col_fun(seq(-5,8,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_UTR5_top[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_UTR5$ID,cellwidth = 8,main = "5'UTR",
         breaks = seq(-7,6,by=0.01),treeheight_row = 10 ,treeheight_col = 10)


pheatmap(RBP_KO_models_UTR3_top[2:6], col_fun(seq(-5,8,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_UTR3_top[2:6]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =RBP_KO_models_UTR3$ID, cellwidth = 8,main = "3'UTR",
         breaks = seq(-7,6,by=0.01),treeheight_row = 10 ,treeheight_col = 10)



```


```{r NT vs ELAVL1}

RBP_KO_models$LFC_ELAVL1_NT <- RBP_KO_models$NT - RBP_KO_models$ELAVL1


ggplot(RBP_KO_models,aes(x=NT,y=ELAVL1,color=abs(LFC_ELAVL1_NT)))+
  geom_point()+
  geom_abline()+
  geom_abline(intercept = c(-2,2), color='red')+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "TTTT") & (RBP_KO_models$region=="UTR3") & (RBP_KO_models$LFC_ELAVL1_NT)< -2,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = -10, nudge_y = 5,size=3)+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "TTTT") & (RBP_KO_models$region=="UTR3") & (RBP_KO_models$LFC_ELAVL1_NT)>2,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = 10, nudge_y = -5,size=3)+
  theme(aspect.ratio = 1)

ggplot(RBP_KO_models[RBP_KO_models$region=="UTR3",],aes(x=NT,y=ELAVL1))+
  geom_point()+
  #geom_point(data=RBP_KO_models[(RBP_KO_models$ID %like% "TTTT") & (RBP_KO_models$region=="UTR3") & abs(RBP_KO_models$LFC_ELAVL1_NT)>1,],color='green')+
  geom_point(data=RBP_KO_models[(RBP_KO_models$ID %like% "ATTTA") & (RBP_KO_models$region=="UTR3") & abs(RBP_KO_models$LFC_ELAVL1_NT)>1,],color='red')+
  geom_abline(color="blue")+
  scale_x_continuous(limits = c(-10,10))+
  scale_y_continuous(limits = c(-10,10))+
  geom_abline(intercept = c(-1,1), color='red')+
  theme(aspect.ratio = 1)





ggplot(RBP_KO_models[RBP_KO_models$region=="UTR3",],aes(x=NT,y=ELAVL1))+
  geom_point()+
  geom_abline(color="blue")+
  scale_x_continuous(limits = c(-10,10))+
  scale_y_continuous(limits = c(-10,10))+
  geom_abline(intercept = c(-1,1), color='red')+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "") & (RBP_KO_models$region=="UTR3") & abs(RBP_KO_models$LFC_ELAVL1_NT)>1 & RBP_KO_models$ELAVL1>0,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = -10, nudge_y = 5,size=3)+
  theme(aspect.ratio = 1)

```



```{r}

RBP_KO_models[(RBP_KO_models$ID %like% "TTT") & (RBP_KO_models$region=="UTR3"),]

ggplot(RBP_KO_models[(RBP_KO_models$ID %like% "TT") & (RBP_KO_models$region=="UTR3"),],aes(x=NT,y=ELAVL1))+
  geom_point()+
  geom_point(data=RBP_KO_models[(RBP_KO_models$ID %like% "ATTTA") & (RBP_KO_models$region=="UTR3") & abs(RBP_KO_models$LFC_ELAVL1_NT)>1,],color='red')+
  geom_abline()+
  scale_x_continuous(limits = c(-25,10))+
  scale_y_continuous(limits = c(-25,10))+
  geom_abline(intercept = c(-1,1), color='red')+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "ATTTA") & (RBP_KO_models$region=="UTR3") & (RBP_KO_models$LFC_ELAVL1_NT)< -1 & RBP_KO_models$ELAVL1>-10,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = -10, nudge_y = 5,size=3)+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "ATTTA") & (RBP_KO_models$region=="UTR3") & (RBP_KO_models$LFC_ELAVL1_NT)>1 & RBP_KO_models$NT>-10,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = 10, nudge_y = -5,size=3)+
  theme(aspect.ratio = 1)




ggplot(RBP_KO_models[(RBP_KO_models$ID %like% "TT") & (RBP_KO_models$region=="UTR3"),],aes(x=NT,y=ELAVL1))+
  geom_point()+
  geom_point(data=RBP_KO_models[(RBP_KO_models$ID %like% "TTTTT") & (RBP_KO_models$region=="UTR3") & abs(RBP_KO_models$LFC_ELAVL1_NT)>1,],color='red')+
  geom_abline()+
  scale_x_continuous(limits = c(-25,10))+
  scale_y_continuous(limits = c(-25,10))+
  geom_abline(intercept = c(-1,1), color='red')+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "TTTTT") & (RBP_KO_models$region=="UTR3") & (RBP_KO_models$LFC_ELAVL1_NT)< -1 & RBP_KO_models$ELAVL1>-10,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = -10, nudge_y = 5,size=3)+
  ggrepel::geom_text_repel(data=RBP_KO_models[(RBP_KO_models$ID %like% "TTTTT") & (RBP_KO_models$region=="UTR3") & (RBP_KO_models$LFC_ELAVL1_NT)>1 & RBP_KO_models$NT>-10,], aes(label=ID),color='black', min.segment.length = 0.0000001,nudge_x = 10, nudge_y = -5,size=3)+
  theme(aspect.ratio = 1)




ggplot(RBP_KO_models[(RBP_KO_models$ID %like% "TT") & (RBP_KO_models$region=="UTR3"),],aes(x=NT,y=ELAVL1,color=abs(LFC_ELAVL1_NT)))+
  geom_point()+
  geom_abline()+
  scale_x_continuous(limits = c(-25,10))+
  scale_y_continuous(limits = c(-25,10))+
  geom_abline(intercept = c(-1,1), color='red')+
  theme(aspect.ratio = 1)


#NT_ELAVL1 <- RBP_KO_models[(RBP_KO_models$ID %like% "TTT") & (RBP_KO_models$region=="UTR3"),]
NT_ELAVL1 <- RBP_KO_models[RBP_KO_models$region=="UTR3",]
NT_ELAVL1 <- subset(NT_ELAVL1,abs(NT_ELAVL1$LFC_ELAVL1_NT)>1)
NT_ELAVL1$FXR2 <- NULL
NT_ELAVL1$FUBP3 <- NULL
NT_ELAVL1$PDCD4 <- NULL
NT_ELAVL1$region <- NULL
NT_ELAVL1$LFC_ELAVL1_NT <- NULL
NT_ELAVL1 <- subset(NT_ELAVL1,NT_ELAVL1$NT>-2 | NT_ELAVL1$ELAVL1>-2)



NT_ELAVL1[is.na(NT_ELAVL1)]<-min(NT_ELAVL1[2:3],na.rm = T)
pheatmap(NT_ELAVL1[2:3], col_fun(seq(-5,8,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(NT_ELAVL1[2:3]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =NT_ELAVL1$ID, cellwidth = 8,main = "3'UTR",
         breaks = seq(-7,6,by=0.01),treeheight_row = 10 ,treeheight_col = 10)




# 
# t.test(NT_ELAVL1$NT, NT_ELAVL1$ELAVL1, paired=T,var.equal = F)
# 
# 
# 
# NT_ELAVL1 <- melt(NT_ELAVL1)
# 
# ggplot(NT_ELAVL1,aes(y=value, x=variable))+
#   geom_point(position = 'jitter')+
#   theme(aspect.ratio = 3)
# 
# 
# ggplot(NT_ELAVL1,aes(x=value, group=variable,color=variable))+
#   geom_density()+
#   theme(aspect.ratio = 1)


```




```{r corrplot}
## now in correlation plots ## 

col_fun= colorRamp2(c(-3, -1, 0, 1, 3), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))

# corrplot::corrplot(cor(RBP_KO_models_top[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-8.5,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
#                    tl.cex = 0.7,addrect = T)

corrplot::corrplot(cor(RBP_KO_models_UTR5_top[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "color", col=col_fun(seq(-3,2.5,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T, title = "5'UTR top")

corrplot::corrplot(cor(RBP_KO_models_CDS_top[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "color", col=col_fun(seq(-3,2.5,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T, title = "CDS top")

corrplot::corrplot(cor(RBP_KO_models_UTR3_top[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "color", col=col_fun(seq(-3,2.5,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T, title = "3'UTR top")

corrplot::corrplot(cor(RBP_KO_models_Other_top[,2:6],method = "spearman",use = "pairwise.complete.obs"),method = "color", col=col_fun(seq(-3,2.5,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T, title = "other top")




```


```{r importance per region}

RBP_KO_models_for_imp_plot <- RBP_KO_models
RBP_KO_models_for_imp_plot[RBP_KO_models_for_imp_plot==min(RBP_KO_models_for_imp_plot[,2:6])]=NA
RBP_KO_models_for_imp_plot <- melt(RBP_KO_models_for_imp_plot)


ggplot(RBP_KO_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 1)

```



```{r capturing the variance of regions}
RBP_KO_models_var <- data.frame(RBP_KO_models)
RBP_KO_models_var[2:6] <- 2^(RBP_KO_models_var[2:6])
# RBP_KO_models_var[RBP_KO_models_var==min(RBP_KO_models_var[,2:6],na.rm = T)]=NA
# RBP_KO_models_var[2:6] <- log2(RBP_KO_models_var[2:6])
RBP_KO_models_var[is.na(RBP_KO_models_var)]=min(RBP_KO_models_var[,2:6],na.rm = T)
RBP_KO_models_var$row_mean <- rowMeans(as.matrix(RBP_KO_models_var[2:6]))

RBP_KO_models_var$row_sd <- matrixStats::rowSds(as.matrix(RBP_KO_models_var[2:6]))
RBP_KO_models_var$corrected_sd <- RBP_KO_models_var$row_sd/RBP_KO_models_var$row_mean
RBP_KO_models_var$row_mean <- log2(RBP_KO_models_var$row_mean)

plot(RBP_KO_models_var$row_sd,RBP_KO_models_var$row_mean)
plot(RBP_KO_models_var$corrected_sd,RBP_KO_models_var$row_mean)

ggplot(RBP_KO_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point()+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(RBP_KO_models_var,aes(x=corrected_sd, fill=NULL, color=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(RBP_KO_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  theme(aspect.ratio = 0.5)

ggplot(RBP_KO_models_var,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

RBP_KO_models_var$region <- factor(RBP_KO_models_var$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(RBP_KO_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(RBP_KO_models_var$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)



mean(RBP_KO_models_var[RBP_KO_models_var$region=="UTR5",]$corrected_sd,na.rm = T) # = -2.38509
mean(RBP_KO_models_var[RBP_KO_models_var$region=="CDS",]$corrected_sd,na.rm = T) # = -2.031128
mean(RBP_KO_models_var[RBP_KO_models_var$region=="UTR3",]$corrected_sd,na.rm = T) # = -2.061535
mean(RBP_KO_models_var[RBP_KO_models_var$region=="Other",]$corrected_sd,na.rm = T) # = -2.781053

median(RBP_KO_models_var$corrected_sd,na.rm = T)

table(RBP_KO_models_var[RBP_KO_models_var$corrected_sd>1.5,]$region)/table(RBP_KO_models_var$region)*100



```
```{r core vs variable features}

hist(RBP_KO_models_var$corrected_sd)
median(RBP_KO_models_var$corrected_sd,na.rm = T)
mean(RBP_KO_models_var$corrected_sd,na.rm = T)

core_features_CN <- data.frame(table(RBP_KO_models_var[RBP_KO_models_var$corrected_sd<mean(RBP_KO_models_var$corrected_sd,na.rm = T),]$region)/table(RBP_KO_models_var$region)*100)
core_features_CN$group <- "B_core"

core_features_CN <- rbind(core_features_CN, data.frame("Var1"=c("UTR5","CDS","UTR3","Other"),"Freq"= 100-core_features_CN$Freq,"group"="A_variable"))


ggplot(core_features_CN,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  # scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)
  


```





```{r k-means clustering CD8 only}
# RBP_KO_models[is.na(RBP_KO_models)]=0
# RBP_KO_models_tops[6:8] <- NULL
RBP_KO_models_tops <- data.frame(RBP_KO_models)
# RBP_KO_models_tops[2:6] <- 2^RBP_KO_models_tops[2:6]
RBP_KO_models_tops$row_var <- matrixStats::rowVars(as.matrix(RBP_KO_models_tops[2:6]))
RBP_KO_models_tops$row_mean <- rowMeans(as.matrix(RBP_KO_models_tops[2:6]))

RBP_KO_models_tops$row_sd <- matrixStats::rowSds(as.matrix(RBP_KO_models_tops[2:6]))
RBP_KO_models_tops$corrected_sd <- RBP_KO_models_tops$row_sd/RBP_KO_models_tops$row_mean
RBP_KO_models_tops$row_mean <- log10(RBP_KO_models_tops$row_mean)
RBP_KO_models_tops$row_var <- log10(RBP_KO_models_tops$row_var)
RBP_KO_models_tops[2:6] <- log10(RBP_KO_models_tops[2:6])

plot(RBP_KO_models_tops$row_mean,RBP_KO_models_tops$corrected_sd)

RBP_KO_models_tops <- subset(RBP_KO_models_tops,RBP_KO_models_tops$corrected_sd>median(RBP_KO_models_tops$corrected_sd,na.rm = T))
dim(RBP_KO_models_tops)

RBP_KO_models_scaled <- data.frame(RBP_KO_models_tops[1:1],scale(RBP_KO_models_tops[2:6]),RBP_KO_models_tops[24:24])

col_fun= colorRamp2(c(-2, -0.5, 0, 0.5, 2), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))

set.seed(1234)
RBP_KO_models_kmeans <- pheatmap(RBP_KO_models_scaled[2:6], col_fun(seq(-5,5,by=0.01)) ,scale = "row",
         clustering_distance_cols = "correlation", cluster_cols = T, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_scaled[2:6]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="correlation",
         labels_row =RBP_KO_models_scaled$ID,cellwidth = 10,kmeans_k = 20,clustering_method = "complete",
         treeheight_row = 5,treeheight_col = 5)


RBP_KO_models_kmeans$kmeans$cluster

RBP_KO_models_scaled$cluster <- RBP_KO_models_kmeans$kmeans$cluster

RBP_KO_models_scaled <- RBP_KO_models_scaled[order(RBP_KO_models_scaled$cluster,decreasing = F),]

pheatmap(RBP_KO_models_scaled[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "row",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RBP_KO_models_scaled[2:6]), fontsize_row=0.001,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="correlation",
         labels_row =RBP_KO_models_scaled$ID,cellwidth = 7)

RBP_KO_models_scaled <- data.frame(RBP_KO_models_scaled)
RBP_KO_models_scaled$rowvar <- matrixStats::rowVars(as.matrix(RBP_KO_models_scaled[2:6]))
RBP_KO_models_scaled <- subset(RBP_KO_models_scaled,RBP_KO_models_scaled$rowvar>0.1)
RBP_KO_models_scaled$rowvar <- NULL
RBP_KO_models_scaled <- data.table::data.table(RBP_KO_models_scaled)

RBP_KO_models_scaled_melted <- data.table::melt.data.table(RBP_KO_models_scaled,id.vars = c("ID","cluster"))

RBP_KO_models_scaled_melted$value <- as.numeric(RBP_KO_models_scaled_melted$value)

RBP_KO_models_scaled_melted <- RBP_KO_models_scaled_melted[RBP_KO_models_scaled_melted$variable!="region"]


ggplot(RBP_KO_models_scaled_melted,aes(x=variable,y=value))+
  geom_line(aes(group = ID), alpha=0.05)+
  facet_wrap(. ~ cluster, ncol = 5)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

#write.table(RBP_KO_models_scaled,"/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Protein_naive_t_models_integrated_with_kmers.csv", sep=";",dec = ",",row.names = F)





## frequencies in clusters ##

region_freq_in_clusters_Prot <- rbind(data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==1,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==1])[1]*100, "cluster"=1),
data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==2,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==2])[1]*100, "cluster"=2),
data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==3,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==3])[1]*100, "cluster"=3),
data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==4,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==4])[1]*100, "cluster"=4),
data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==5,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==5])[1]*100, "cluster"=5),
data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==6,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==6])[1]*100, "cluster"=6),
data.frame(table(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==7,]$region)/dim(RBP_KO_models_scaled[RBP_KO_models_scaled$cluster==7])[1]*100, "cluster"=7)

)

region_freq_in_clusters_Prot


ggplot(region_freq_in_clusters_Prot,aes(y=Freq, x=Var1, fill=Var1))+
  geom_bar(stat = 'identity')+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  facet_wrap(. ~ cluster, ncol = 2)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))





```







```{r }
RBP_KO_models_codons <- RBP_KO_models
RBP_KO_models_codons$region[grep("codon", RBP_KO_models_codons$ID)] <- "codon"

RBP_KO_models_codons <- subset(RBP_KO_models_codons, RBP_KO_models_codons$region=="codon")

dim(RBP_KO_models_codons)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

RBP_KO_models_codons <- RBP_KO_models_codons[order(RBP_KO_models_codons$ID),]

pheatmap(RBP_KO_models_codons[2:6], col_fun(seq(-3,3,by=0.01)) ,scale = "column",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = F,
         labels_col=colnames(RBP_KO_models_codons[2:6]), fontsize_row=4,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",
         labels_row =RBP_KO_models_codons$ID, gaps_col = 4,cellwidth = 7, cellheight = 3)


```



```{r }
RBP_KO_models_aminos <- RBP_KO_models
RBP_KO_models_aminos$region[grep("amino", RBP_KO_models_aminos$ID)] <- "amino"

RBP_KO_models_aminos <- subset(RBP_KO_models_aminos, RBP_KO_models_aminos$region=="amino")

dim(RBP_KO_models_aminos)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(RBP_KO_models_aminos[2:8], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_aminos[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RBP_KO_models_aminos$ID, gaps_col = 4,cellwidth = 20, cellheight = 6)


```



```{r AGAA}


RBP_KO_models_AGAA <- subset(RBP_KO_models, RBP_KO_models$region=="UTR3")

RBP_KO_models_AGAA <- RBP_KO_models_AGAA[grep("AGAA",RBP_KO_models_AGAA$ID),]

RBP_KO_models_AGAA <- subset(RBP_KO_models_AGAA, rowMeans(RBP_KO_models_AGAA[,2:6])>-6)
dim(RBP_KO_models_AGAA)

#154360

col_fun_blue = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#2980b9","#000000")) 

pheatmap(RBP_KO_models_AGAA[2:6], col_fun_blue(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_AGAA[2:6]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RBP_KO_models_AGAA$ID, gaps_col = 4,cellwidth = 10, cellheight = 5)


```




```{r CTTT}

RBP_KO_models_median_cor <- RBP_KO_models
RBP_KO_models_median_cor[,2:6] <- 2^RBP_KO_models_median_cor[,2:6]
RBP_KO_models_median_cor[is.na(RBP_KO_models_median_cor)]=0
RBP_KO_models_median_cor[,2:6] <- RBP_KO_models_median_cor[,2:6] - mapply(median,RBP_KO_models_median_cor[,2:6])
RBP_KO_models_median_cor[,2:6] <- log2(RBP_KO_models_median_cor[,2:6])
RBP_KO_models_median_cor <- do.call(data.frame,lapply(RBP_KO_models_median_cor,function(x) replace(x, is.infinite(x),NA)))
RBP_KO_models_median_cor[is.na(RBP_KO_models_median_cor)]=min(RBP_KO_models_median_cor[,2:6],na.rm = T)

RBP_KO_models_CTTT <- subset(RBP_KO_models_median_cor, RBP_KO_models_median_cor$region=="UTR3")

RBP_KO_models_CTTT <- RBP_KO_models_CTTT[grep("CTTT",RBP_KO_models_CTTT$ID),]

RBP_KO_models_CTTT <- subset(RBP_KO_models_CTTT, rowMeans(RBP_KO_models_CTTT[,2:6])>-6)
dim(RBP_KO_models_CTTT)

col_fun_red = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#c0392b","#000000")) 

pheatmap(RBP_KO_models_CTTT[2:6], col_fun_red(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_CTTT[2:6]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RBP_KO_models_CTTT$ID,cellwidth = 10, cellheight = 10)

```




```{r ATTTA}


RBP_KO_models_ATTTA <- subset(RBP_KO_models, RBP_KO_models$region=="UTR3")

RBP_KO_models_ATTTA <- RBP_KO_models_ATTTA[grep("ATTTA",RBP_KO_models_ATTTA$ID),]

RBP_KO_models_ATTTA <- subset(RBP_KO_models_ATTTA,
                     RBP_KO_models_ATTTA$CD8_Tn_CN>-2|
                     RBP_KO_models_ATTTA$CD8_Tcm_CN>-2 |  
                     RBP_KO_models_ATTTA$CD8_Tem_CN>-2 |
                     RBP_KO_models_ATTTA$CD8_Teff_CN>-2 |
                     RBP_KO_models_ATTTA$CD4_Tn_CN>-2 |
                     RBP_KO_models_ATTTA$CD4_Tcm_CN>-2 |  
                     RBP_KO_models_ATTTA$CD4_Tem_CN>-2)
dim(RBP_KO_models_ATTTA)

col_fun_purple = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#8e44ad","#000000")) 

pheatmap(RBP_KO_models_ATTTA[2:8], col_fun_purple(seq(-10,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_ATTTA[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RBP_KO_models_ATTTA$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```




```{r CTCAGG}


RBP_KO_models_CTCAGG <- subset(RBP_KO_models, RBP_KO_models$region=="UTR3")

RBP_KO_models_CTCAGG <- RBP_KO_models_CTCAGG[grep("CTCAGG",RBP_KO_models_CTCAGG$ID),]

# RBP_KO_models_CTCAGG <- subset(RBP_KO_models_CTCAGG,
                     # RBP_KO_models_CTCAGG$CD8_Tn_CN>-2|
                     # RBP_KO_models_CTCAGG$CD8_Tcm_CN>-2 |  
                     # RBP_KO_models_CTCAGG$CD8_Tem_CN>-2 |
                     # RBP_KO_models_CTCAGG$CD8_Teff_CN>-2 |
                     # RBP_KO_models_CTCAGG$CD4_Tn_CN>-2 |
                     # RBP_KO_models_CTCAGG$CD4_Tcm_CN>-2 |  
                     # RBP_KO_models_CTCAGG$CD4_Tem_CN>-2)
dim(RBP_KO_models_CTCAGG)

col_fun_orange = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#f39c12","#000000")) 

pheatmap(RBP_KO_models_CTCAGG[2:8], col_fun_orange(seq(-8,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(RBP_KO_models_CTCAGG[2:8]), fontsize_row=6,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =RBP_KO_models_CTCAGG$ID, gaps_col = 4,cellwidth = 20, cellheight = 10)


```


## exporting CN versus features ##

```{r importing CN data}

CN_CD8_Tcm_test <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/CN_CD8_Tcm_test", sep=";")
CN_CD8_Tcm_train <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/CN_CD8_Tcm_train", sep=";")

CN_CD8_Tcm <- rbind.data.frame(CN_CD8_Tcm_test,CN_CD8_Tcm_train)


```



```{r RNA modif prot}

ggplot(CN_CD8_Tcm, aes(x=log10(total_m1A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m5C),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m6A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m7G),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_AtoI),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```


```{r PTM modif CN models}

ggplot(CN_CD8_Tcm, aes(x=log10(Acetylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Ubiquitination_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Malonylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Methylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=(drerio_homolog_perc_id_r1),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```




```{r m6a regions}


ggplot(CN_CD8_Tcm, aes(x=log10(UTR5_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(CDS_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(UTR3_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m6A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

```


```{r region m7G}


ggplot(CN_CD8_Tcm, aes(x=UTR5_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=CDS_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=UTR3_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```


```{r 5UTR}

ggplot(CN_CD8_Tcm, aes(x=log10(UTR5_length_UTR5),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



ggplot(CN_CD8_Tcm, aes(x=GC_percentage_UTR5,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=GCGC_UTR5,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)





```



```{r CDS Prot}

ggplot(CN_CD8_Tcm, aes(x=log10(CTTC_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(AGAA_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(S_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(C_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(D_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=(R_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=GC_percentage_CDS,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r condon usage}

ggplot(CN_CD8_Tcm, aes(x=(AAG_codon_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,30))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(TGC_codon_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,10))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```



```{r 3UTR prot}

ggplot(CN_CD8_Tcm, aes(x=GC_percentage_UTR3,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(TTTT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(ATTTA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(AATAAA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



```




```{r selected motifs}

ggplot(CN_CD8_Tcm, aes(x=log10(TATTTA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(AGAAGA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(CTTTCTT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(CTCAGGT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```








## Translation efficiency estimates 



```{r importing TE models}

TE_rest <- readRDS("/home/ben/Analysis/RF_human/xgb_TE_rest_08-11-2021.RDS")
TE_PMA <- readRDS("/home/ben/Analysis/RF_human/xgb_TE_PMA_08-11-2021.RDS")


TE_rest<- data.frame(varImp(TE_rest)$importance)
TE_PMA<- data.frame(varImp(TE_PMA)$importance)

TE_rest$ID <- rownames(TE_rest)
TE_PMA$ID <- rownames(TE_PMA)


colnames(TE_rest)[1] <- "TE_rest"
colnames(TE_PMA)[1] <- "TE_PMA"

TE_models <- list(TE_rest,TE_PMA)

TE_models <- join_all(TE_models,by = "ID")
TE_models[1:2] <- TE_models[2:1]
colnames(TE_models)[1:2] <- colnames(TE_models)[2:1]
dim(TE_models)

TE_models[2:3] <- log2(TE_models[2:3])

TE_models <- do.call(data.frame,lapply(TE_models,function(x) replace(x, is.infinite(x),NA)))


dim(TE_models)
TE_models$region <- "Other"
TE_models$region[grep("CDS", TE_models$ID)] <- "CDS"
TE_models$region[grep("UTR5", TE_models$ID)] <- "UTR5"
TE_models$region[grep("UTR3", TE_models$ID)] <- "UTR3"
#TE_models$region[grep("codon", TE_models$ID)] <- "codon"

dim(TE_models)

TE_models_NAs <- TE_models

TE_models[is.na(TE_models)]=-5





TE_models_CDS <- subset(TE_models, TE_models$region=="CDS")
TE_models_CDS <- subset(TE_models_CDS,
                     TE_models_CDS$TE_rest>2.63|
                     TE_models_CDS$TE_PMA>2.63 )
dim(TE_models_CDS)

TE_models_UTR5 <- subset(TE_models, TE_models$region=="UTR5")
TE_models_UTR5 <- subset(TE_models_UTR5,
                     TE_models_UTR5$TE_rest>1.34 |
                     TE_models_UTR5$TE_PMA>1.34 )
dim(TE_models_UTR5)

TE_models_UTR3 <- subset(TE_models, TE_models$region=="UTR3")
TE_models_UTR3 <- subset(TE_models_UTR3,
                     TE_models_UTR3$TE_rest>1.01|
                     TE_models_UTR3$TE_PMA>1.01 )
dim(TE_models_UTR3)

TE_models_Other <- subset(TE_models, TE_models$region=="Other")
TE_models_Other <- subset(TE_models_Other,
                     TE_models_Other$TE_rest>1|
                     TE_models_Other$TE_PMA>1)
dim(TE_models_Other)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(TE_models_Other[2:3], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_Other[2:3]), fontsize_row=5,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =TE_models_Other$ID,cellwidth = 20)


pheatmap(TE_models_CDS[2:3], col_fun(seq(-5,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_CDS[2:3]), fontsize_row=4 ,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_CDS$ID,cellwidth = 20)


pheatmap(TE_models_UTR5[2:3], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_UTR5[2:3]), fontsize_row=4,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_UTR5$ID, cellwidth = 20)


pheatmap(TE_models_UTR3[2:3], col_fun(seq(-4,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_UTR3[2:3]), fontsize_row=4,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_UTR3$ID,cellwidth = 20)


```


```{r importance per region}

TE_models_for_imp_plot <- TE_models_NAs
TE_models_for_imp_plot <- melt(TE_models_for_imp_plot)


ggplot(TE_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.7), size = 0.75, stroke=0)+
  scale_color_manual(values = c("#3498db","#C70039"))+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.6, color="red",position = position_dodge(preserve = "total",width = 0.7))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.6, color="blue",position = position_dodge(preserve = "total",width = 0.7))+
  theme_minimal()+
  theme(aspect.ratio = 2)

TE_models_for_imp_plot$miRNA <- "Not miRNA"
TE_models_for_imp_plot$miRNA[grep("hsa", TE_models$ID)] <- "miRNA"


ggplot(TE_models_for_imp_plot,aes(x=miRNA,y=value,color=miRNA))+
  geom_point(position = "jitter")+
  # geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.7), size = 0.75, stroke=0)+
  scale_color_manual(values = c("#C70039","#3498db"))+
  scale_x_discrete(limits = c("Not miRNA","miRNA"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.6, color="red",position = position_dodge(preserve = "total",width = 0.7))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.6, color="blue",position = position_dodge(preserve = "total",width = 0.7))+
  theme_minimal()+
  theme(aspect.ratio = 2)

```




```{r CDS}
TE_models_CDS_dotplot <-TE_models_NAs[TE_models_NAs$region=="CDS",]

TE_models_CDS_dotplot$LFC_rest_PMA <- TE_models_CDS_dotplot$TE_PMA - TE_models_CDS_dotplot$TE_rest

TE_models_CDS_dotplot <- do.call(data.frame,lapply(TE_models_CDS_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_CDS_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[abs(TE_models_CDS_dotplot$LFC_rest_PMA)>1 & (TE_models_CDS_dotplot$TE_rest>2 | TE_models_CDS_dotplot$TE_PMA>2),], aes(label=ID),size=2.5)+
  theme_minimal()+
  theme(aspect.ratio = 1)




TE_models_CDS_dotplot$rank_TE_rest <- rank(-TE_models_CDS_dotplot$TE_rest)

ggplot(TE_models_CDS_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[TE_models_CDS_dotplot$TE_rest>4.3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1500))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




TE_models_CDS_dotplot$rank_TE_PMA <- rank(-TE_models_CDS_dotplot$TE_PMA)

ggplot(TE_models_CDS_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[TE_models_CDS_dotplot$TE_PMA>4.3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1400))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


```




```{r 3UTR}
TE_models_UTR3_dotplot <-TE_models_NAs[TE_models_NAs$region=="UTR3",]

TE_models_UTR3_dotplot$LFC_rest_PMA <- TE_models_UTR3_dotplot$TE_PMA - TE_models_UTR3_dotplot$TE_rest

TE_models_UTR3_dotplot <- do.call(data.frame,lapply(TE_models_UTR3_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_UTR3_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[abs(TE_models_UTR3_dotplot$LFC_rest_PMA)>1 & (TE_models_UTR3_dotplot$TE_rest>1.5 | TE_models_UTR3_dotplot$TE_PMA>1.5),], aes(label=ID),size=2)+
  theme_minimal()+
  theme(aspect.ratio = 1)



TE_models_UTR3_dotplot$rank_TE_rest <- rank(-TE_models_UTR3_dotplot$TE_rest)

ggplot(TE_models_UTR3_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[TE_models_UTR3_dotplot$TE_rest>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1200))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




TE_models_UTR3_dotplot$rank_TE_PMA <- rank(-TE_models_UTR3_dotplot$TE_PMA)

ggplot(TE_models_UTR3_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[TE_models_UTR3_dotplot$TE_PMA>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1200))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




```






```{r 5UTR}
TE_models_UTR5_dotplot <-TE_models_NAs[TE_models_NAs$region=="UTR5",]

TE_models_UTR5_dotplot$LFC_rest_PMA <- TE_models_UTR5_dotplot$TE_PMA - TE_models_UTR5_dotplot$TE_rest

TE_models_UTR5_dotplot <- do.call(data.frame,lapply(TE_models_UTR5_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_UTR5_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[abs(TE_models_UTR5_dotplot$LFC_rest_PMA)>1 & (TE_models_UTR5_dotplot$TE_rest>1.5 | TE_models_UTR5_dotplot$TE_PMA>1.5),], aes(label=ID),size=2)+
  theme_minimal()+
  theme(aspect.ratio = 1)



TE_models_UTR5_dotplot$rank_TE_rest <- rank(-TE_models_UTR5_dotplot$TE_rest)

ggplot(TE_models_UTR5_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[TE_models_UTR5_dotplot$TE_rest>3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,800))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)






TE_models_UTR5_dotplot$rank_TE_PMA <- rank(-TE_models_UTR5_dotplot$TE_PMA)

ggplot(TE_models_UTR5_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[TE_models_UTR5_dotplot$TE_PMA>3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,700))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




```



```{r Other}
TE_models_Other_dotplot <-TE_models_NAs[TE_models_NAs$region=="Other",]

TE_models_Other_dotplot$LFC_rest_PMA <- TE_models_Other_dotplot$TE_PMA - TE_models_Other_dotplot$TE_rest

TE_models_Other_dotplot <- do.call(data.frame,lapply(TE_models_Other_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_Other_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[abs(TE_models_Other_dotplot$LFC_rest_PMA)>1 & (TE_models_Other_dotplot$TE_rest>0 | TE_models_Other_dotplot$TE_PMA>0),], aes(label=ID),size=2.5)+
  theme_minimal()+
  theme(aspect.ratio = 1)





TE_models_Other_dotplot$rank_TE_rest <- rank(-TE_models_Other_dotplot$TE_rest)

ggplot(TE_models_Other_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[TE_models_Other_dotplot$TE_rest>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 10,force = 1)+
  scale_x_continuous(limits = c(0,80))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)






TE_models_Other_dotplot$rank_TE_PMA <- rank(-TE_models_Other_dotplot$TE_PMA)

ggplot(TE_models_Other_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[TE_models_Other_dotplot$TE_PMA>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 10,force = 1)+
  scale_x_continuous(limits = c(0,80))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)

```


