---
title: "Visualizing XGB output in immune subsets"
author: "Benoit Nicolet"
date: "03/05/2022"
output: html_document
---


```{r setup, include=FALSE}

#install.packages("pheatmap")

library(plyr)
library(dplyr)
library(ggplot2)
library(ggfortify)
library(tidyverse)
library(circlize)
library(pheatmap)
library(caret)
library(ggpointdensity)
library(viridis)
library(reshape)
# library(limma)

# BiocManager::install("limma")

knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set("/home/ben/Analysis/RF_human/immune_system/")
setwd("/home/ben/Analysis/RF_human/immune_system/")

```


```{r importing RNA library}

# Sequence_parameters <- read.delim("/home/ben/Analysis/RF_human/Library_V2_Aug2021/RNA_library_v2_RBP_GC_length_codon_AA_m6A_m5C_AtoI_m1A_m7G_CD8miRDB.csv",sep = ";", dec=",")

```




## CN Models ##

```{r importing CN models}

CD8_Tn_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD8_Tn_18_10_2021.RDS")
CD8_Tcm_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD8_Tcm_18_10_2021.RDS")
CD8_Tem_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD8_Tem_18_10_2021.RDS")
CD8_Teff_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD8_Teff_18_10_2021.RDS")

CD4_Tn_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD4_Tn_18_10_2021 (1).RDS")
CD4_Tcm_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD4_Tcm_18_10_2021.RDS")
CD4_Tem_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD4_Tem_18_10_2021.RDS")

CD8_Tn_CN<- data.frame(caret::varImp(CD8_Tn_CN,scale=F)$importance)
CD8_Tcm_CN<- data.frame(caret::varImp(CD8_Tcm_CN,scale=F)$importance)
CD8_Tem_CN<- data.frame(caret::varImp(CD8_Tem_CN,scale=F)$importance)
CD8_Teff_CN<- data.frame(caret::varImp(CD8_Teff_CN,scale=F)$importance)
CD4_Tn_CN<- data.frame(caret::varImp(CD4_Tn_CN,scale=F)$importance)
CD4_Tcm_CN<- data.frame(caret::varImp(CD4_Tcm_CN,scale=F)$importance)
CD4_Tem_CN<- data.frame(caret::varImp(CD4_Tem_CN,scale=F)$importance)


B_naive_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_B_naive_02_05_2022.RDS")
B_mem_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_B_mem_02_05_2022.RDS")
B_plasma_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_B_plasma_02_05_2022.RDS")

Basophils_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_Basophils_02_05_2022.RDS")
Eosinophils_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_Eosinophils_02_05_2022.RDS")
Neutrophils_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_Neutrophils_02_05_2022.RDS")

mDC_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_mDC_02_05_2022.RDS")
pDC_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_pDC_02_05_2022.RDS")


B_naive_CN<- data.frame(caret::varImp(B_naive_CN,scale=F)$importance)
B_mem_CN<- data.frame(caret::varImp(B_mem_CN,scale=F)$importance)
B_plasma_CN<- data.frame(caret::varImp(B_plasma_CN,scale=F)$importance)
Basophils_CN<- data.frame(caret::varImp(Basophils_CN,scale=F)$importance)
Eosinophils_CN<- data.frame(caret::varImp(Eosinophils_CN,scale=F)$importance)
Neutrophils_CN<- data.frame(caret::varImp(Neutrophils_CN,scale=F)$importance)
mDC_CN<- data.frame(caret::varImp(mDC_CN,scale=F)$importance)
pDC_CN<- data.frame(caret::varImp(pDC_CN,scale=F)$importance)

Mono_classical_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_Mono_classical_02_05_2022.RDS")
Mono_intermediate_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_Mono_intermediate_02_05_2022.RDS")
Mono_non_classical_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_Mono_non_classical_02_05_2022.RDS")

mTregs_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_mTregs_02_05_2022.RDS")
nTregs_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_nTregs_02_05_2022.RDS")

NK_bright_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_NK_bright_02_05_2022.RDS")
NK_dim_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_NK_dim_02_05_2022.RDS")



HEK293_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/cell_lines_ref/models/xgb_CN_HEK293_07_06_2022.RDS")
HeLa_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/cell_lines_ref/models/xgb_CN_HeLa_07_06_2022.RDS")
K562_CN <- readRDS("/home/ben/Analysis/RF_human/immune_system/cell_lines_ref/models/xgb_CN_K562_07_06_2022.RDS")




Mono_classical_CN<- data.frame(caret::varImp(Mono_classical_CN,scale=F)$importance)
Mono_intermediate_CN<- data.frame(caret::varImp(Mono_intermediate_CN,scale=F)$importance)
Mono_non_classical_CN<- data.frame(caret::varImp(Mono_non_classical_CN,scale=F)$importance)
mTregs_CN<- data.frame(caret::varImp(mTregs_CN,scale=F)$importance)
nTregs_CN<- data.frame(caret::varImp(nTregs_CN,scale=F)$importance)
NK_bright_CN<- data.frame(caret::varImp(NK_bright_CN,scale=F)$importance)
NK_dim_CN<- data.frame(caret::varImp(NK_dim_CN,scale=F)$importance)

HEK293_CN<- data.frame(caret::varImp(HEK293_CN,scale=F)$importance)
HeLa_CN<- data.frame(caret::varImp(HeLa_CN,scale=F)$importance)
K562_CN<- data.frame(caret::varImp(K562_CN,scale=F)$importance)

CD8_Tn_CN$ID <- rownames(CD8_Tn_CN)
CD8_Tcm_CN$ID <- rownames(CD8_Tcm_CN)
CD8_Tem_CN$ID <- rownames(CD8_Tem_CN)
CD8_Teff_CN$ID <- rownames(CD8_Teff_CN)
CD4_Tn_CN$ID <- rownames(CD4_Tn_CN)
CD4_Tcm_CN$ID <- rownames(CD4_Tcm_CN)
CD4_Tem_CN$ID <- rownames(CD4_Tem_CN)

B_naive_CN$ID <- rownames(B_naive_CN)
B_mem_CN$ID <- rownames(B_mem_CN)
B_plasma_CN$ID <- rownames(B_plasma_CN)
Basophils_CN$ID <- rownames(Basophils_CN)
Eosinophils_CN$ID <- rownames(Eosinophils_CN)
Neutrophils_CN$ID <- rownames(Neutrophils_CN)
mDC_CN$ID <- rownames(mDC_CN)
pDC_CN$ID <- rownames(pDC_CN)
Mono_classical_CN$ID <- rownames(Mono_classical_CN)
Mono_intermediate_CN$ID <- rownames(Mono_intermediate_CN)
Mono_non_classical_CN$ID <- rownames(Mono_non_classical_CN)
mTregs_CN$ID <- rownames(mTregs_CN)
nTregs_CN$ID <- rownames(nTregs_CN)
NK_bright_CN$ID <- rownames(NK_bright_CN)
NK_dim_CN$ID <- rownames(NK_dim_CN)


HEK293_CN$ID <- rownames(HEK293_CN)
HeLa_CN$ID <- rownames(HeLa_CN)
K562_CN$ID <- rownames(K562_CN)


colnames(CD8_Tn_CN)[1] <- "CD8_Tn_CN"
colnames(CD8_Tcm_CN)[1] <- "CD8_Tcm_CN"
colnames(CD8_Tem_CN)[1] <- "CD8_Tem_CN"
colnames(CD8_Teff_CN)[1] <- "CD8_Teff_CN"
colnames(CD4_Tn_CN)[1] <- "CD4_Tn_CN"
colnames(CD4_Tcm_CN)[1] <- "CD4_Tcm_CN"
colnames(CD4_Tem_CN)[1] <- "CD4_Tem_CN"

colnames(B_naive_CN)[1] <- "B_naive_CN"
colnames(B_mem_CN)[1] <- "B_mem_CN"
colnames(B_plasma_CN)[1] <- "B_plasma_CN"
colnames(Basophils_CN)[1] <- "Basophils_CN"
colnames(Eosinophils_CN)[1] <- "Eosinophils_CN"
colnames(Neutrophils_CN)[1] <- "Neutrophils_CN"
colnames(mDC_CN)[1] <- "mDC_CN"
colnames(pDC_CN)[1] <- "pDC_CN"
colnames(Mono_classical_CN)[1] <- "Mono_classical_CN"
colnames(Mono_intermediate_CN)[1] <- "Mono_intermediate_CN"
colnames(Mono_non_classical_CN)[1] <- "Mono_non_classical_CN"
colnames(mTregs_CN)[1] <- "mTregs_CN"
colnames(nTregs_CN)[1] <- "nTregs_CN"
colnames(NK_bright_CN)[1] <- "NK_bright_CN"
colnames(NK_dim_CN)[1] <- "NK_dim_CN"

colnames(HEK293_CN)[1] <- "HEK293_CN"
colnames(HeLa_CN)[1] <- "HeLa_CN"
colnames(K562_CN)[1] <- "K562_CN"


CN_subsets_models <- list(CD8_Tn_CN,CD8_Tcm_CN,CD8_Tem_CN,CD8_Teff_CN,CD4_Tn_CN,CD4_Tcm_CN,CD4_Tem_CN,
                          mTregs_CN,nTregs_CN,
                          NK_bright_CN,NK_dim_CN,
                          B_naive_CN,B_mem_CN,B_plasma_CN,
                          Basophils_CN,Eosinophils_CN,Neutrophils_CN,
                          Mono_classical_CN,Mono_intermediate_CN,Mono_non_classical_CN,
                          mDC_CN,pDC_CN,
                          HEK293_CN,HeLa_CN,K562_CN)

CN_subsets_models <- plyr::join_all(CN_subsets_models,by = "ID")
CN_subsets_models[1:2] <- CN_subsets_models[2:1]
colnames(CN_subsets_models)[1:2] <- colnames(CN_subsets_models)[2:1]
dim(CN_subsets_models)



## median correction ##
# CN_subsets_models <- data.frame(CN_subsets_models[1:1],scale(CN_subsets_models[2:26],center = F))

CN_subsets_models[2:26] <- log10(CN_subsets_models[2:26])
CN_subsets_models <- do.call(data.frame,lapply(CN_subsets_models,function(x) replace(x, is.infinite(x),NA)))
CN_subsets_models[is.na(CN_subsets_models)]=min(CN_subsets_models[2:26],na.rm = T)
CN_subsets_models[2:26] <- CN_subsets_models[2:26] - min(CN_subsets_models[2:26])

# rownames(CN_subsets_models) <- CN_subsets_models$ID
# 
# CN_subsets_models <- as.matrix(CN_subsets_models[2:26])
# 
# CN_subsets_models <- limma::voom(CN_subsets_models,design = NULL)
# CN_subsets_models <- data.frame(CN_subsets_models)
# CN_subsets_models <- data.frame("ID"=rownames(CN_subsets_models),CN_subsets_models)



# CN_subsets_models_median_cor <- CN_subsets_models
# CN_subsets_models_median_cor[,2:26] <- 2^CN_subsets_models_median_cor[,2:26]
# CN_subsets_models_median_cor[is.na(CN_subsets_models_median_cor)]=0
# CN_subsets_models_median_cor[,2:26] <- CN_subsets_models_median_cor[,2:26] - mapply(median,CN_subsets_models_median_cor[,2:26])
# CN_subsets_models_median_cor[,2:26] <- log2(CN_subsets_models_median_cor[,2:26])
# CN_subsets_models_median_cor <- do.call(data.frame,lapply(CN_subsets_models_median_cor,function(x) replace(x, is.infinite(x),NA)))
# CN_subsets_models_median_cor[is.na(CN_subsets_models_median_cor)]=min(CN_subsets_models_median_cor[,2:26],na.rm = T)
# CN_subsets_models <- CN_subsets_models_median_cor

#CN_subsets_models[is.na(CN_subsets_models)]=min(CN_subsets_models[2:26],na.rm = T)



dim(CN_subsets_models)
CN_subsets_models$region <- "Other"
CN_subsets_models$region[grep("CDS", CN_subsets_models$ID)] <- "CDS"
CN_subsets_models$region[grep("UTR5", CN_subsets_models$ID)] <- "UTR5"
CN_subsets_models$region[grep("UTR3", CN_subsets_models$ID)] <- "UTR3"
#CN_subsets_models$region[grep("codon", CN_subsets_models$ID)] <- "codon"

dim(CN_subsets_models)


CN_subsets_models_CDS <- subset(CN_subsets_models, CN_subsets_models$region=="CDS")
dim(CN_subsets_models_CDS)

CN_subsets_models_UTR5 <- subset(CN_subsets_models, CN_subsets_models$region=="UTR5")
dim(CN_subsets_models_UTR5)

CN_subsets_models_UTR3 <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")
dim(CN_subsets_models_UTR3)

CN_subsets_models_Other <- subset(CN_subsets_models, CN_subsets_models$region=="Other")
dim(CN_subsets_models_Other)


```




```{r corplot}

col_fun= colorRamp2(c(-3, -1, 0, 0.5, 2), c("#053061","#579EC9","#FFFFFF","#D96651","#67001F"))


corrplot::corrplot(cor(CN_subsets_models[,2:26],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "full lib",mar = c(0,0,1,0))

corrplot::corrplot(cor(CN_subsets_models_UTR5[,2:26],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "5'UTR",mar = c(0,0,1,0))

corrplot::corrplot(cor(CN_subsets_models_CDS[,2:26],method = "spearman",use = "pairwise.complete.obs"),method = "shade",col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)),diag = F,order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "CDS",mar = c(0,0,1,0))


corrplot::corrplot(cor(CN_subsets_models_UTR3[,2:26],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "3'UTR",mar = c(0,0,1,0))

corrplot::corrplot(cor(CN_subsets_models_Other[,2:26],method = "spearman",use = "pairwise.complete.obs"),method = "shade", col.lim = c(0,1), col=col_fun(seq(-9,3,by=0.01)), diag = F, order = "hclust",hclust.method = "complete",
                   tl.cex = 0.7,addrect = T,title = "other",mar = c(0,0,1,0))

# cor(CN_subsets_models[,2:8],method = "pearson",use = "pairwise.complete.obs")

```



```{r corplot and heatmap on top features}



CN_subsets_models_CDS_top <- CN_subsets_models_CDS
CN_subsets_models_CDS_top[2:26] <- 10^(CN_subsets_models_CDS_top[2:26])
CN_subsets_models_CDS_top$rowmean <- rowMeans(CN_subsets_models_CDS_top[,2:26])
CN_subsets_models_CDS_top <- CN_subsets_models_CDS_top[order(CN_subsets_models_CDS_top$rowmean,decreasing = T),]
CN_subsets_models_CDS_top <- CN_subsets_models_CDS_top[1:100,]
CN_subsets_models_CDS_top[2:26] <- log10(CN_subsets_models_CDS_top[2:26])
CN_subsets_models_CDS_top <- do.call(data.frame,lapply(CN_subsets_models_CDS_top,function(x) replace(x, is.infinite(x),NA)))
dim(CN_subsets_models_CDS_top)
#CN_subsets_models_CDS_top$rowmean <- NULL

CN_subsets_models_UTR5_top <- CN_subsets_models_UTR5
CN_subsets_models_UTR5_top[2:26] <- 10^(CN_subsets_models_UTR5_top[2:26])
CN_subsets_models_UTR5_top$rowmean <- rowMeans(CN_subsets_models_UTR5_top[,2:26])
CN_subsets_models_UTR5_top <- CN_subsets_models_UTR5_top[order(CN_subsets_models_UTR5_top$rowmean,decreasing = T),]
CN_subsets_models_UTR5_top <- CN_subsets_models_UTR5_top[1:100,]
CN_subsets_models_UTR5_top[2:26] <- log10(CN_subsets_models_UTR5_top[2:26])
CN_subsets_models_UTR5_top <- do.call(data.frame,lapply(CN_subsets_models_UTR5_top,function(x) replace(x, is.infinite(x),NA)))
dim(CN_subsets_models_UTR5_top)


CN_subsets_models_UTR3_top <- CN_subsets_models_UTR3
CN_subsets_models_UTR3_top[2:26] <- 10^(CN_subsets_models_UTR3_top[2:26])
CN_subsets_models_UTR3_top$rowmean <- rowMeans(CN_subsets_models_UTR3_top[,2:26])
CN_subsets_models_UTR3_top <- CN_subsets_models_UTR3_top[order(CN_subsets_models_UTR3_top$rowmean,decreasing = T),]
CN_subsets_models_UTR3_top <- CN_subsets_models_UTR3_top[1:100,]
CN_subsets_models_UTR3_top[2:26] <- log10(CN_subsets_models_UTR3_top[2:26])
CN_subsets_models_UTR3_top <- do.call(data.frame,lapply(CN_subsets_models_UTR3_top,function(x) replace(x, is.infinite(x),NA)))
dim(CN_subsets_models_UTR3_top)


CN_subsets_models_Other_top <- CN_subsets_models_Other
CN_subsets_models_Other_top[2:26] <- 10^(CN_subsets_models_Other_top[2:26])
CN_subsets_models_Other_top$rowmean <- rowMeans(CN_subsets_models_Other_top[,2:26])
CN_subsets_models_Other_top <- CN_subsets_models_Other_top[order(CN_subsets_models_Other_top$rowmean,decreasing = T),]
CN_subsets_models_Other_top <- CN_subsets_models_Other_top[1:25,]
CN_subsets_models_Other_top[2:26] <- log10(CN_subsets_models_Other_top[2:26])
CN_subsets_models_Other_top <- do.call(data.frame,lapply(CN_subsets_models_Other_top,function(x) replace(x, is.infinite(x),NA)))
dim(CN_subsets_models_Other_top)


## Heatmap time ##
col_fun = colorRamp2(c(6.5,7.5,9.5), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(CN_subsets_models_Other_top[2:26], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_Other_top[2:26]), fontsize_row=4,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =CN_subsets_models_Other$ID, cellwidth = 8, main = "general",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)


pheatmap(CN_subsets_models_CDS_top[2:26], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_CDS_top[2:26]), fontsize_row=3 ,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =CN_subsets_models_CDS$ID,cellwidth = 8,main = "CDS",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)


pheatmap(CN_subsets_models_UTR5_top[2:26], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_UTR5_top[2:26]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =CN_subsets_models_UTR5$ID,cellwidth = 8,main = "5'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)


pheatmap(CN_subsets_models_UTR3_top[2:26], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_UTR3_top[2:26]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="euclidean",fontsize_col=5,
         labels_row =CN_subsets_models_UTR3$ID, cellwidth = 8,main = "3'UTR",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)



```


```{r ranking var_imp}

dat <- list()
dat_list <- list()
for (column in 2:26){
  x <- CN_subsets_models[column]
  x <- x[order(x,decreasing = T),]
  x <- 10^x
  dat$cumSum <- cumsum(x)
  dat$rank <- 1:length(x)
  dat$group <- colnames(CN_subsets_models[column])
  dat_list[[column]] <- do.call(cbind,dat)
  
}

ranking <- data.frame(do.call(rbind,dat_list))
head(ranking[1:10,])
# 
# 
# CN_subsets_models_ranking <- CN_subsets_models_ranking[order(CN_subsets_models_ranking$CD8_Tn_CN,decreasing = T),]
# CN_subsets_models_ranking$CD8_Tn_CN <- 10^CN_subsets_models_ranking$CD8_Tn_CN
# CN_subsets_models_ranking$cumSum_CD8_Tn_CN <- cumsum(CN_subsets_models_ranking$CD8_Tn_CN)
# CN_subsets_models_ranking$rank_CD8_Tn_CN <- 1:nrow(CN_subsets_models_ranking)
# CN_subsets_models_ranking$CD8_Tn_CN <- log10(CN_subsets_models_ranking$CD8_Tn_CN)
# 
# CN_subsets_models_ranking <- CN_subsets_models_ranking[order(CN_subsets_models_ranking$CD8_Tcm_CN,decreasing = T),]
# CN_subsets_models_ranking$CD8_Tcm_CN <- 10^CN_subsets_models_ranking$CD8_Tcm_CN
# CN_subsets_models_ranking$cumSum_CD8_Tcm_CN <- cumsum(CN_subsets_models_ranking$CD8_Tcm_CN)
# CN_subsets_models_ranking$rank_CD8_Tcm_CN <- 1:nrow(CN_subsets_models_ranking)
# CN_subsets_models_ranking$CD8_Tcm_CN <- log10(CN_subsets_models_ranking$CD8_Tcm_CN)
# 
# CN_subsets_models_ranking <- CN_subsets_models_ranking[order(CN_subsets_models_ranking$HEK293_CN,decreasing = T),]
# CN_subsets_models_ranking$CD8_Tn_CN <- 10^CN_subsets_models_ranking$CD8_Tn_CN
# CN_subsets_models_ranking$cumSum_CD8_Tn_CN <- cumsum(CN_subsets_models_ranking$CD8_Tn_CN)
# CN_subsets_models_ranking$rank_CD8_Tn_CN <- 1:nrow(CN_subsets_models_ranking)
# CN_subsets_models_ranking$CD8_Tn_CN <- log10(CN_subsets_models_ranking$CD8_Tn_CN)
# 
# 
# CN_subsets_models_ranking <- CN_subsets_models_ranking[order(CN_subsets_models_ranking$HEK293_CN,decreasing = T),]
# CN_subsets_models_ranking$HEK293_CN <- 10^CN_subsets_models_ranking$HEK293_CN
# CN_subsets_models_ranking$cumSum_HEK293_CN<- cumsum(CN_subsets_models_ranking$HEK293_CN)
# CN_subsets_models_ranking$rank_HEK293_CN <- 1:nrow(CN_subsets_models_ranking)
# CN_subsets_models_ranking$HEK293_CN <- log10(CN_subsets_models_ranking$HEK293_CN)
# 
# 
# CN_subsets_models_ranking[2:27] <- NULL
# 
# CN_subsets_models_ranking <- melt(CN_subsets_models_ranking,measure.vars = c(""))
ranking$cumSum <- as.numeric(ranking$cumSum)
ranking$rank <- as.numeric(ranking$rank)
ranking$color <- "Primary cells"
ranking$color[ranking$group== "HEK293_CN" | ranking$group== "HeLa_CN" | ranking$group== "K562_CN"]= "Cell lines"

table(ranking$group)
table(ranking$color)

ggplot(ranking, aes(x=rank,y=cumSum, group=group,color=color))+
  geom_point(size=0.25)+
  scale_y_continuous(limits = c(0,7e+10),breaks = c(0,3.5e+10,7e+10))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(ranking, aes(x=rank,y=cumSum, group=group,color=color))+
  geom_point(size=0.5)+
  scale_x_continuous(limits = c(0,2000))+
  scale_y_continuous(limits = c(3.5e+10,7e+10),breaks = c(3.5e+10,7e+10))+
  theme_minimal()+
  theme(aspect.ratio = 1)

# What is the diversity of gene expression

```





```{r importance per region}

CN_subsets_models_for_imp_plot <- CN_subsets_models
# CN_subsets_models_for_imp_plot[CN_subsets_models_for_imp_plot==min(CN_subsets_models_for_imp_plot[,2:26])]=NA
# CN_subsets_models_for_imp_plot[2:26] <- CN_subsets_models_for_imp_plot[2:26] - min(CN_subsets_models_for_imp_plot[2:26])
CN_subsets_models_for_imp_plot <- melt(CN_subsets_models_for_imp_plot)


ggplot(CN_subsets_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.4, dodge.width = 0.8), size = 0.5, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =1, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 1)



## average feature imp ##
CN_subsets_models_avg_imp <- data.frame("ID"=CN_subsets_models$ID,"rowmean"=rowMeans(CN_subsets_models[,2:26]),"region"=CN_subsets_models$region)
CN_subsets_models_avg_imp <- melt(CN_subsets_models_avg_imp)


ggplot(CN_subsets_models_avg_imp,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = region), position = position_jitterdodge(jitter.width = 0.7, dodge.width = 0.8), size = 0.75, stroke=0) +
  #scale_color_manual(values = c("#a9cce3","#5499c7","#2471a3","#1a5276","#f5cba7","#e59866","#d35400"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.7, color="red",position = position_dodge(preserve = "total",width = 0.8))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.7, color="blue",position = position_dodge(preserve = "total",width = 0.8))+
  theme_minimal()+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 3)



```



```{r capturing the variance of regions}

CN_subsets_models_var <- data.frame(CN_subsets_models)
CN_subsets_models_var[2:26] <- 10^(CN_subsets_models_var[2:26])
# CN_subsets_models_var[CN_subsets_models_var==min(CN_subsets_models_var[,2:26],na.rm = T)]=NA

CN_subsets_models_var[is.na(CN_subsets_models_var)]=0
CN_subsets_models_var$row_mean <- rowMeans(as.matrix(CN_subsets_models_var[2:26]))

CN_subsets_models_var$row_sd <- matrixStats::rowSds(as.matrix(CN_subsets_models_var[2:26]))
CN_subsets_models_var$corrected_sd <- CN_subsets_models_var$row_sd/CN_subsets_models_var$row_mean
CN_subsets_models_var$row_mean <- log10(CN_subsets_models_var$row_mean)
CN_subsets_models_var[2:26] <- log10(CN_subsets_models_var[2:26])

plot(CN_subsets_models_var$row_sd,CN_subsets_models_var$row_mean)
plot(CN_subsets_models_var$corrected_sd,CN_subsets_models_var$row_mean)

ggplot(CN_subsets_models_var,aes(x=corrected_sd,y=row_mean))+
  geom_point()+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_subsets_models_var,aes(x=corrected_sd, fill=NULL, color=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(CN_subsets_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  theme(aspect.ratio = 0.5)

ggplot(CN_subsets_models_var,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

CN_subsets_models_var$region <- factor(CN_subsets_models_var$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(CN_subsets_models_var,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(CN_subsets_models_var$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)



mean(CN_subsets_models_var[CN_subsets_models_var$region=="UTR5",]$corrected_sd,na.rm = T) # = -2.38509
mean(CN_subsets_models_var[CN_subsets_models_var$region=="CDS",]$corrected_sd,na.rm = T) # = -2.031128
mean(CN_subsets_models_var[CN_subsets_models_var$region=="UTR3",]$corrected_sd,na.rm = T) # = -2.061535
mean(CN_subsets_models_var[CN_subsets_models_var$region=="Other",]$corrected_sd,na.rm = T) # = -2.781053

median(CN_subsets_models_var$corrected_sd,na.rm = T)

table(CN_subsets_models_var[CN_subsets_models_var$corrected_sd>1.5,]$region)/table(CN_subsets_models_var$region)*100



```
```{r core vs variable features}

hist(CN_subsets_models_var$corrected_sd)
median(CN_subsets_models_var$corrected_sd,na.rm = T)
mean(CN_subsets_models_var$corrected_sd,na.rm = T)

core_features_CN <- data.frame(table(CN_subsets_models_var[CN_subsets_models_var$corrected_sd<mean(CN_subsets_models_var$corrected_sd,na.rm = T),]$region)/table(CN_subsets_models_var$region)*100)
core_features_CN$group <- "B_core"

core_features_CN <- rbind(core_features_CN, data.frame("Var1"=c("UTR5","CDS","UTR3","Other"),"Freq"= 100-core_features_CN$Freq,"group"="A_variable"))


ggplot(core_features_CN,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  # scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)
  


```




```{r full variablity assessment}

CN_subsets_models_var$row_min <- do.call(pmin, as.data.frame(CN_subsets_models_var[2:26]))
table(CN_subsets_models_var$row_min==0)

dim(CN_subsets_models_var[(CN_subsets_models_var$corrected_sd < median(CN_subsets_models_var$corrected_sd,na.rm = T)) & (CN_subsets_models_var$row_min>0),])

dim(CN_subsets_models_var[(CN_subsets_models_var$row_min>0),])

core_features_CN <- data.frame(
  table(CN_subsets_models_var[(CN_subsets_models_var$corrected_sd < median(CN_subsets_models_var$corrected_sd,na.rm = T)) & (CN_subsets_models_var$row_min>0),]$region)/table(CN_subsets_models_var$region)*100)
core_features_CN$group <- "C_core"

HV_core_features_CN <- data.frame(
  table(CN_subsets_models_var[(CN_subsets_models_var$corrected_sd < median(CN_subsets_models_var$corrected_sd,na.rm = T)) & (CN_subsets_models_var$row_min==0),]$region)/table(CN_subsets_models_var$region)*100)
HV_core_features_CN$group <- "A_hyper_variable"

core_features_CN <- rbind(core_features_CN, HV_core_features_CN, data.frame("Var1"=c("UTR5","CDS","UTR3","Other"),"Freq"= 100-(core_features_CN$Freq + HV_core_features_CN$Freq),"group"="B_variable"))


ggplot(core_features_CN,aes(x=Var1, y=Freq, fill=group))+
  geom_bar(stat="identity")+
  # scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  scale_y_continuous(expand = c(0,0))+
  theme(aspect.ratio = 2)
  


```


```{r capturing the variance of regions}
non_zero_CN_subsets_models <- data.frame(non_zero_CN_subsets_models)
# non_zero_CN_subsets_models[2:26] <- 10^(non_zero_CN_subsets_models[2:26])
# # non_zero_CN_subsets_models[non_zero_CN_subsets_models==min(non_zero_CN_subsets_models[,2:26],na.rm = T)]=NA
# 
# non_zero_CN_subsets_models[is.na(non_zero_CN_subsets_models)]=min(non_zero_CN_subsets_models[,2:26],na.rm = T)
# non_zero_CN_subsets_models$row_mean <- rowMeans(as.matrix(non_zero_CN_subsets_models[2:26]))
# 
# non_zero_CN_subsets_models$row_sd <- matrixStats::rowSds(as.matrix(non_zero_CN_subsets_models[2:26]))
# non_zero_CN_subsets_models$corrected_sd <- non_zero_CN_subsets_models$row_sd/non_zero_CN_subsets_models$row_mean
# non_zero_CN_subsets_models$row_mean <- log10(non_zero_CN_subsets_models$row_mean)
# non_zero_CN_subsets_models[2:26] <- log10(non_zero_CN_subsets_models[2:26])

plot(non_zero_CN_subsets_models$row_sd,non_zero_CN_subsets_models$row_mean)
plot(non_zero_CN_subsets_models$corrected_sd,non_zero_CN_subsets_models$row_mean)

ggplot(non_zero_CN_subsets_models,aes(x=corrected_sd,y=row_mean))+
  geom_point()+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(non_zero_CN_subsets_models,aes(x=corrected_sd, fill=NULL, color=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


ggplot(non_zero_CN_subsets_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  #scale_x_continuous(limits = c(-7,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  theme(aspect.ratio = 0.5)

ggplot(non_zero_CN_subsets_models,aes(x=region, y=corrected_sd, fill=region))+
  geom_violin(alpha=0.6)+
  # scale_y_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = -2.5, type="dotted")+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width = 0.7, color="black")+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  theme(aspect.ratio = 2)

non_zero_CN_subsets_models$region <- factor(non_zero_CN_subsets_models$region, levels=c("UTR5","CDS","UTR3","Other"))

ggplot(non_zero_CN_subsets_models,aes(x=corrected_sd, fill=region))+
  geom_density(alpha=0.6)+
  # scale_x_continuous(limits = c(0,3))+
  theme_minimal()+
  geom_vline(xintercept = median(non_zero_CN_subsets_models$corrected_sd,na.rm = T), linetype="dotted")+
  facet_wrap(.~region,ncol = 1)+
  theme(aspect.ratio = 0.5)






```





```{r }
CN_subsets_models_codons <- CN_subsets_models
CN_subsets_models_codons$region[grep("codon", CN_subsets_models_codons$ID)] <- "codon"

CN_subsets_models_codons <- subset(CN_subsets_models_codons, CN_subsets_models_codons$region=="codon")

dim(CN_subsets_models_codons)


#col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green
col_fun = colorRamp2(c(6.5,7.5,9.5), c("#FFFFFF","#1E8449","#000000")) 

CN_subsets_models_codons <- CN_subsets_models_codons[order(CN_subsets_models_codons$ID),]


pheatmap(CN_subsets_models_codons[2:26], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_codons[2:26]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =CN_subsets_models_codons$ID, cellwidth = 8,main = "codons",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)

```



```{r }
CN_subsets_models_aminos <- CN_subsets_models
CN_subsets_models_aminos$region[grep("amino", CN_subsets_models_aminos$ID)] <- "amino"

CN_subsets_models_aminos <- subset(CN_subsets_models_aminos, CN_subsets_models_aminos$region=="amino")

dim(CN_subsets_models_aminos)


col_fun = colorRamp2(c(6.5,7.5,9.5), c("#FFFFFF","#1E8449","#000000")) 

pheatmap(CN_subsets_models_aminos[2:26], col_fun(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_aminos[2:26]), fontsize_row=3,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="manhattan",fontsize_col=5,
         labels_row =CN_subsets_models_aminos$ID, cellwidth = 8,main = "amino acids",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)



```



```{r AGAA}


CN_subsets_models_AGAA <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")

CN_subsets_models_AGAA <- CN_subsets_models_AGAA[grep("AGAA",CN_subsets_models_AGAA$ID),]

CN_subsets_models_AGAA <- subset(CN_subsets_models_AGAA, rowMeans(CN_subsets_models_AGAA[,2:26])>6)
dim(CN_subsets_models_AGAA)

#154360

col_fun_blue = colorRamp2(c(6,8,10), c("#FFFFFF","#2980b9","#000000")) 

pheatmap(CN_subsets_models_AGAA[2:26], col_fun_blue(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_AGAA[2:26]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =CN_subsets_models_AGAA$ID, cellwidth = 8,main = "AGAA",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)



```




```{r CTTT}

CN_subsets_models_CTTT <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")
CN_subsets_models_CTTT <- CN_subsets_models_CTTT[grep("CTTT",CN_subsets_models_CTTT$ID),]

CN_subsets_models_CTTT <- subset(CN_subsets_models_CTTT, rowMeans(CN_subsets_models_CTTT[,2:26])>4)
dim(CN_subsets_models_CTTT)


col_fun_red = colorRamp2(c(6,8,10), c("#FFFFFF","#c0392b","#000000")) 

pheatmap(CN_subsets_models_CTTT[2:26], col_fun_red(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_CTTT[2:26]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =CN_subsets_models_CTTT$ID, cellwidth = 8,main = "CTTT",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10,gaps_col = 22)

```




```{r ATTTA}


CN_subsets_models_ATTTA <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")
CN_subsets_models_ATTTA <- CN_subsets_models_ATTTA[grep("ATTTA",CN_subsets_models_ATTTA$ID),]

CN_subsets_models_ATTTA <- subset(CN_subsets_models_ATTTA, rowMeans(CN_subsets_models_ATTTA[,2:26])>4)
dim(CN_subsets_models_ATTTA)


col_fun_red = colorRamp2(c(6,8,10), c("#FFFFFF","#8e44ad","#000000")) 

pheatmap(CN_subsets_models_ATTTA[2:26], col_fun_red(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_ATTTA[2:26]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =CN_subsets_models_ATTTA$ID, cellwidth = 8,main = "ATTTA",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)


```




```{r CTCAGG}


CN_subsets_models_CTCAGG <- subset(CN_subsets_models, CN_subsets_models$region=="UTR3")
CN_subsets_models_CTCAGG <- CN_subsets_models_CTCAGG[grep("CTCAGG",CN_subsets_models_CTCAGG$ID),]

CN_subsets_models_CTCAGG <- subset(CN_subsets_models_CTCAGG, rowMeans(CN_subsets_models_CTCAGG[,2:26])>0)
dim(CN_subsets_models_CTCAGG)


col_fun_red = colorRamp2(c(6,8,10),  c("#FFFFFF","#f39c12","#000000")) 

pheatmap(CN_subsets_models_CTCAGG[2:26], col_fun_red(seq(3,10,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(CN_subsets_models_CTCAGG[2:26]), fontsize_row=5,na_col = "#515a5a", 
         border_color = NA, clustering_distance_rows ="canberra",fontsize_col=5,
         labels_row =CN_subsets_models_CTCAGG$ID, cellheight = 8, cellwidth = 8,main = "CTCAGG",
         breaks = seq(3,10,by=0.01),treeheight_row = 10 ,treeheight_col = 10, gaps_col = 22)


```

















```{r trying to explain models better with SHAP}

# https://www.r-bloggers.com/2021/01/treeshap%E2%80%8A-%E2%80%8Aexplain-tree-based-models-with-shap-values/

# install.packages("devtools")
# devtools::install_github('ModelOriented/treeshap')
library(treeshap)

CD8Tn_CN_model <- readRDS("/home/ben/Analysis/RF_human/immune_system/Protein/models/xgb_CN_CD8_Tn_18_10_2021.RDS")
CD8Tn_CN_train <- read.delim("/home/ben/Analysis/RF_human/immune_system/Protein/split_datasets/CN_CD8_Tn_train",sep = ";")
# CD8Tn_CN_train_minus_CN <- CD8Tn_CN_train[2:dim(CD8Tn_CN_train)[2]]
# CD8Tn_CN_model$modelInfo

CD8Tn_CN_model_unified <- xgboost.unify(CD8Tn_CN_model$finalModel, CD8Tn_CN_train, recalculate = FALSE)
# 
# CD8Tn_CN_model_unified$model <- CD8Tn_CN_model_unified$model[(CD8Tn_CN_model_unified$model$Feature)%in%(colnames(CD8Tn_CN_train[2:dim(CD8Tn_CN_train)[2]]))==TRUE,]
# 
# CD8Tn_CN_model_unified$model <- CD8Tn_CN_model_unified$model[!is.na(CD8Tn_CN_model_unified$model$Feature)==TRUE,]

CD8Tn_CN_model_treeshap_res <- treeshap(CD8Tn_CN_model_unified,CD8Tn_CN_train[1:1000,])
# CD8Tn_CN_model_treeshap_res$shaps

plot_feature_dependence(CD8Tn_CN_model_treeshap_res, 'CDS_m6a')+
  geom_smooth()
CD8Tn_CN_model_treeshap_res$shaps$UTR3_m7G



# CD8Tn_CN_model_treeshap_res <- treeshap(CD8Tn_CN_model_unified,CD8Tn_CN_train_minus_CN[1:100,],interactions = T)
# plot_interaction(treeshap_interactions, 'total_m7G', 'UTR3_length_UTR3')


```


























## exporting CN versus features ##

```{r importing CN data}

CN_CD8_Tcm_test <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/CN_CD8_Tcm_test", sep=";")
CN_CD8_Tcm_train <- read.delim("~/Analysis/RF_human/T_cell_subsets/protein_model/CN_CD8_Tcm_train", sep=";")

CN_CD8_Tcm <- rbind.data.frame(CN_CD8_Tcm_test,CN_CD8_Tcm_train)


```



```{r RNA modif prot}

ggplot(CN_CD8_Tcm, aes(x=log10(total_m1A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m5C),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m6A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m7G),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_AtoI),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```


```{r PTM modif CN models}

ggplot(CN_CD8_Tcm, aes(x=log10(Acetylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Ubiquitination_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Malonylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(Methylation_count),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=(drerio_homolog_perc_id_r1),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)



```




```{r m6a regions}


ggplot(CN_CD8_Tcm, aes(x=log10(UTR5_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(CDS_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(UTR3_m6a),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(total_m6A),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  theme_minimal()+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme(aspect.ratio = 1)

```


```{r region m7G}


ggplot(CN_CD8_Tcm, aes(x=UTR5_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=CDS_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=UTR3_m7G,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,80))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```


```{r 5UTR}

ggplot(CN_CD8_Tcm, aes(x=log10(UTR5_length_UTR5),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



ggplot(CN_CD8_Tcm, aes(x=GC_percentage_UTR5,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=GCGC_UTR5,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)





```



```{r CDS Prot}

ggplot(CN_CD8_Tcm, aes(x=log10(CTTC_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(AGAA_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(S_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(C_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(D_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,20))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=(R_amino_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,40))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=GC_percentage_CDS,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(1,100))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


```

```{r condon usage}

ggplot(CN_CD8_Tcm, aes(x=(AAG_codon_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,30))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=(TGC_codon_CDS),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,10))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```



```{r 3UTR prot}

ggplot(CN_CD8_Tcm, aes(x=GC_percentage_UTR3,y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  #scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(TTTT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(ATTTA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(AATAAA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)



```




```{r selected motifs}

ggplot(CN_CD8_Tcm, aes(x=log10(TATTTA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(AGAAGA_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

ggplot(CN_CD8_Tcm, aes(x=log10(CTTTCTT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)


ggplot(CN_CD8_Tcm, aes(x=log10(CTCAGGT_UTR3),y=CN))+
  geom_pointdensity(stroke=0, size=1.5, show.legend = F)+
  scale_color_viridis(option = "A",begin = 0, end = 0.9)+
  geom_smooth(method = "gam")+
  geom_smooth(method = "lm", color="red")+
  scale_x_continuous(limits = c(0,4))+
  scale_y_continuous(limits = c(0,8))+
  theme_minimal()+
  theme(aspect.ratio = 1)

```








## Translation efficiency estimates 



```{r importing TE models}

TE_rest <- readRDS("/home/ben/Analysis/RF_human/xgb_TE_rest_08-11-2021.RDS")
TE_PMA <- readRDS("/home/ben/Analysis/RF_human/xgb_TE_PMA_08-11-2021.RDS")


TE_rest<- data.frame(varImp(TE_rest,scale=F)$importance)
TE_PMA<- data.frame(varImp(TE_PMA,scale=F)$importance)

TE_rest$ID <- rownames(TE_rest)
TE_PMA$ID <- rownames(TE_PMA)


colnames(TE_rest)[1] <- "TE_rest"
colnames(TE_PMA)[1] <- "TE_PMA"

TE_models <- list(TE_rest,TE_PMA)

TE_models <- join_all(TE_models,by = "ID")
TE_models[1:2] <- TE_models[2:1]
colnames(TE_models)[1:2] <- colnames(TE_models)[2:1]
dim(TE_models)

TE_models[2:3] <- log2(TE_models[2:3])

TE_models <- do.call(data.frame,lapply(TE_models,function(x) replace(x, is.infinite(x),NA)))


dim(TE_models)
TE_models$region <- "Other"
TE_models$region[grep("CDS", TE_models$ID)] <- "CDS"
TE_models$region[grep("UTR5", TE_models$ID)] <- "UTR5"
TE_models$region[grep("UTR3", TE_models$ID)] <- "UTR3"
#TE_models$region[grep("codon", TE_models$ID)] <- "codon"

dim(TE_models)

TE_models_NAs <- TE_models

TE_models[is.na(TE_models)]=-5





TE_models_CDS <- subset(TE_models, TE_models$region=="CDS")
TE_models_CDS <- subset(TE_models_CDS,
                     TE_models_CDS$TE_rest>2.63|
                     TE_models_CDS$TE_PMA>2.63 )
dim(TE_models_CDS)

TE_models_UTR5 <- subset(TE_models, TE_models$region=="UTR5")
TE_models_UTR5 <- subset(TE_models_UTR5,
                     TE_models_UTR5$TE_rest>1.34 |
                     TE_models_UTR5$TE_PMA>1.34 )
dim(TE_models_UTR5)

TE_models_UTR3 <- subset(TE_models, TE_models$region=="UTR3")
TE_models_UTR3 <- subset(TE_models_UTR3,
                     TE_models_UTR3$TE_rest>1.01|
                     TE_models_UTR3$TE_PMA>1.01 )
dim(TE_models_UTR3)

TE_models_Other <- subset(TE_models, TE_models$region=="Other")
TE_models_Other <- subset(TE_models_Other,
                     TE_models_Other$TE_rest>1|
                     TE_models_Other$TE_PMA>1)
dim(TE_models_Other)


col_fun = colorRamp2(c(-3, 0, 3), c("#FFFFFF","#1E8449","#000000")) # green

pheatmap(TE_models_Other[2:3], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_Other[2:3]), fontsize_row=5,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="canberra",
         labels_row =TE_models_Other$ID,cellwidth = 20)


pheatmap(TE_models_CDS[2:3], col_fun(seq(-5,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_CDS[2:3]), fontsize_row=4 ,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_CDS$ID,cellwidth = 20)


pheatmap(TE_models_UTR5[2:3], col_fun(seq(-3,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "canberra", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_UTR5[2:3]), fontsize_row=4,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_UTR5$ID, cellwidth = 20)


pheatmap(TE_models_UTR3[2:3], col_fun(seq(-4,3,by=0.01)) ,scale = "none",
         clustering_distance_cols = "correlation", cluster_cols = F, cluster_rows = T,
         labels_col=colnames(TE_models_UTR3[2:3]), fontsize_row=4,na_col = "#515a5a", 
         border_color = "black", clustering_distance_rows ="manhattan",
         labels_row =TE_models_UTR3$ID,cellwidth = 20)


```


```{r importance per region}

TE_models_for_imp_plot <- TE_models_NAs
TE_models_for_imp_plot <- melt(TE_models_for_imp_plot)


ggplot(TE_models_for_imp_plot,aes(x=region,y=value,group=variable))+
  geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.7), size = 0.75, stroke=0)+
  scale_color_manual(values = c("#3498db","#C70039"))+
  scale_x_discrete(limits = c("UTR5","CDS","UTR3","Other"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.6, color="red",position = position_dodge(preserve = "total",width = 0.7))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.6, color="blue",position = position_dodge(preserve = "total",width = 0.7))+
  theme_minimal()+
  theme(aspect.ratio = 2)

TE_models_for_imp_plot$miRNA <- "Not miRNA"
TE_models_for_imp_plot$miRNA[grep("hsa", TE_models$ID)] <- "miRNA"


ggplot(TE_models_for_imp_plot,aes(x=miRNA,y=value,color=miRNA))+
  geom_point(position = "jitter")+
  # geom_jitter(aes(color = variable), position = position_jitterdodge(jitter.width = 0.5, dodge.width = 0.7), size = 0.75, stroke=0)+
  scale_color_manual(values = c("#C70039","#3498db"))+
  scale_x_discrete(limits = c("Not miRNA","miRNA"))+
  stat_summary(fun = median, fun.min = median, fun.max = median,
               geom = "crossbar", width =0.6, color="red",position = position_dodge(preserve = "total",width = 0.7))+
  stat_summary(fun = mean, fun.min = mean, fun.max = mean,
               geom = "crossbar", width =0.6, color="blue",position = position_dodge(preserve = "total",width = 0.7))+
  theme_minimal()+
  theme(aspect.ratio = 2)

```




```{r CDS}
TE_models_CDS_dotplot <-TE_models_NAs[TE_models_NAs$region=="CDS",]

TE_models_CDS_dotplot$LFC_rest_PMA <- TE_models_CDS_dotplot$TE_PMA - TE_models_CDS_dotplot$TE_rest

TE_models_CDS_dotplot <- do.call(data.frame,lapply(TE_models_CDS_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_CDS_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[abs(TE_models_CDS_dotplot$LFC_rest_PMA)>1 & (TE_models_CDS_dotplot$TE_rest>2 | TE_models_CDS_dotplot$TE_PMA>2),], aes(label=ID),size=2.5)+
  theme_minimal()+
  theme(aspect.ratio = 1)




TE_models_CDS_dotplot$rank_TE_rest <- rank(-TE_models_CDS_dotplot$TE_rest)

ggplot(TE_models_CDS_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[TE_models_CDS_dotplot$TE_rest>4.3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1500))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




TE_models_CDS_dotplot$rank_TE_PMA <- rank(-TE_models_CDS_dotplot$TE_PMA)

ggplot(TE_models_CDS_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_CDS_dotplot[TE_models_CDS_dotplot$TE_PMA>4.3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1400))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)


```




```{r 3UTR}
TE_models_UTR3_dotplot <-TE_models_NAs[TE_models_NAs$region=="UTR3",]

TE_models_UTR3_dotplot$LFC_rest_PMA <- TE_models_UTR3_dotplot$TE_PMA - TE_models_UTR3_dotplot$TE_rest

TE_models_UTR3_dotplot <- do.call(data.frame,lapply(TE_models_UTR3_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_UTR3_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[abs(TE_models_UTR3_dotplot$LFC_rest_PMA)>1 & (TE_models_UTR3_dotplot$TE_rest>1.5 | TE_models_UTR3_dotplot$TE_PMA>1.5),], aes(label=ID),size=2)+
  theme_minimal()+
  theme(aspect.ratio = 1)



TE_models_UTR3_dotplot$rank_TE_rest <- rank(-TE_models_UTR3_dotplot$TE_rest)

ggplot(TE_models_UTR3_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[TE_models_UTR3_dotplot$TE_rest>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1200))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




TE_models_UTR3_dotplot$rank_TE_PMA <- rank(-TE_models_UTR3_dotplot$TE_PMA)

ggplot(TE_models_UTR3_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR3_dotplot[TE_models_UTR3_dotplot$TE_PMA>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,1200))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




```






```{r 5UTR}
TE_models_UTR5_dotplot <-TE_models_NAs[TE_models_NAs$region=="UTR5",]

TE_models_UTR5_dotplot$LFC_rest_PMA <- TE_models_UTR5_dotplot$TE_PMA - TE_models_UTR5_dotplot$TE_rest

TE_models_UTR5_dotplot <- do.call(data.frame,lapply(TE_models_UTR5_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_UTR5_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[abs(TE_models_UTR5_dotplot$LFC_rest_PMA)>1 & (TE_models_UTR5_dotplot$TE_rest>1.5 | TE_models_UTR5_dotplot$TE_PMA>1.5),], aes(label=ID),size=2)+
  theme_minimal()+
  theme(aspect.ratio = 1)



TE_models_UTR5_dotplot$rank_TE_rest <- rank(-TE_models_UTR5_dotplot$TE_rest)

ggplot(TE_models_UTR5_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[TE_models_UTR5_dotplot$TE_rest>3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,800))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)






TE_models_UTR5_dotplot$rank_TE_PMA <- rank(-TE_models_UTR5_dotplot$TE_PMA)

ggplot(TE_models_UTR5_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_UTR5_dotplot[TE_models_UTR5_dotplot$TE_PMA>3,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 100,force = 1)+
  scale_x_continuous(limits = c(0,700))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)




```



```{r Other}
TE_models_Other_dotplot <-TE_models_NAs[TE_models_NAs$region=="Other",]

TE_models_Other_dotplot$LFC_rest_PMA <- TE_models_Other_dotplot$TE_PMA - TE_models_Other_dotplot$TE_rest

TE_models_Other_dotplot <- do.call(data.frame,lapply(TE_models_Other_dotplot,function(x) replace(x, is.infinite(x),NA)))

ggplot(TE_models_Other_dotplot,aes(x=TE_rest,y=TE_PMA,color=abs(LFC_rest_PMA)))+
  geom_point()+
  scale_color_viridis(option = "G",direction = -1,end = 0.85)+
  geom_abline(slope = 1,intercept = c(0))+
  geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[abs(TE_models_Other_dotplot$LFC_rest_PMA)>1 & (TE_models_Other_dotplot$TE_rest>0 | TE_models_Other_dotplot$TE_PMA>0),], aes(label=ID),size=2.5)+
  theme_minimal()+
  theme(aspect.ratio = 1)





TE_models_Other_dotplot$rank_TE_rest <- rank(-TE_models_Other_dotplot$TE_rest)

ggplot(TE_models_Other_dotplot,aes(y=TE_rest,x=rank_TE_rest,color=TE_rest))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[TE_models_Other_dotplot$TE_rest>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 10,force = 1)+
  scale_x_continuous(limits = c(0,80))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)






TE_models_Other_dotplot$rank_TE_PMA <- rank(-TE_models_Other_dotplot$TE_PMA)

ggplot(TE_models_Other_dotplot,aes(y=TE_PMA,x=rank_TE_PMA,color=TE_PMA))+
  geom_point()+
  scale_color_viridis(option = "G",direction = 1,end = 0.85)+
  #geom_abline(slope = 1,intercept = c(0))+
  #geom_abline(slope = 1,intercept = c(-1,1), linetype="dashed", color="blue", size=1)+
  ggrepel::geom_text_repel(data = TE_models_Other_dotplot[TE_models_Other_dotplot$TE_PMA>2,],min.segment.length = 0, aes(label=ID), size=3, nudge_x = 10,force = 1)+
  scale_x_continuous(limits = c(0,80))+
  theme_minimal()+
  theme(aspect.ratio = 0.5)

```





```{r radar plot RNA + CN models}

CN_model_perf <- read.delim("/home/ben/Analysis/RF_human/immune_system/CN_model_perf_w-wo_RNA_data.csv", sep=";", dec=",")
# colnames(CN_model_perf) <- c("population","Rsq")
# CN_model_perf <- CN_model_perf[2:20,]
CN_model_perf <- rbind(data.frame("X"="","RNA.seq.only"=0,"library.only"=0,"RNA.seq...library"=0),CN_model_perf)
CN_model_perf$X <- factor(CN_model_perf$X, levels =CN_model_perf$X)
dim(CN_model_perf)


ggplot(CN_model_perf, aes(x=X, y=library.only, fill=library.only))+
  geom_bar(stat = "identity",show.legend = F, color="black")+
  scale_y_continuous(limits = c(0,0.61),breaks = c(0.1,0.2,0.3,0.4,0.5,0.6))+
  # geom_hline(yintercept = 0.3872, linetype="dotted")+
  scale_fill_viridis(option = "H", direction = 1, end = 1, begin = 0.15)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.title.y = element_blank(),axis.text.y = element_blank())+
  annotate(x = 1, y = 0.2, label = "0.2", geom = "text")+
  annotate(x = 1, y = 0.3, label = "0.3", geom = "text")+
  annotate(x = 1, y = 0.4, label = "0.4", geom = "text")+
  annotate(x = 1, y = 0.5, label = "0.5", geom = "text")+
  annotate(x = 1, y = 0.6, label = "0.6", geom = "text")+
  coord_polar(start = -0.15)+
  geom_text(aes(label = round(library.only,digits = 2)), hjust = 0, nudge_y = -0.05, nudge_x = -0.25, colour = "black", size = 3, bg.r = 0)




# rownames(CN_model_perf) <- CN_model_perf$X
# CN_model_perf$X <- NULL
# CN_model_perf <- data.frame(t(CN_model_perf))
# CN_model_perf$group <- rownames(CN_model_perf)
# CN_model_perf <- cbind(CN_model_perf[22:22],CN_model_perf[1:21])
# 
# ggradar::ggradar(CN_model_perf,values.radar = c(0.4,0.5,0.6),grid.max = 0.6,centre.y = 0,legend.position = "bottom",)
CN_model_perf <- reshape2::melt(CN_model_perf)
CN_model_perf <- CN_model_perf[order(CN_model_perf$variable,decreasing = T),]

ggplot(CN_model_perf,aes(x=X,y=value,group=variable, color=variable))+
  geom_line()+
  geom_point()+
  scale_y_continuous(limits = c(0,0.7), expand = c(0,0),breaks = c(0.2,0.3,0.4,0.5,0.6,0.7))+
  coord_polar(start = -0.15)+
  theme_minimal()+
  theme(aspect.ratio = 1, axis.text.y = element_blank(), axis.title.y = element_blank())+
  annotate(x = 1, y = 0.2, label = "0.2", geom = "text")+
  annotate(x = 1, y = 0.3, label = "0.3", geom = "text")+
  annotate(x = 1, y = 0.4, label = "0.4", geom = "text")+
  annotate(x = 1, y = 0.5, label = "0.5", geom = "text")+
  annotate(x = 1, y = 0.6, label = "0.6", geom = "text")+
  annotate(x = 1, y = 0.6, label = "0.7", geom = "text")
  # geom_text(aes(label = round(value,digits = 2)), hjust = 0, nudge_y = -0.05, colour = "black", size = 3)

# 
# ggplot(CN_model_perf, aes(fill=variable, y=value, x=X)) + 
#   geom_bar(stat="identity",position="identity")+
#   scale_y_continuous(limits = c(0,0.7), expand = c(0,0),breaks = c(0.2,0.3,0.4,0.5,0.6))+
#   coord_polar(start = -0.15)+
#   theme_minimal()+
#   theme(aspect.ratio = 1, axis.text.y = element_blank(), axis.title.y = element_blank())+
#   annotate(x = 1, y = 0.2, label = "0.2", geom = "text")+
#   annotate(x = 1, y = 0.3, label = "0.3", geom = "text")+
#   annotate(x = 1, y = 0.4, label = "0.4", geom = "text")+
#   annotate(x = 1, y = 0.5, label = "0.5", geom = "text")+
#   annotate(x = 1, y = 0.6, label = "0.6", geom = "text")





```








