---
title: "Random Forest modeling of mRNA level"
author: "Benoit Nicolet"
date: "27/01/2021"
output: html_document
---

```{r setup, include=FALSE}

library(plyr)
library(dplyr)
library(doMC)
library(randomForest)
library(biomaRt)
library(ggplot2)
library(tidyverse)
library(caret)
library(e1071)


knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set("/home/ben/Analysis/RF_human/Jurkat_stability/")
setwd("/home/ben/Analysis/RF_human/Jurkat_stability/")


```


```{r BiomaRt}

listEnsemblArchives()

ensembl <- useEnsembl(biomart="ensembl",dataset="hsapiens_gene_ensembl", host = "http://apr2018.archive.ensembl.org")

## Here I wanted to get the geneID + transcriptID + Gene name + type of gene, all this for coding genes only:
tx2gene <- getBM(attributes=c("ensembl_transcript_id_version","external_gene_name"), mart = ensembl)

tx_annotations <- getBM(attributes=c("ensembl_gene_id","external_gene_name","gene_biotype"), mart = ensembl)
tx_annotations <- tx_annotations[!duplicated(tx_annotations$ensembl_gene_id),]
#dim(tx2gene)

```


```{r getting parameters}

Sequence_parameters_RNA <- read.delim("/home/ben/Analysis/RF_human/sequences/Sequence_parameters_RNA_23-04-21.csv", sep=";", dec=",")

```


## Prepararing the data of Tactivated
```{r prep with Tactivated data}

TPM_0h <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Tnaive_0h/quant.sf")
TPM_6h <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Tnaive_0h/quant.sf")
TPM_24h <- read.delim("/home/ben/Analysis/RF_human/CD4_activation_Roger_Geiger/Tnaive_0h/quant.sf")
dim(TPM_0h)

TPM_0h <- data.frame("ID"=TPM_0h$Name,"TPM"=TPM_0h$TPM)
TPM_0h <- subset(TPM_0h, TPM_0h$TPM>0)
TPM_0h <- merge(TPM_0h, tx2gene, by.x="ID", by.y="ensembl_transcript_id_version", all.x=T)

TPM_6h <- data.frame("ID"=TPM_6h$Name,"TPM"=TPM_6h$TPM)
TPM_6h <- subset(TPM_6h, TPM_0h$TPM>0)
TPM_6h <- merge(TPM_6h, tx2gene, by.x="ID", by.y="ensembl_transcript_id_version", all.x=T)

TPM_24h <- data.frame("ID"=TPM_24h$Name,"TPM"=TPM_24h$TPM)
TPM_24h <- subset(TPM_24h, TPM_24h$TPM>0)
TPM_24h <- merge(TPM_24h, tx2gene, by.x="ID", by.y="ensembl_transcript_id_version", all.x=T)

registerDoMC(4)

TPM_0h <- ddply(TPM_0h,"external_gene_name", numcolwise(sum), .parallel = T, .progress = T)
TPM_0h$TPM <- log10(TPM_0h$TPM)
TPM_0h <- subset(TPM_0h, TPM_0h$TPM>=-1)
colnames(TPM_0h)[1]="ID"
dim(TPM_0h) # 14081 genes 

TPM_0h <- merge(TPM_0h,Sequence_parameters_RNA,by="ID", all.x=T)
TPM_0h[is.na(TPM_0h)]= 0


rownames(TPM_0h) <- TPM_0h$ID
TPM_0h$ID <- NULL
# 
# dim(TPM_0h)
# TPM_2G[1:(ncol(TPM_2G))] <- lapply(TPM_2G[1:(ncol(TPM_2G))], function(x) {
#     if(is.character(x)) as.numeric(x) else x
# })

# TPM_CX3CR1_Tn<- do.call(data.frame,lapply(TPM_CX3CR1_Tn,function(x) replace(x, is.infinite(x),NA)))
```



## Selecting promising features for mRNA model
```{r selecting promising features for mRNA model}

## Here I remove the features with near zero variance & found in less than 1% of all genes in the dataset ##
registerDoMC(4)

## Tact ##
TPM_0h[is.na(TPM_0h)]=0
TPM_0h <- TPM_0h[, -nearZeroVar(TPM_0h, allowParallel = T, uniqueCut = dim(TPM_0h)[1]*0.01)]
dim(TPM_0h)



```


## Making space in RAM
```{r making space in RAM, echo=FALSE}

#rm(Gene_wise_5UTR_list,Gene_wise_CDS_list,Gene_wise_3UTR_list,Gene_annotation,UTR3_param,UTR5_param,CDS_param,Gene_wise_CDS_list)
gc()

```



## Training the RF model for Tactivated
```{r Training the RF model for Tactivated}
dim(TPM_2G) # 12999  2722

gc()
registerDoMC(10)

set.seed(12345)

# This bit of code is used to set a custom seed to each core to be passed on to seeds= of trainControl()
#length is = (n_repeats*nresampling)+1
# seeds <- vector(mode = "list", length = 16)
# #(3 is the number of tuning parameter, mtry for rf, here equal to ncol(iris)-2)
# for(i in 1:12) seeds[[i]]<- sample.int(n=1000, 1)
# #for the last model
# seeds[[13]]<-sample.int(1000, 1)

## For the choice of cv of repeated CV, Here is a good post : ##
# http://appliedpredictivemodeling.com/blog/2014/11/27/vpuig01pqbklmi72b8lcl3ij5hj2qm #


control <- trainControl(method="cv",
                        number=10,
                        #repeats = 2,
                        verboseIter = TRUE,
                        allowParallel = TRUE,
                        seeds = NULL) # NULL will set the seeds using a random set of integers (from doc)


start_time <- Sys.time()
rf_TPM_Tn_0h <- train(TPM~.,
                       data=TPM_0h,
                       method="rf",
                       ntree=100,
                       trControl=control,
                       metric="Rsquared",
                       tuneGrid= data.frame(mtry = round((ncol(TPM_0h)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)

end_time <- Sys.time()
end_time - start_time # Time difference of 16.78211 hours


rf_TPM_Tn_0h

# Random Forest
#
# 12999 samples
#  2721 predictor
#
# No pre-processing
# Resampling: Cross-Validated (10 fold)
# Summary of sample sizes: 11699, 11699, 11699, 11699, 11699, 11699, ...
# Resampling results:
#
#   RMSE       Rsquared   MAE
#   0.6057513  0.2010541  0.4847257
#
# Tuning parameter 'mtry' was held constant at a value of 907


rf_TPM_Tn_0h$finalModel
#                Type of random forest: regression
#                      Number of trees: 1000
# No. of variables tried at each split: 907
#
#           Mean of squared residuals: 0.3664218
#                     % Var explained: 19

saveRDS(rf_TPM_Tn_0h, "/home/ben/Analysis/RF_human/final_models/rf_TPM_Tn_0h_30_04_2021.RDS")
rf_TPM_Tn_0h <- readRDS("/home/ben/Analysis/RF_human/final_models/rf_TPM_Tn_0h_30_04_2021.RDS")


imp_rf_TPM_Tn_0h<- data.frame(varImp(rf_TPM_Tn_0h$finalModel))
imp_rf_TPM_Tn_0h$ID <- rownames(imp_rf_TPM_Tn_0h)

write.table(imp_rf_TPM_2G_RNA,"/home/ben/Analysis/RF_human/final_models/imp_rf_2G_RNA_04_02_2021.csv",sep=";",dec=",")

```



## training a model on all features for Tn protein levels
```{r training a model on all features for Tn protein levels}
# dim(TPM_Tn_param) # 11523  2680
# 
# gc()
# registerDoMC(10)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 2,
#                         verboseIter = TRUE,
#                         allowParallel = TRUE,
#                         seeds = NULL)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPMTn_RNA <- train(Tn_TPM~., 
#                        data=TPM_Tn_param, 
#                        method="rf", 
#                        ntree=1000,
#                        trControl=control, 
#                        metric="Rsquared",
#                        tuneGrid= data.frame(mtry = round((ncol(TPM_Tn_param)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 13.28999 hours
# 
# rf_TPMTn_RNA
# # Random Forest 
# # 
# # 11523 samples
# #  2679 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 10370, 10370, 10371, 10371, 10371, 10370, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared  MAE    
# #   0.5489673  0.233446  0.42663
# # 
# # Tuning parameter 'mtry' was held constant at a value of 893
# 
# rf_TPMTn_RNA$finalModel
# #                Type of random forest: regression
# #                      Number of trees: 1000
# # No. of variables tried at each split: 893
# # 
# #           Mean of squared residuals: 0.3008455
# #                     % Var explained: 22.44
# 
# saveRDS(rf_TPMTn_RNA, "/home/ben/Analysis/RF_human/final_models/rf_Tn_RNA_05_02_2021.RDS")
# 
# 
# imp_rf_TPMTn_RNA <- data.frame(varImp(rf_TPMTn_RNA$finalModel))
# imp_rf_TPMTn_RNA$ID <- rownames(imp_rf_TPMTn_RNA)
# 
# write.table(imp_rf_TPMTn_RNA,"/home/ben/Analysis/RF_human/final_models/imp_rf_Tn_RNA_05_02_2021.csv",sep=";",dec=",")

```


## training a model on all features for Tcm protein levels
```{r training a model on all features for Tcm protein levels}
# dim(TPM_Tcm_param) # 11687  2683
# 
# gc()
# registerDoMC(10)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 2,
#                         verboseIter = TRUE,
#                         allowParallel = TRUE,
#                         seeds=NULL)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPMTcm_RNA <- train(Tcm_TPM~., 
#                        data=TPM_Tcm_param, 
#                        method="rf", 
#                        ntree=1000,
#                        trControl=control, 
#                        metric="Rsquared",
#                        tuneGrid= data.frame(mtry = round((ncol(TPM_Tcm_param)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 13.60298 hours
# 
# rf_TPMTcm_RNA
# 
# # Random Forest 
# # 
# # 11687 samples
# #  2682 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 10518, 10518, 10518, 10518, 10518, 10518, ... 
# # Resampling results:
# # 
# #   RMSE      Rsquared   MAE      
# #   0.558513  0.2272691  0.4371935
# # 
# # Tuning parameter 'mtry' was held constant at a value of 894
# 
# rf_TPMTcm_RNA$finalModel
# 
# #                Type of random forest: regression
# #                      Number of trees: 1000
# # No. of variables tried at each split: 894
# # 
# #           Mean of squared residuals: 0.3121242
# #                     % Var explained: 21.77
# 
# saveRDS(rf_TPMTcm_RNA, "/home/ben/Analysis/RF_human/final_models/rf_Tcm_RNA_06_02_2021.RDS")
# 
# 
# imp_rf_TPMTcm_RNA <- data.frame(varImp(rf_TPMTcm_RNA$finalModel))
# imp_rf_TPMTcm_RNA$ID <- rownames(imp_rf_TPMTcm_RNA)
# 
# write.table(imp_rf_TPMTcm_RNA,"/home/ben/Analysis/RF_human/final_models/imp_rf_Tcm_RNA_06_02_2021.csv",sep=";",dec=",")

```




## training a model on all features for Tem protein levels
```{r training a model on all features for Tem protein levels}
# dim(TPM_Tem_param) # 11474  2686
# 
# gc()
# registerDoMC(10)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 2,
#                         verboseIter = TRUE,
#                         allowParallel = TRUE,
#                         seeds=NULL)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPMTem_RNA <- train(Tem_TPM~., 
#                        data=TPM_Tem_param, 
#                        method="rf", 
#                        ntree=1000,
#                        trControl=control, 
#                        metric="Rsquared",
#                        tuneGrid= data.frame(mtry = round((ncol(TPM_Tem_param)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 13.27845 hours
# 
# rf_TPMTem_RNA
# 
# # Random Forest 
# # 
# # 11474 samples
# #  2685 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 10328, 10326, 10327, 10327, 10326, 10326, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared   MAE      
# #   0.5612216  0.2192643  0.4375711
# # 
# # Tuning parameter 'mtry' was held constant at a value of 895
# 
# rf_TPMTem_RNA$finalModel
# 
# #                Type of random forest: regression
# #                      Number of trees: 1000
# # No. of variables tried at each split: 895
# # 
# #           Mean of squared residuals: 0.3147301
# #                     % Var explained: 21.16
# 
# saveRDS(rf_TPMTem_RNA, "/home/ben/Analysis/RF_human/final_models/rf_Tem_RNA_07_02_2021.RDS")
# 
# 
# imp_rf_TPMTem_RNA <- data.frame(varImp(rf_TPMTem_RNA$finalModel))
# imp_rf_TPMTem_RNA$ID <- rownames(imp_rf_TPMTem_RNA)
# 
# write.table(imp_rf_TPMTem_RNA,"/home/ben/Analysis/RF_human/final_models/imp_rf_Tem_RNA_07_02_2021.csv",sep=";",dec=",")

```



## training a model on all features for random data
```{r training a model on all features for random data}
# dim(TPM_2G) # 12999  2722
# 
# TPM_2G_random <- TPM_2G[sample(nrow(TPM_2G)),]
# 
# mean(TPM_2G$TPM_2G[TPM_2G$TPM_2G>0]) # 1.287768
# sd(TPM_2G$TPM_2G[TPM_2G$TPM_2G>0]) # 0.6726224
# dim(TPM_2G) # 12999  2722
# 
# TPM_2G_random$TPM_2G <- rnorm(12999, mean=1.287768, sd=0.6726224)
# 
# 
# plot(TPM_2G_random$TPM_2G, TPM_2G_random$ATTTA_UTR3)
# plot(TPM_2G$TPM_2G, TPM_2G_random$ATTTA_UTR3)
# 
# plot(TPM_2G_random$TPM_2G, TPM_2G$TPM_2G)
# 
# 
# gc()
# registerDoMC(10)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 2,
#                         verboseIter = TRUE,
#                         allowParallel = TRUE,
#                         seeds=NULL)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_random_RNA <- train(TPM_2G~., 
#                        data=TPM_2G_random, 
#                        method="rf", 
#                        ntree=100,
#                        trControl=control, 
#                        metric="Rsquared",
#                        tuneGrid= data.frame(mtry = round((ncol(TPM_2G_random)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 2.110838 hours (100 trees...)
# 
# rf_TPM_random_RNA
# 
# # Random Forest 
# # 
# # 12999 samples
# #  2721 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 11699, 11699, 11699, 11699, 11699, 11699, ... 
# # Resampling results:
# # 
# #   RMSE      Rsquared     MAE      
# #   0.676439  0.001030089  0.5389773
# # 
# # Tuning parameter 'mtry' was held constant at a value of 907
# 
# rf_TPM_random_RNA$finalModel
# 
# #                Type of random forest: regression
# #                      Number of trees: 100
# # No. of variables tried at each split: 907
# # 
# #           Mean of squared residuals: 0.4667284
# #                     % Var explained: -3.07
# 
# saveRDS(rf_TPM_random_RNA, "/home/ben/Analysis/RF_human/final_models/rf_random_RNA_29_01_2021.RDS")
# 
# 
# imp_rf_TPM_random_RNA <- data.frame(varImp(rf_TPM_random_RNA$finalModel))
# imp_rf_TPM_random_RNA$ID <- rownames(imp_rf_TPM_random_RNA)
# 
# write.table(imp_rf_TPM_random_RNA,"/home/ben/Analysis/RF_human/final_models/imp_rf_random_RNA_29_01_2021.csv",sep=";",dec=",")

```







```{r visualizing motifs}
# ggplot(TPM_2G, aes(x=TATATT_UTR3, y=TPM_2G))+
#       geom_point()+
#       geom_smooth(method = "lm")+
#       #ylab("% increase MSE for CN")+
#       #xlab("CTCAGGT occurence")+
#       ggtitle("")+
#       scale_y_continuous(limit=c(-1,5))+
#       scale_x_continuous(limit=c(0,50))+
#       theme_classic()+
#       theme(aspect.ratio = 1)
# 
# ggplot(TPM_Tem_param, aes(x=AGAAGA_CDS, y=TPM_Tem_param$Tem_TPM))+
#       geom_point()+
#       geom_smooth(method = "lm")+
#       #ylab("% increase MSE for CN")+
#       xlab("CTCA occurence")+
#       ggtitle("")+
#       scale_y_continuous(limit=c(-1,5))+
#       scale_x_continuous(limit=c(0,50))+
#       theme_classic()+
#       theme(aspect.ratio = 1)

```


# region-specific contribution to the models
```{r preping region-specific stuff}

TPM_Tn_param_UTR5 <- cbind(TPM_Tn_param %>% dplyr::select(contains(c("TPM","UTR5"))))
TPM_Tn_param_CDS <- cbind(TPM_Tn_param %>% dplyr::select(contains(c("TPM","CDS"))))
TPM_Tn_param_UTR3 <- cbind(TPM_Tn_param %>% dplyr::select(contains(c("TPM","UTR3"))))
TPM_Tn_param_else <- cbind(TPM_Tn_param %>% dplyr::select(!contains(c("TPM","UTR5","CDS","UTR3"))))

TPM_Tcm_param_UTR5 <- cbind(TPM_Tcm_param %>% dplyr::select(contains(c("TPM","UTR5"))))
TPM_Tcm_param_CDS <- cbind(TPM_Tcm_param %>% dplyr::select(contains(c("TPM","CDS"))))
TPM_Tcm_param_UTR3 <- cbind(TPM_Tcm_param %>% dplyr::select(contains(c("TPM","UTR3"))))
TPM_Tcm_param_else <- cbind(TPM_Tcm_param %>% dplyr::select(!contains(c("TPM","UTR5","CDS","UTR3"))))

TPM_Tem_param_UTR5 <- cbind(TPM_Tem_param %>% dplyr::select(contains(c("TPM","UTR5"))))
TPM_Tem_param_CDS <- cbind(TPM_Tem_param %>% dplyr::select(contains(c("TPM","CDS"))))
TPM_Tem_param_UTR3 <- cbind(TPM_Tem_param %>% dplyr::select(contains(c("TPM","UTR3"))))
TPM_Tem_param_else <- cbind(TPM_Tem_param %>% dplyr::select(!contains(c("TPM","UTR5","CDS","UTR3"))))

TPM_2G_UTR5 <- cbind(TPM_2G %>% dplyr::select(contains(c("TPM","UTR5"))))
TPM_2G_CDS <- cbind(TPM_2G %>% dplyr::select(contains(c("TPM","CDS"))))
TPM_2G_UTR3 <- cbind(TPM_2G %>% dplyr::select(contains(c("TPM","UTR3"))))
TPM_2G_else <- cbind(TPM_2G %>% dplyr::select(!contains(c("TPM","UTR5","CDS","UTR3"))))


## Lets see how to devide the number of trees (1000 for all features) to allow approximation of feature importance values ##
dim(TPM_2G) # 2722
dim(TPM_2G_CDS) # 1365  = 500 trees
dim(TPM_2G_UTR5) # 257 = 100 trees
dim(TPM_2G_UTR3) # 1096 = 400 trees

#1096/2722*1000


```



```{r RF CDS contribution T act}
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_2G_CDS <- train(TPM_2G~., 
#                        TPM_2G_CDS, 
#                        method="rf", 
#                        ntree=500,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(TPM_2G_CDS)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 4.494687 hours
# 
# rf_TPM_2G_CDS
# # Random Forest 
# # 
# # 12999 samples
# #  1365 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 11699, 11699, 11699, 11699, 11699, 11699, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared   MAE      
# #   0.6325819  0.1211445  0.5112261
# # 
# # Tuning parameter 'mtry' was held constant at a value of 455
# 
# 
# rf_TPM_2G_CDS$finalModel
# #                Type of random forest: regression
# #                      Number of trees: 500
# # No. of variables tried at each split: 455
# # 
# #           Mean of squared residuals: 0.3989556
# #                     % Var explained: 11.81
# 
# saveRDS(rf_TPM_2G_CDS,"/home/ben/Analysis/RF_human/final_models/rf_TPM_2G_CDS_22032021.RDS")

```



```{r RF CDS contribution Tn}

gc()
registerDoMC(22)
control <- trainControl(method="cv",
                        number=10,
                        #repeats = 5,
                        verboseIter = TRUE,
                        allowParallel = T)

set.seed(12345)
start_time <- Sys.time()
rf_TPM_Tn_param_CDS <- train(Tn_TPM~., 
                       TPM_Tn_param_CDS, 
                       method="rf", 
                       ntree=500,
                       trControl=control, 
                       tuneGrid = data.frame(mtry = round((ncol(TPM_Tn_param_CDS)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)


end_time <- Sys.time()
end_time - start_time # Time difference of 3.572417 hours

rf_TPM_Tn_param_CDS

# Random Forest 
# 
# 11523 samples
#  1346 predictor
# 
# No pre-processing
# Resampling: Cross-Validated (10 fold) 
# Summary of sample sizes: 10370, 10370, 10371, 10371, 10371, 10370, ... 
# Resampling results:
# 
#   RMSE       Rsquared   MAE      
#   0.5651159  0.1803214  0.4398406
# 
# Tuning parameter 'mtry' was held constant at a value of 449

rf_TPM_Tn_param_CDS$finalModel

#                Type of random forest: regression
#                      Number of trees: 500
# No. of variables tried at each split: 449
# 
#           Mean of squared residuals: 0.3195749
#                     % Var explained: 17.61
          
saveRDS(rf_TPM_Tn_param_CDS,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tn_param_CDS_24032021.RDS")

```


```{r RF CDS contribution Tcm}

gc()
registerDoMC(22)
control <- trainControl(method="cv",
                        number=10,
                        #repeats = 5,
                        verboseIter = TRUE,
                        allowParallel = T)

set.seed(12345)
start_time <- Sys.time()
rf_TPM_Tcm_param_CDS <- train(Tcm_TPM~., 
                       TPM_Tcm_param_CDS, 
                       method="rf", 
                       ntree=500,
                       trControl=control, 
                       tuneGrid = data.frame(mtry = round((ncol(TPM_Tcm_param_CDS)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)


end_time <- Sys.time()
end_time - start_time # Time difference of 3.648669 hours

rf_TPM_Tcm_param_CDS

# Random Forest 
# 
# 11687 samples
#  1348 predictor
# 
# No pre-processing
# Resampling: Cross-Validated (10 fold) 
# Summary of sample sizes: 10518, 10518, 10518, 10518, 10518, 10518, ... 
# Resampling results:
# 
#   RMSE       Rsquared   MAE      
#   0.5738121  0.1786782  0.4504573
# 
# Tuning parameter 'mtry' was held constant at a value of 449

rf_TPM_Tcm_param_CDS$finalModel
#                Type of random forest: regression
#                      Number of trees: 500
# No. of variables tried at each split: 449
# 
#           Mean of squared residuals: 0.3309899
#                     % Var explained: 17.05

saveRDS(rf_TPM_Tcm_param_CDS,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tcm_param_CDS_24032021.RDS")

```


```{r RF CDS contribution Tem}

gc()
registerDoMC(22)
control <- trainControl(method="cv",
                        number=10,
                        #repeats = 5,
                        verboseIter = TRUE,
                        allowParallel = T)

set.seed(12345)
start_time <- Sys.time()
rf_TPM_Tem_param_CDS <- train(Tem_TPM~., 
                       TPM_Tem_param_CDS, 
                       method="rf", 
                       ntree=500,
                       trControl=control, 
                       tuneGrid = data.frame(mtry = round((ncol(TPM_Tem_param_CDS)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)


end_time <- Sys.time()
end_time - start_time # Time difference of 3.553261 hours

rf_TPM_Tem_param_CDS

# Random Forest 
# 
# 11474 samples
#  1353 predictor
# 
# No pre-processing
# Resampling: Cross-Validated (10 fold) 
# Summary of sample sizes: 10328, 10326, 10327, 10327, 10326, 10326, ... 
# Resampling results:
# 
#   RMSE       Rsquared   MAE      
#   0.5766638  0.1709906  0.4513378
# 
# Tuning parameter 'mtry' was held constant at a value of 451

rf_TPM_Tem_param_CDS$finalModel
#                Type of random forest: regression
#                      Number of trees: 500
# No. of variables tried at each split: 451
# 
#           Mean of squared residuals: 0.3320364
#                     % Var explained: 16.83

saveRDS(rf_TPM_Tem_param_CDS,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tem_param_CDS_24032021.RDS")

```










```{r RF 5UTR contribution T act}
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_2G_UTR5 <- train(TPM_2G~., 
#                        TPM_2G_UTR5, 
#                        method="rf", 
#                        ntree=100,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(TPM_2G_UTR5)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 10.69706 mins
# 
# rf_TPM_2G_UTR5
# 
# # Random Forest 
# # 
# # 12999 samples
# #   257 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 11699, 11699, 11699, 11699, 11699, 11699, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared    MAE     
# #   0.6465254  0.07667401  0.521755
# # 
# # Tuning parameter 'mtry' was held constant at a value of 86
# 
# 
# rf_TPM_2G_UTR5$finalModel
# #                Type of random forest: regression
# #                      Number of trees: 100
# # No. of variables tried at each split: 86
# # 
# #           Mean of squared residuals: 0.4259605
# #                     % Var explained: 5.84
# 
# saveRDS(rf_TPM_2G_UTR5,"/home/ben/Analysis/RF_human/final_models/rf_TPM_2G_UTR5_22032021.RDS")

```



```{r RF UTR5 contribution Tn}
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_Tn_param_UTR5 <- train(Tn_TPM~., 
#                        TPM_Tn_param_UTR5, 
#                        method="rf", 
#                        ntree=100,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(TPM_Tn_param_UTR5)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 8.285611 mins
# 
# rf_TPM_Tn_param_UTR5
# 
# Random Forest 
# 
# # 11523 samples
# #   247 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 10370, 10370, 10371, 10371, 10371, 10370, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared   MAE      
# #   0.6018966  0.0663666  0.4761971
# # 
# # Tuning parameter 'mtry' was held constant at a value of 82
# 
# rf_TPM_Tn_param_UTR5$finalModel
# 
# #                Type of random forest: regression
# #                      Number of trees: 100
# # No. of variables tried at each split: 82
# # 
# #           Mean of squared residuals: 0.3676859
# #                     % Var explained: 5.21
# 
# saveRDS(rf_TPM_Tn_param_UTR5,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tn_param_UTR5_22032021.RDS")

```


```{r RF UTR5 contribution Tcm}
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_Tcm_param_UTR5 <- train(Tcm_TPM~., 
#                        TPM_Tcm_param_UTR5, 
#                        method="rf", 
#                        ntree=100,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(TPM_Tcm_param_UTR5)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 8.383726 mins
# 
# rf_TPM_Tcm_param_UTR5
# 
# # Random Forest 
# # 
# # 11687 samples
# #   247 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 10518, 10518, 10518, 10518, 10518, 10518, ... 
# # Resampling results:
# # 
# #   RMSE      Rsquared    MAE      
# #   0.611652  0.06328233  0.4866148
# # 
# # Tuning parameter 'mtry' was held constant at a value of 82
# 
# rf_TPM_Tcm_param_UTR5$finalModel
# 
# #               Type of random forest: regression
# #                      Number of trees: 100
# # No. of variables tried at each split: 82
# # 
# #           Mean of squared residuals: 0.3803609
# #                     % Var explained: 4.67
# 
# saveRDS(rf_TPM_Tcm_param_UTR5,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tcm_param_UTR5_22032021.RDS")

```


```{r RF UTR5 contribution Tcm}
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_Tem_param_UTR5 <- train(Tem_TPM~., 
#                        TPM_Tem_param_UTR5, 
#                        method="rf", 
#                        ntree=100,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(TPM_Tem_param_UTR5)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 8.172811 mins
# 
# rf_TPM_Tem_param_UTR5
# 
# # Random Forest 
# # 
# # 11474 samples
# #   245 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 10328, 10326, 10327, 10327, 10326, 10326, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared    MAE      
# #   0.6150646  0.05290693  0.4882962
# # 
# # Tuning parameter 'mtry' was held constant at a value of 82
# 
# rf_TPM_Tem_param_UTR5$finalModel
# #                Type of random forest: regression
# #                      Number of trees: 100
# # No. of variables tried at each split: 82
# # 
# #           Mean of squared residuals: 0.3837966
# #                     % Var explained: 3.86
# 
# saveRDS(rf_TPM_Tem_param_UTR5,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tem_param_UTR5_22032021.RDS")

```










```{r RF 3UTR contribution T act}
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=10,
#                         #repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_TPM_2G_UTR3 <- train(TPM_2G~., 
#                        TPM_2G_UTR3, 
#                        method="rf", 
#                        ntree=400,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(TPM_2G_UTR3)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 3.063828 hours
# 
# rf_TPM_2G_UTR3
# 
# # Random Forest 
# # 
# # 12999 samples
# #  1096 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (10 fold) 
# # Summary of sample sizes: 11699, 11699, 11699, 11699, 11699, 11699, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared    MAE      
# #   0.6463367  0.08236495  0.5233491
# # 
# # Tuning parameter 'mtry' was held constant at a value of 365
# 
# rf_TPM_2G_UTR3$finalModel
# #                Type of random forest: regression
# #                      Number of trees: 400
# # No. of variables tried at each split: 365
# # 
# #           Mean of squared residuals: 0.4189629
# #                     % Var explained: 7.39
# 
# saveRDS(rf_TPM_2G_UTR3,"/home/ben/Analysis/RF_human/final_models/rf_TPM_2G_UTR3_22032021.RDS")

```



```{r RF UTR3 contribution Tn}

gc()
registerDoMC(22)
control <- trainControl(method="cv",
                        number=10,
                        #repeats = 5,
                        verboseIter = TRUE,
                        allowParallel = T)

set.seed(12345)
start_time <- Sys.time()
rf_TPM_Tn_param_UTR3 <- train(Tn_TPM~., 
                       TPM_Tn_param_UTR3, 
                       method="rf", 
                       ntree=400,
                       trControl=control, 
                       tuneGrid = data.frame(mtry = round((ncol(TPM_Tn_param_UTR3)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)


end_time <- Sys.time()
end_time - start_time # Time difference of 2.466456 hours

rf_TPM_Tn_param_UTR3

# Random Forest 
# 
# 11523 samples
#  1083 predictor
# 
# No pre-processing
# Resampling: Cross-Validated (10 fold) 
# Summary of sample sizes: 10370, 10370, 10371, 10371, 10371, 10370, ... 
# Resampling results:
# 
#   RMSE       Rsquared    MAE      
#   0.5943499  0.09147841  0.4650077
# 
# Tuning parameter 'mtry' was held constant at a value of 361

rf_TPM_Tn_param_UTR3$finalModel
#                      Number of trees: 400
# No. of variables tried at each split: 361
# 
#           Mean of squared residuals: 0.3546141
#                     % Var explained: 8.58


saveRDS(rf_TPM_Tn_param_UTR3,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tn_param_UTR3_24032021.RDS")

```


```{r RF UTR3 contribution Tcm}

gc()
registerDoMC(22)
control <- trainControl(method="cv",
                        number=10,
                        #repeats = 5,
                        verboseIter = TRUE,
                        allowParallel = T)

set.seed(12345)
start_time <- Sys.time()
rf_TPM_Tcm_param_UTR3 <- train(Tcm_TPM~., 
                       TPM_Tcm_param_UTR3, 
                       method="rf", 
                       ntree=400,
                       trControl=control, 
                       tuneGrid = data.frame(mtry = round((ncol(TPM_Tcm_param_UTR3)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)


end_time <- Sys.time()
end_time - start_time # Time difference of 2.552044 hours

rf_TPM_Tcm_param_UTR3

# Random Forest 
# 
# 11687 samples
#  1084 predictor
# 
# No pre-processing
# Resampling: Cross-Validated (10 fold) 
# Summary of sample sizes: 10518, 10518, 10518, 10518, 10518, 10518, ... 
# Resampling results:
# 
#   RMSE       Rsquared    MAE    
#   0.6030764  0.09055214  0.47521
# 
# Tuning parameter 'mtry' was held constant at a value of 361

rf_TPM_Tcm_param_UTR3$finalModel
#                Type of random forest: regression
#                      Number of trees: 400
# No. of variables tried at each split: 361
# 
#           Mean of squared residuals: 0.363853
#                     % Var explained: 8.81

saveRDS(rf_TPM_Tcm_param_UTR3,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tcm_param_UTR3_24032021.RDS")

```


```{r RF UTR3 contribution Tcm}

gc()
registerDoMC(22)
control <- trainControl(method="cv",
                        number=10,
                        #repeats = 5,
                        verboseIter = TRUE,
                        allowParallel = T)

set.seed(12345)
start_time <- Sys.time()
rf_TPM_Tem_param_UTR3 <- train(Tem_TPM~., 
                       TPM_Tem_param_UTR3, 
                       method="rf", 
                       ntree=400,
                       trControl=control, 
                       tuneGrid = data.frame(mtry = round((ncol(TPM_Tem_param_UTR3)-1)/3)),
                       na.action = na.omit,
                       importance = TRUE,
                       verbose = TRUE)


end_time <- Sys.time()
end_time - start_time # Time difference of  2.45578 hours

rf_TPM_Tem_param_UTR3

# Random Forest 
# 
# 11474 samples
#  1084 predictor
# 
# No pre-processing
# Resampling: Cross-Validated (10 fold) 
# Summary of sample sizes: 10328, 10326, 10327, 10327, 10326, 10326, ... 
# Resampling results:
# 
#   RMSE       Rsquared    MAE      
#   0.6038329  0.08907385  0.4742389
# 
# Tuning parameter 'mtry' was held constant at a value of 361

rf_TPM_Tem_param_UTR3$finalModel

#                Type of random forest: regression
#                      Number of trees: 400
# No. of variables tried at each split: 361
# 
#           Mean of squared residuals: 0.3658118
#                     % Var explained: 8.37

saveRDS(rf_TPM_Tem_param_UTR3,"/home/ben/Analysis/RF_human/final_models/rf_TPM_Tem_param_UTR3_24032021.RDS")

```





# ```{r RF CDS non amino / codon contribution}
# 
# dim(CDS_motif_param_TPM2G) # 4648 2251
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=3,
#                         repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_CDS_param_nocodon_noAA_TPM2G <- train(avg_log_TPM~., 
#                        CDS_motif_param_TPM2G, 
#                        method="rf", 
#                        ntree=222,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(CDS_motif_param_TPM2G)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # 
# 
# rf_CDS_param_nocodon_noAA_TPM2G
# 
# rf_CDS_param_nocodon_noAA_TPM2G$finalModel
# 
# saveRDS(rf_CDS_param_nocodon_noAA_TPM2G,"rf_CDS_param_nocodon_noAA_TPM2G_contribution.RDS")
# 
# 
# 
# ```
# 
# 
# 
# ```{r RF CDS Amino Acid only contribution}
# 
# dim(CDS_AA_param_TPM2G) # 4648  21
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=3,
#                         repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_CDS_AAonly_param_TPM2G <- train(avg_log_TPM~., 
#                        CDS_AA_param_TPM2G, 
#                        method="rf", 
#                        ntree=222,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(CDS_AA_param_TPM2G)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 1.010237 mins
# 
# rf_CDS_AAonly_param_TPM2G
# # Random Forest 
# # 
# # 4648 samples
# #   20 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (3 fold, repeated 5 times) 
# # Summary of sample sizes: 3098, 3100, 3098, 3099, 3098, 3099, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared  MAE      
# #   0.6210629  0.32338   0.5003818
# # 
# # Tuning parameter 'mtry' was held constant at a value of 7
# 
# rf_CDS_AAonly_param_TPM2G$finalModel
# 
# 
# # Type of random forest: regression
# #                      Number of trees: 500
# # No. of variables tried at each split: 7
# # 
# #           Mean of squared residuals: 0.3835968
# #                     % Var explained: 32.34
# 
# saveRDS(rf_CDS_AAonly_param_TPM2G,"rf_CDS_AA_only_param_TPM2G_contribution.RDS")
# 
# 
# 
# ```
# 
# 
# 
# ```{r RF CDS codon contribution}
# 
# dim(CDS_codon_param_TPM2G) # 4648  65
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=3,
#                         repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_CDS_codon_only_param_TPM2G <- train(avg_log_TPM~., 
#                        CDS_codon_param_TPM2G, 
#                        method="rf", 
#                        ntree=222,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(CDS_codon_param_TPM2G)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # 
# 
# rf_CDS_codon_only_param_TPM2G
# 
# # Random Forest 
# # 
# # 4648 samples
# #   64 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (3 fold, repeated 5 times) 
# # Summary of sample sizes: 3098, 3100, 3098, 3099, 3098, 3099, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared   MAE     
# #   0.6171637  0.3462685  0.498724
# # 
# # Tuning parameter 'mtry' was held constant at a value of 21
# 
# 
# rf_CDS_codon_only_param_TPM2G$finalModel
# #                Type of random forest: regression
# #                      Number of trees: 500
# # No. of variables tried at each split: 21
# # 
# #           Mean of squared residuals: 0.3770835
# #                     % Var explained: 33.49
# 
# saveRDS(rf_CDS_codon_only_param_TPM2G,"rf_CDS_codon_only_param_TPM2G_contribution.RDS")
# 
# 
# 
# ```
# 
# 
# 
# 
# ```{r RF UTR5 contribution}
# 
# dim(UTR5_param_TPM2G) # 4648   30
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=3,
#                         repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_UTR5_param_TPM2G <- train(avg_log_TPM~., 
#                        UTR5_param_TPM2G, 
#                        method="rf", 
#                        ntree=500,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(UTR5_param_TPM2G)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 1.182964 mins
# 
# rf_UTR5_param_TPM2G
# 
# # Random Forest 
# # 
# # 4648 samples
# #   29 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (3 fold, repeated 5 times) 
# # Summary of sample sizes: 3098, 3100, 3098, 3099, 3098, 3099, ... 
# # Resampling results:
# # 
# #   RMSE      Rsquared    MAE      
# #   0.739245  0.04360613  0.5914947
# # 
# # Tuning parameter 'mtry' was held constant at a value of 10
# 
# 
# rf_UTR5_param_TPM2G$finalModel
# 
# #                Type of random forest: regression
# #                      Number of trees: 500
# # No. of variables tried at each split: 10
# # 
# #           Mean of squared residuals: 0.5414322
# #                     % Var explained: 4.5
# 
# saveRDS(rf_UTR5_param_TPM2G,"rf_UTR5_param_TPM2G_contribution.RDS")
# 
# 
# 
# ```
# 
# 
# 
# 
# ```{r RF UTR3 contribution}
# 
# dim(UTR3_param_TPM2G) # 4648  409
# 
# gc()
# registerDoMC(22)
# control <- trainControl(method="cv",
#                         number=3,
#                         repeats = 5,
#                         verboseIter = TRUE,
#                         allowParallel = T)
# 
# set.seed(12345)
# start_time <- Sys.time()
# rf_UTR3_param_TPM2G <- train(avg_log_TPM~., 
#                        UTR3_param_TPM2G, 
#                        method="rf", 
#                        ntree=500,
#                        trControl=control, 
#                        tuneGrid = data.frame(mtry = round((ncol(UTR3_param_TPM2G)-1)/3)),
#                        na.action = na.omit,
#                        importance = TRUE,
#                        verbose = TRUE)
# 
# 
# end_time <- Sys.time()
# end_time - start_time # Time difference of 8.715641 mins
# 
# rf_UTR3_param_TPM2G
# 
# # Random Forest 
# # 
# # 4648 samples
# #  233 predictor
# # 
# # No pre-processing
# # Resampling: Cross-Validated (3 fold, repeated 5 times) 
# # Summary of sample sizes: 3098, 3100, 3098, 3099, 3098, 3099, ... 
# # Resampling results:
# # 
# #   RMSE       Rsquared    MAE      
# #   0.7298375  0.06162596  0.5854489
# # 
# # Tuning parameter 'mtry' was held constant at a value of 78
# 
# rf_UTR3_param_TPM2G$finalModel
# 
# #                Type of random forest: regression
# #                      Number of trees: 500
# # No. of variables tried at each split: 78
# # 
# #           Mean of squared residuals: 0.5343584
# #                     % Var explained: 5.74
# 
# saveRDS(rf_UTR3_param_TPM2G,"rf_UTR3_param_TPM2G_contribution.RDS")
# 
# rf_UTR3_param_TPM2G <- readRDS("rf_UTR3_param_TPM2G_contribution.RDS")
# 
# imp_rf_UTR3_param_TPM2G<- data.frame(varImp(rf_UTR3_param_TPM2G$finalModel))
# imp_rf_UTR3_param_TPM2G$ID <- rownames(imp_rf_UTR3_param_TPM2G)
# 


# ``` 





