---
title: "Pilot screen analysis"
author: "Kaspar Bresser"
output: 
  github_document:
    toc: true
 # html_document: 
#    theme: simplex
#    highlight: pygments
#    code_folding: show
#    self_contained: TRUE
#    toc: yes
#    toc_float:
#      collapsed: no
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message=FALSE,
                      autodep  = TRUE,
                      cache = FALSE,
                      fig.width = 7,
                      fig.asp = 0.618,
                      fig.align = "center")
```


```{r loading, echo=FALSE}
library(rstatix)
library(GGally)
library(ggpubr)
library(tidyverse)
```


## Import and inspect

Import read counts

```{r import_data}
readcounts <- read_tsv("./output_all_filtered.tsv")

readcounts
```

update column names

```{r}
colnames(readcounts) <-  c("Motif", "HEK.GFP.A","HEK.GFP.B", 
                                   "HEK.Top.A","HEK.Top.B", 
                                   "HEK.Bot.A","HEK.Bot.B", 
                                   "HeLa.LoMi.A", "HeLa.LoMi.B", 
                                   "HeLa.LoLo.A", "HeLa.LoLo.B", 
                                   "HeLa.HiMi.A", "HeLa.HiMi.B",
                                   "HeLa.HiHi.A", "HeLa.HiHi.B"  ) 

readcounts
```


Lets keep only the stuff we'll need, and add a numeric identifier.

```{r filter}
readcounts %>% 
  filter(!str_detect(Motif, "__")) %>% 
  summarise(.,
            across(where(is.numeric), sum),
            across(where(is.character), ~"__aligned")) %>% 
  bind_rows(readcounts) -> readcounts

readcounts
```


```{r plot_stats, fig.width=6}
readcounts %>% 
  filter(str_detect(Motif, "__")) %>% 
  pivot_longer(cols = contains("."), names_to = "sample", values_to = "reads") %>%
  filter(!(reads == 0)) %>% 
  ggplot(aes(x = sample, y = reads, fill = fct_rev(Motif)))+
  geom_bar(stat = "identity", position = "fill", color = "black", width = .8)+
  theme_classic()+
  theme(panel.grid.major.y = element_line(), axis.text.x=element_text(angle=45,hjust=1))
```

Keep only the aligned reads

```{r filter_aligned}
readcounts %>% 
  filter(!str_detect(Motif, "__"))  -> aligned

write_tsv(aligned, "./Output/RE_013_readcounts_filt.tsv")
```



## correlation

```{r pearson, fig.width=12}
ggpairs(log10(aligned[1:6]+1))
ggsave("./Figs/cor_plots_HEK.png", width = 6, height = 5, scale = 1.5)

ggpairs(log10(aligned[7:14]+1))
ggsave("./Figs/cor_plots_HeLa.png", width = 6, height = 5, scale = 1.5)
```

## Filtering

HEK

```{r, fig.width=15}
aligned %>%
  select(Motif, contains("HEK")) %>% 
  rowwise() %>%
  filter(sum(c_across(2:7)) > 10) %>%
  ungroup() %>%
  mutate(across(where(is.numeric), ~.+1),
         across(where(is.numeric), ~./sum(.))*10000) %>%
  mutate(avg.top = ( HEK.Top.A + HEK.Top.B)/2,
         avg.bot = (HEK.Bot.A + HEK.Bot.B)/2,
         avg.gfp = (HEK.GFP.A + HEK.GFP.B)/2) -> aligned.HEK



write_tsv(aligned.HEK, "./Output/RE_013_HEK_normalized.tsv")


ggpairs(log10(aligned.HEK[2:7]))
ggsave("./Figs/cor_plots_normalized_HEK.png", width = 12, height = 11, scale = 1.5)
```

HeLa

```{r, fig.width=15}
aligned %>%
  select(Motif, contains("HeLa")) %>% 
  rowwise() %>%
  filter(sum(c_across(2:7)) > 10) %>%
  ungroup() %>%
  mutate(across(where(is.numeric), ~.+1),
         across(where(is.numeric), ~./sum(.))*10000) %>%
  mutate(LoMi = ( HeLa.LoMi.A + HeLa.LoMi.B)/2,
         LoLo = (HeLa.LoLo.A + HeLa.LoLo.A)/2,
         HiMi = (HeLa.HiMi.A + HeLa.HiMi.B)/2,
         HiHi = (HeLa.HiHi.A + HeLa.HiHi.B)/2) -> aligned.HeLa



write_tsv(aligned.HeLa, "./Output/RE_013_HeLa_normalized.tsv")


ggpairs(log10(aligned.HeLa[2:9]))
ggsave("./Figs/cor_plots_normalized_HeLa.png", width = 12, height = 11, scale = 1.5)
```

## Compare samples

Compare half-samples HEK


```{r fig.width=10, fig.asp=.35}
var.list1 <- list("HEK.Top.A", "HEK.Bot.A", "HEK.GFP.A")
var.list2 <- list("HEK.Top.B", "HEK.Bot.B", "HEK.GFP.B")

p.list <- map2( var.list1,
                var.list2, 
                ~ggplot(aligned.HEK, aes(.data[[.x]], .data[[.y]]))+
                    geom_point()+
                    scale_y_log10()+
                    scale_x_log10()+
                    theme_classic()+
                    theme(panel.grid.major = element_line())  +
                    stat_cor(method = "pearson", label.x = -.8, label.y = 2.6)
                )

ggarrange(plotlist = p.list, nrow = 1)

ggsave("./Figs/scatter_halfsamples_HEK.pdf", width = 9, height = 3)
```


Use the averages of the half-samples to compare the experimental samples

```{r fig.width=10, fig.asp=.35}
var.list1 <- list("avg.top", "avg.top", "avg.bot")
var.list2 <- list("avg.bot", "avg.gfp", "avg.gfp")

p.list <- map2( var.list1,
                var.list2, 
                ~ggplot(aligned.HEK, aes(.data[[.x]], .data[[.y]]))+
                    geom_point()+
                    scale_y_log10()+
                    scale_x_log10()+
                    theme_classic()+
                    theme(panel.grid.major = element_line()) +
                    stat_cor(method = "pearson", label.x = -.8, label.y = 2.6)
                )

ggarrange(plotlist = p.list, ncol = 3)

ggsave("./Figs/scattter_experimental_HEK.pdf", width = 9, height = 3)
```


Compare half-samples HeLa


```{r fig.width=10, fig.asp=.25}
var.list1 <- list("HeLa.LoLo.A", "HeLa.LoMi.A", "HeLa.HiMi.A", "HeLa.HiHi.A")
var.list2 <- list("HeLa.LoLo.B", "HeLa.LoMi.B", "HeLa.HiMi.B","HeLa.HiHi.B")

p.list <- map2( var.list1,
                var.list2, 
                ~ggplot(aligned.HeLa, aes(.data[[.x]], .data[[.y]]))+
                    geom_point()+
                    scale_y_log10()+
                    scale_x_log10()+
                    theme_classic()+
                    theme(panel.grid.major = element_line())  +
                    stat_cor(method = "pearson", label.x = -.8, label.y = 2.6)
                )

ggarrange(plotlist = p.list, nrow = 1)

ggsave("./Figs/scatter_halfsamples_HeLa.pdf", width = 12, height = 3)
```


Use the averages of the half-samples to compare the experimental samples

```{r fig.width=10, fig.asp=.35}
var.list1 <- list("LoLo", "LoLo", "HiHi")
var.list2 <- list("HiHi", "HiMi", "HiMi")

p.list <- map2( var.list1,
                var.list2, 
                ~ggplot(aligned.HeLa, aes(.data[[.x]], .data[[.y]]))+
                    geom_point()+
                    scale_y_log10()+
                    scale_x_log10()+
                    theme_classic()+
                    theme(panel.grid.major = element_line()) +
                    stat_cor(method = "pearson", label.x = -.8, label.y = 2.6)
                )

ggarrange(plotlist = p.list, ncol = 3)

ggsave("./Figs/scattter_experimental_HeLa.pdf", width = 9, height = 3)
```